---
title: "`r params$project_title`"
editor_options:
  chunk_output_type: console
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: no
params:
  organisation: Karolinska Institutet
  pi_mail: prin.cipal@ki.se
  pi_name: Prin Cipal
  project_title: Broliden_5325
  redmine_issue: '#5325'
  requester_mail: kristina.broliden@ki.se
  requester_name: Kristina Broliden
  staff_mail: supp.ortstaff@nbis.se
  staff_name: Supp Ortstaff
  staff_web: https://nbis.se/about/staff/supp-ortstaff/
---

```{r Setup, echo = FALSE}
knitr::opts_chunk$set(fig.width  = 10,
                      results    = "hold",
                      message    = FALSE,
                      warning    = FALSE)

setwd("/Users/vilkal/work/Brolidens_work/Projects/Gabriella_repo/reports/rmarkdown")
```

#Load libraries and other scripts

```{r Load data, message=FALSE, warning=FALSE, include=FALSE}
##################
# LOAD LIBRARIES #
##################
suppressWarnings({suppressMessages({suppressPackageStartupMessages({
library(igraph)
library(rafalib)
library(sva)
library(batchelor)
remotes::install_github("czarnewski/niceRplots",upgrade = "always",force = T)
library(niceRplots)
library(tidyverse)
})  })  })

#############
# LODA DATA #
#############
PATH <- "../../" #Path to the data

# Load metadata
dataset_names <- c("ASV_tissue_V3_B2.csv",   # Tissue, Boston run 2 (95 samples)
                     "ASV_tissue_V3_B1.csv", # Tissue, Boston run 1 (1 sample)
                     #"ASV_CVL_V2_B1.csv", # CVL V2, Boston run 1 (27 samples)
                     #"ASV_CVL_V2_B2.csv", # CVL V2, Boston run 2 (49 samples)
                     #"ASV_CVL_V2_S.csv",  # CVL V2, CTMR         (62 samples)
                     "ASV_CVL_V3_B1.csv") # CVL V3, Boston run 1 (111 samples)

datasets_ <- map(dataset_names, ~read.csv( paste0(PATH,"data/",.x))) %>% set_names(., dataset_names)
metadata <- read.csv(paste0(PATH,"data/Clinical_visit_2_3_updatefeb22.csv"))
rownames(metadata) <- as.character(metadata$ID)

#################
# COLOR PALETTS #
#################
pal <- c(RColorBrewer::brewer.pal(9,"Set1"),RColorBrewer::brewer.pal(8,"Set2"),
         RColorBrewer::brewer.pal(9,"Pastel1"),RColorBrewer::brewer.pal(8,"Pastel2")) #color pallete for plots
heat_pal <- c("#000000", colorRampPalette(c("#000000","grey5","grey30","orange3","yellow","yellow","white"))(90))
cor_pal <- colorRampPalette(c("navy","white","firebrick"))(90)
```

```{r Functions, include=FALSE}
# create "other" taxonomy
other_taxa.fun <- function(df) {
  other_taxa <- df %>%
  unite("taxa_other", Kingdom:Species, sep = ";", remove = F, na.rm = T) %>%
  mutate(taxa_other = ifelse(is.na(.$Species), paste0(.$taxa_other, ";other"), .$taxa_other)) %>%
  #mutate(taxa_other = str_replace(.$taxa_other, ";NA.+|;NA", ";other")) %>%
  mutate(taxa_other = str_extract(.$taxa_other, "([^;]+);([^;]+)$")) %>% # get the two last taxa levels
  mutate(taxa_other = sub('^(.*);(other)', '\\2;\\1', .$taxa_other)) # place "other" first
return(other_taxa)
}

# create genus lacto taxonomy
genus_lacto.fun <- function(df){
  genus_lacto <- df %>%
    mutate(Genus_lacto = .$Genus, .after=Seq_ID) %>% 
    mutate(Genus_lacto = ifelse(grepl("crispatus",.$Species), paste("L.", "crispatus/acidophilus"), 
                                ifelse(grepl("reuteri",.$Species), paste("L.", "reuteri/oris/frumenti/antri"),
                                  ifelse(grepl("gasseri",.$Species), paste("L.","gasseri/johnsonii/taiwanensis"),
                                    ifelse(grepl("murinus",.$Species), paste("L.","murinus/animalis/apodemi/salivarius"),
                                      ifelse(grepl("plantarum",.$Species), 
                                           paste("L.", "plantarum/fabifermentans/composti/paraplantarum/graminis/fuchuensis"),
                                        ifelse(grepl("Lactobacillus",.$Genus), paste("L.", .$Species), 
                                              as.character(.$Genus_lacto))))))) )
  return(genus_lacto)
} 

aggregate.fun <- function(df, level){
  level <- enquo(level)
  aggregated <- df %>%
    dplyr::rename(Taxonomy = !!level) %>%
    unite(., "Full_taxonomy", Kingdom:Species, sep = ";") %>%
    dplyr::select(-any_of(c("Sequence", "Seq_ID", "Genus_lacto", "taxa_other", "Full_taxonomy"))) %>%
    group_by(Taxonomy) %>%
    summarise(across(where(is.numeric), sum)) %>%
    ungroup()
return(aggregated)
}
```


# Taxonomy agglomeration

```{r Agglomerate Taxa, echo = FALSE}
temp <- map(datasets_, ~.x  %>%
            select(Sequence, Seq_ID, Kingdom:Species, everything()) %>%
            filter(!(is.na(.$Kingdom))) %>% # Removes any kingdom being NA
            other_taxa.fun(.) %>% # create "other" taxonomy 
            genus_lacto.fun(.)  # create genus lacto taxonomy
            ) %>%  
        {. ->> temp_ } %>% 
        map(., ~aggregate.fun(., "Genus_lacto")) # agglomerate on genus/lacto spp. lvl.


# Extract taxonomic information to file
seq_taxa <- bind_rows(temp_, .id = "dataset") %>% select(1:12) %>% unique() 
#write.csv(seq_taxa, paste0(PATH,"results/","SeqID_to_gen-lacto_tax-oth",".csv"),row.names = F )

# make long format raw non agglom ASV files
# remove_from_CVL_V2 <- intersect(colnames(temp_[[5]]), colnames(temp_[[3]]))[-c(1:11)]
# names <- c("ASV_tissue_V3_Boston run2","ASV_tissue_V3_Boston run1", "ASV_CVL_V2_Boston run1", 
#                      "ASV_CVL_V2_Boston run2", "ASV_CVL_V2_Stockholm", "ASV_CVL_V3_Boston run 1") 
# ASV_long <- temp_ %>%
#   modify_at(., c("ASV_CVL_V2_C.csv"), ~select(.x, -any_of(remove_from_CVL_V2))) %>% 
#   map(., ~pivot_longer(.x, matches("^P\\d"), names_to = "ID", values_to = "Reads")) %>%
#   set_names(names ) %>%
#   bind_rows(., .id = "dataset") %>%
#   separate(., dataset, into=c(NA, "Sample_type", "Visit","Place_sequenced"), sep = "_", remove = F) %>% 
#   mutate(dataset = str_extract(.$dataset, "CVL_V\\d|tissue_V3")) %>% 
#   group_split(dataset, .keep=F) %>% set_names(c("tissue", "CVL_V2", "CVL_V3"))
# 
# saveRDS(ASV_long, paste0(PATH,"results/","ASV_all_data_long_format.RDS"))

#arrange all datasets to have the same rownames with all bacteria
common_microbes_fun <- function(x) {
  t <- bind_rows(temp) %>%
    select(Taxonomy) %>%
    unique() %>%
    left_join(x, by="Taxonomy") %>%
    mutate(across(Taxonomy, ~replace_na(.x, "Not assigned"))) %>%
    mutate(across(-Taxonomy, ~replace_na(.x, 0))) %>%
    column_to_rownames(var = "Taxonomy")
  return(t)
}
datasets_ <- map(temp, ~common_microbes_fun(.x))
lapply(datasets_,dim)


# datasets <- datasets[3] #in case you wanna test in 1 dataset

# Transform phyloseq data into normalized matricies ( log2[CP1K] )
# datasets <- lapply(datasets,function(x){
#     x <- t(log2( t(x) / colSums(x) * 1000 + 1))
# })
# lapply(datasets,dim)

```

# Merging, renaming and correcting batch effects on datasets

```{r merge sequencing runs, echo = FALSE}
datasets <- datasets_
names(datasets)

#No batch correction needed in tissue_V3 dataset, only add it to the other samples
datasets[["ASV_tissue_V3"]] <- cbind( datasets[["ASV_tissue_V3_B2.csv"]],datasets[["ASV_tissue_V3_B1.csv"]] )

datasets <- datasets[-c(1,2)]
lapply(datasets,dim)


```


```{r save raw and norm microbiome files, echo = FALSE,fig.height=5,fig.width=5}

datasets <- lapply(datasets,function(x){
  x <- x[,sort(colnames(x))]
  return(x)
})

# raw files
n <- c("ASV_CVL_V3", "ASV_tissue_V3")
map2(datasets, n,  ~write.csv(.x, file=paste0(PATH,"results/",.y,"_raw_counts.csv"),row.names = T) )

# normalized files
datasets[["ASV_CVL_V3.csv"]] <- round( t(log2( t(datasets[["ASV_CVL_V3_B1.csv"]]) / colSums(datasets[["ASV_CVL_V3_B1.csv"]]) * 1000 + 1 )),4)
datasets[["ASV_tissue_V3"]] <- round(t( log2( t(datasets[["ASV_tissue_V3"]]) / colSums(datasets[["ASV_tissue_V3"]]) * 1000 + 1 )),4)

write.csv( datasets[["ASV_tissue_V3"]] , paste0(PATH,"results/ASV_tissue_V3_normalized_batch_corrected.csv"),row.names = T )
write.csv( datasets[["ASV_CVL_V3.csv"]] , paste0(PATH,"results/ASV_CVL_V3_normalized_batch_corrected.csv"),row.names = T )
```



```{r save raw and norm trx files, echo = FALSE}
trx <- read.csv(paste0(PATH,"data/TRX_rawcounts.csv"),row.names = 1)
trx <- rowsum(trx[,-c(1,2)],trx$symbol)
colnames(trx) <- sub("ML","",colnames(trx))
trx  <- trx[ rownames(trx) != "" , ]
trx <- trx[ rownames(trx) != "NA" , ]
colnames(trx) <- as.character(metadata$ID)[ match(colnames(trx) , as.character(metadata$PatID) ) ]
trx <- trx[ rowSums(trx > 5) >= 3 , ]
trx <- t( log2( t( trx ) / colSums(trx) * 1e6 + 1 ) )
write.csv( trx , paste0(PATH,"results/Tissue_RNAseq_V3_normalized.csv"),row.names = T )
```








