---
title: "`r params$project_title`"
editor_options:
  chunk_output_type: console
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: no
params:
  organisation: Karolinska Institutet
  pi_mail: prin.cipal@ki.se
  pi_name: Prin Cipal
  project_title: Broliden_5325
  redmine_issue: '#5325'
  requester_mail: kristina.broliden@ki.se
  requester_name: Kristina Broliden
  staff_mail: supp.ortstaff@nbis.se
  staff_name: Supp Ortstaff
  staff_web: https://nbis.se/about/staff/supp-ortstaff/
---

```{r Setup, echo = FALSE}
knitr::opts_chunk$set(fig.width  = 10,
                      results    = "hold",
                      message    = FALSE,
                      warning    = FALSE)
```

#Load libraries and other scripts

```{r, echo = FALSE}
library(igraph)
library(rafalib)
library(sva)
library(batchelor)
source("~/repos/niceRplots/R/plotting_functions.R")
source("~/repos/niceRplots/R/add_fig_label.R")
source("~/repos/niceRplots/R/helper_functions.R")

```

#Defining some variables for the analysis

```{r, echo = FALSE}
PATH <- "~/Desktop/NBIS/SMS_Projects/broliden_5325/" #Path to the data

#color palettes
pal <- RColorBrewer::brewer.pal(8,"Set1") #color pallete for plots
heat_pal <- c("#000000", colorRampPalette(c("#000000","grey5","grey30","orange3","yellow","yellow","white"))(90))
cor_pal <- colorRampPalette(c("navy","white","firebrick"))(90)


#Graph construction  
fct <- .5 # FC threshold for differential expression
pvt <- 0.01 # Pvalue threshold for differential expression
min_pct <- .1 # minimun level of detected bacteria in each sample group
```

# Loading data and metadata

```{r, echo = FALSE}
# Load metadata
metadata <- read.csv(paste0(PATH,"/data/Clinical_visit_2_3_updatesept28.csv"))
rownames(metadata) <- as.character(metadata$ID)

# Load datasets
dataset_names <- c("Tissue_RNAseq_V3_normalized",
                     "ASV_tissue_V3_normalized_batch_corrected",
                     "ASV_CVL_V3_normalized_batch_corrected",
                     "ASV_CVL_V2_normalized_batch_corrected",
                     "ASV_CVL_V2_normalized_NOT_batch_corrected")

#import datasets and return matrix with taxa-level resolution
datasets <- lapply(dataset_names, function(x){
  x <- read.csv( paste0(PATH,"results/",x,".csv"),row.names = 1)
  return(as.matrix(x)) })
names(datasets) <- dataset_names


#fill all samples with 0s to make plotting easier
datasets_all_samples <- lapply(datasets, function(x) {
  temp <- matrix(0,nrow = nrow(x),
                 ncol = nrow(metadata),
                 dimnames = list(rownames(x),rownames(metadata)))
  temp[rownames(x),colnames(x)] <- x
  return(temp)
})
lapply(datasets_all_samples,dim)


```

# Calculate QC metrics

```{r, echo = FALSE}
counts <- t(as.data.frame( lapply(datasets_all_samples,colSums) ))
counts[is.na(counts)] <- 0
detected_counts <- t(as.data.frame( lapply(datasets_all_samples,function(x) colSums( x > 0 )) ))
detected_counts[is.na(detected_counts)] <- 0
rownames(detected_counts) <- paste0(rownames(detected_counts),">0")

mypar(mar=c(2,20,2,1))
barlist( data = rbind( counts , detected_counts),
         genes = c(rownames(counts),rownames(detected_counts)),
         clustering = rep(1,nrow(metadata)),
         col=ifelse( (colSums(counts == 0)>=3)*1,"red","grey" ),draw_mean_lines=F)

sum((colSums(counts == 0)>=3)*1)
colnames(counts)[(colSums(counts == 0)>=3)]

metadata[colnames(counts)[(colSums(counts == 0)>=3)],]
```

# Comparing CVL2 samples before and after batch correction

```{r, echo = FALSE}
mypar(1,2,mar=c(2,15,1,1))

for( dataset in c("ASV_CVL_V2_normalized_NOT_batch_corrected",
            "ASV_CVL_V2_normalized_batch_corrected")){
  
  # Computing differential expression across batches
  all_microbiome <- cbind(datasets_all_samples[[dataset]])
  datasets <- read.csv(paste0(PATH,"/results/batches_CVL2.csv"))[,2]
  
  all_microbiome <- all_microbiome[,as.character(read.csv(paste0(PATH,"/results/batches_CVL2.csv"))[,1])]
  
  NN <- min(table(datasets))
  
  
  res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
  for(i in levels(datasets)){
    for(j in rownames(all_microbiome) ){
      
      set.seed(1)
      a <- c(sample(all_microbiome[j,datasets == i],NN))
      set.seed(1)
      b <- sample(all_microbiome[j,datasets != i],NN)
      
      perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
      perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
      
      temp <- wilcox.test(x=a, y=b,exact = F)
      
      fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
      
      res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                            c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
      colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
    }
  }
  

  res <- res[-1,]
  res$pvalue <- as.numeric(res$pvalue)
  res$perc.1 <- as.numeric(res$perc.1)
  res$perc.2 <- as.numeric(res$perc.2)
  res$perc.diff <- res$perc.1 - res$perc.2
  
  res$fc <- as.numeric(res$fc)
  res$FDR <- p.adjust(res$pvalue)
  res <- res[order(res$pvalue),]
  res <- res[res$fc > fct,]
  res <- res[abs(res$pvalue) < 0.05,]
  res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
  dim(res)
  
  ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))
  
  plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets, main = dataset,show_grid = T,cex.main=1,font.main=1,
            cex.row = .8,cex.col = .8,srt = 0)
}
```



```{r, echo = FALSE}
# Computing differential expression across datasets

all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V2_normalized_batch_corrected"]])
datasets <- factor(c( rep("tissue_v3",121) , rep("CVL_v3",121) , rep("CVL_v2",121) ))

NN <- min(table(datasets))


res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    set.seed(1)
    a <- c(sample(all_microbiome[j,datasets == i],NN))
    set.seed(1)
    b <- sample(all_microbiome[j,datasets != i],NN)
    
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
dim(res)

ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

mypar(mar=c(2,15,1,1))
plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.row = .8,cex.col = .8,srt = 0)

```


# Correlation CVL3 and RNAseq

```{r, echo = FALSE}

TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
CVL3 <- datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]]
samples_TRX_and_CVL3 <- colnames(counts) [ colSums(counts[c("Tissue_RNAseq_V3_normalized","ASV_CVL_V3_normalized_batch_corrected"),] > 0)==2 ]

#filter TRX dataset
TRX <- TRX [ , samples_TRX_and_CVL3 ]
TRX <- TRX [ rowSums(TRX>0)>= 2 , ]
top_vars_TRX <- names(sort(apply(TRX,1,var),decreasing = T)[1:3000] )
plot(apply(TRX,1,mean),apply(TRX,1,var),col=ifelse(rownames(TRX) %in% top_vars_TRX,"red","grey20"),cex=.2)
TRX <- TRX[ top_vars_TRX , ]
dim(TRX)

#filter CVL3 dataset
CVL3 <- CVL3 [ , samples_TRX_and_CVL3 ]
CVL3 <- CVL3 [ rowSums(CVL3>0)>= 2 , ]
top_vars_CVL3 <- names(sort(apply(CVL3,1,var),decreasing = T)[1:100] )
dim(CVL3)


mypar()
cors <- cor( t(rbind(TRX) ) ,  t(rbind(CVL3) ) )
hist(cors,xlim=c(-1,1),breaks = 100)
dim(cors)


library(fgsea)
bacteria_use <- colnames(cors)
gmt <- gmtPathways("../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
enrichments <- lapply(bacteria_use,function(x){
  res <- fgsea(pathways = gmt,stats = sort(cors[,x],decreasing = T),nper=1000)
  return(res)
})
names(enrichments) <- bacteria_use

pvalues <- lapply(enrichments,function(x) setNames(x$pval,x$pathway) )
pvalues <- as.data.frame(pvalues)
pvalues <- -log10( pvalues )
top_pathways <- names(sort(apply(pvalues,1,median),T))[1:50]
mypar(mar=c(4,15,2,2))
boxplot( t(pvalues[rev(top_pathways),]),cex=.1,horizontal=T,las=1 )

pvalues[ pvalues < -log10(0.05)] <- 0
sort(rowSums(pvalues>0),T)
sort(colSums(pvalues>0),T)

NESes <- lapply(enrichments,function(x) setNames(x$NES,x$pathway) )
NESes <- t(as.data.frame(NESes))
# 
# o_kegg <- hclust( as.dist( (1-cor(NESes))/2 ),"ward.D2")$order
# o_bacs <- hclust( as.dist( (1-cor(t(NESes)))/2 ),"ward.D2")$order
o_kegg <- hclust( dist( t(NESes) ),"ward.D2")$order
o_bacs <- hclust( dist( NESes ),"ward.D2")$order
NESes <- NESes[o_bacs,o_kegg]

mypar(mar=c(15,2,2,15))
image( t(NESes[nrow(NESes):1,]),col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-4,4,length.out = 92),
       axes=F,border=NA,main="KEGG pathways",xlab="",ylab="bacterias",line=.4,cex.main=1,font.main=1)
text(  par("usr")[c(4)]+.02 , seq(1,0,length.out = nrow(NESes)),
      labels = rownames(NESes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=.3)
text( seq(0,1,length.out = nrow(NESes)) , par("usr")[c(1)]-.02,
      labels = colnames(NESes), srt = 90, adj = c(1,.5), xpd = TRUE, cex=.3)



eee <- data.frame(x=c(sapply(colnames(NESes),function(x)rep(x,nrow(NESes)))),
                  y=rep(rownames(NESes),ncol(NESes)),
                  w=c(NESes),
                  p=c(t(pvalues)),
                  x1=0,
                  y1=c(sapply(ncol(NESes):1,function(x)rep(x,nrow(NESes))))/ncol(NESes) ,
                  x2=1,
                  y2=rep(nrow(NESes):1,ncol(NESes))/nrow(NESes) )

eee$w2 <- eee$w / max(eee$w)
eee <- eee[eee$w > 1.5,]
eee <- eee[eee$p > 2,]

eee$y1 <- rank(eee$y1) / length(eee$y1)
eee$y2 <- rank(eee$y2) / length(eee$y2)




plot(c(0,1),c(0,1),type="n")
for( i in 1:nrow(eee)){
  lines(c(eee$x1[i],eee$x2[i]),c(eee$y1[i],eee$y2[i]),
        col=colorRampPalette(c("grey95","grey","firebrick"))(99)[ eee$w2 * 98 + 1 ] )
}




```






```{r, echo = FALSE}

g <- graph_from_incidence_matrix( 1-cors, directed = F,weighted = T  )

l <- layout_with_fr(g)
plot( g , vertex.label.cex=0.000001 , layout=l,
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,
      vertex.size=1,vertex.color=ifelse(V(g)$name %in% colnames(CVL3),"red","blue") )

```




