---
title: "`r params$project_title`"
editor_options:
  chunk_output_type: console
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: no
params:
  organisation: Karolinska Institutet
  pi_mail: prin.cipal@ki.se
  pi_name: Prin Cipal
  project_title: Broliden_5325
  redmine_issue: '#5325'
  requester_mail: kristina.broliden@ki.se
  requester_name: Kristina Broliden
  staff_mail: supp.ortstaff@nbis.se
  staff_name: Supp Ortstaff
  staff_web: https://nbis.se/about/staff/supp-ortstaff/
---

```{r Setup, echo = FALSE}
knitr::opts_chunk$set(fig.width  = 10,
                      results    = "hide",
                      message    = FALSE,
                      warning    = FALSE)
#setwd("/Users/vilkal/work/Brolidens_work/Projects/Gabriella_repo/reports/rmarkdown")
```

#Load libraries and other scripts

```{r, echo = FALSE}
library(igraph)
library(rafalib)
library(fgsea)
library(scales)
remotes::install_github("czarnewski/niceRplots",upgrade = "always",force = T)
library(niceRplots)
library(tidyverse)
```

#Defining some variables for the analysis

```{r, echo = FALSE}
PATH <- "../../" #Path to the data
Set1 <- c("#4DAF4A", "#377EB8", "#E41A1C", "#FF7F00", "#984EA3", "#FFFF33", "#A65628", "#F781BF", "#999999")

#color palettes
pal <- c(Set1,RColorBrewer::brewer.pal(8,"Set2"),
         RColorBrewer::brewer.pal(9,"Pastel1"),RColorBrewer::brewer.pal(8,"Pastel2")) #color pallete for plots
heat_pal <- c("#000000", colorRampPalette(c("#000000","grey5","grey30","orange3","yellow","yellow","white"))(90))
cor_pal <- colorRampPalette(c("navy","white","firebrick"))(90)


#Graph construction  
fct <- .25 # FC threshold for differential expression
pvt <- 0.01 # Pvalue threshold for differential expression
min_pct <- .05 # minimun level of detected bacteria in each sample group

```

# Loading data and metadata

```{r, echo = FALSE}
# Load meta data
meta <- read.csv(paste0(PATH,"/data/Clinical_visit_2_3_updatefeb22.csv"))
rownames(meta) <- as.character(meta$ID)

# Load datasets
dataset_names <- c("Tissue_RNAseq_V3_normalized",
                     "ASV_tissue_V3_normalized_batch_corrected",
                     "ASV_CVL_V3_normalized_batch_corrected",
                     "ASV_CVL_V2_normalized_NOT_batch_corrected")


#import datasets and return matrix with genus_lacto level resolution
# datasets <- map(dataset_names, 
#                 ~read.csv(paste0(PATH,"results/",.x, ".csv"),row.names = 1)) %>%
#   set_names(., dataset_names)


#fill all samples with 0s to make plotting easier
#x <- datasets[[2]]
# all_samples <- unique(unlist(lapply(datasets,colnames)))
# datasets_all_samples <- lapply(datasets, function(x) {
#   col <- setdiff(all_samples, colnames(x))
#   temp <- tibble(t = rownames(x))  %>% 
#      tibble::add_column(!!!set_names(as.list(rep(0, length(col))),nm=col)) %>% 
#     column_to_rownames(., var = "t") %>% bind_cols(x) 
#   return(temp)
# })

#import datasets and return matrix with genus_lacto level resolution
datasets <- lapply(dataset_names, function(x){
  x <- read.csv( paste0(PATH,"results/",x,".csv"),row.names = 1)
  return(as.matrix(x)) })
names(datasets) <- dataset_names


# Define what samples to use
samples_to_use <- intersect(colnames(datasets[["ASV_CVL_V3_normalized_batch_corrected"]]),
                            colnames(datasets[["Tissue_RNAseq_V3_normalized"]]))


#fill all samples with 0s to make plotting easier
x <- datasets[[1]]
datasets_all_samples <- lapply(datasets, function(x) {
  temp <- matrix(0,nrow = nrow(x),
                 ncol = nrow(meta),
                 dimnames = list(rownames(x),rownames(meta))
                 )
  temp[rownames(x),colnames(x)] <- x
  return(temp)
})
lapply(datasets_all_samples,dim)

# Clean up metadata
tissue_gr <- read.csv(paste0(PATH,"/data/210609_TissuegroupsJune.csv"))
metadata <- meta %>%
  left_join(tissue_gr, by="ID") %>%
  mutate(Luminal_names = ifelse(.$ModifiedM4 == "NonIners", "L. crispatus/jensenii",
                        ifelse(.$ModifiedM4 == "Iners", "L. iners",
                          ifelse(.$ModifiedM4 == "Gard", "Gardnerella",
                            ifelse(.$ModifiedM4 == "Prev", "High diverse", 
                                    "Other")))) ) %>%
  mutate(Luminal_gr = ifelse(.$ModifiedM4 == "NonIners", "L1",
                        ifelse(.$ModifiedM4 == "Iners", "L2", 
                          ifelse(.$ModifiedM4 == "Gard", "L3",
                            ifelse(.$ModifiedM4 == "Prev", "L4", 
                                    "L5")))) ) %>%
  mutate(Tissue_names = ifelse(.$TissuegroupsJune == "Crisp", "L. crispatus/jensenii",
                        ifelse(.$TissuegroupsJune == "Iners", "L. iners", 
                          ifelse(.$TissuegroupsJune == "Gard", "Gardnerella",
                            ifelse(.$TissuegroupsJune == "Prev", "High diverse", 
                                    "Other")))) ) %>%
  mutate(Tissue_gr = ifelse(.$TissuegroupsJune == "Crisp", "T1",
                        ifelse(.$TissuegroupsJune == "Iners", "T2", 
                          ifelse(.$TissuegroupsJune == "Gard", "T3",
                            ifelse(.$TissuegroupsJune == "Prev", "T4", 
                                    "T5")))) ) %>%
  select(ID:Yeast_v3, group_M9_CVLv3, Luminal_names, Tissue_names, Luminal_gr,Tissue_gr) 

rownames(metadata) <- as.character(metadata$ID)
```

# Calculate QC metrics

\clearpage

```{r, echo = FALSE,fig.width=10,fig.height=5}
counts <- t(as.data.frame( lapply(datasets_all_samples,colSums) ))
counts[is.na(counts)] <- 0
detected_counts <- t(as.data.frame( lapply(datasets_all_samples,function(x) colSums( x > 0 )) ))
detected_counts[is.na(detected_counts)] <- 0
rownames(detected_counts) <- paste0(rownames(detected_counts),">0")

mypar(2,1,mar=c(2,10,2,1))
barlist( data = rbind( counts ),main = "total counts",
         genes = c(rownames(counts)), labels = sub("_norm.*","",rownames(counts)),
         clustering = rep(1,nrow(metadata)),
         col=ifelse( (colSums(counts == 0)==3)*1,"red", ifelse( (colSums(counts == 0)==2)*1,"orange3","grey" ) ),draw_mean_lines=F)
add_letter("a")

barlist( data = rbind( detected_counts),main = "detected features",
         genes = c(rownames(detected_counts)),labels = sub("_norm.*","",rownames(counts)),
         clustering = rep(1,nrow(metadata)),
         col=ifelse( (colSums(counts == 0)==3)*1,"red", ifelse( (colSums(counts == 0)==2)*1,"orange3","grey" ) ),draw_mean_lines=F)
add_letter("b")

sum((colSums(counts == 0)>=3)*1)
colnames(counts)[(colSums(counts == 0)==3)]
paste0(colnames(counts)[(colSums(counts == 0)==3)],collapse = ",")

metadata[colnames(counts)[(colSums(counts == 0)>=3)],]

datasets_all_samples <- lapply( datasets_all_samples , function(x) { x [ , samples_to_use] } )
metadata <- metadata[ samples_to_use , ]


detected_counts <- detected_counts[,samples_to_use]
barlist( data = rbind( detected_counts),main = "detected features",
         genes = c(rownames(detected_counts)),labels = sub("_norm.*","",rownames(counts)),
         clustering = rep(1,nrow(metadata)),
         col="grey",draw_mean_lines=F)
add_letter("b")

saveRDS(datasets_all_samples,"../../results/datasets_all_samples.RDS")
```

**Figure 1.** Comparative barplot for the **a)** total counts and **b)** number of non-zero detected features (genes / bacteria) for each of the sequencing datasets. Samples are ordered alfabetically according to the patient ID. Samples that are present in exactly two datasets are shown in orange (`r paste0(colnames(counts)[(colSums(counts == 0)==2)],collapse = ",")`). Samples that are present in exactly two datasets are shown in red (`r paste0(colnames(counts)[(colSums(counts == 0)==3)],collapse = ",")`).


```{r, echo = FALSE,fig.width=10,fig.height=9}
# all_microbiome <- rbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
#                         datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
#                         datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])
all_microbiome <- datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]]
# all_microbiome <- t( t(all_microbiome) / colSums(all_microbiome) )
all_microbiome <- all_microbiome[ rowSums(all_microbiome>0) >= 3 ,  ]
# all_microbiome <- all_microbiome + rnorm( length(all_microbiome) , mean = 0, sd = .00000001)
dim(all_microbiome)

zscore <- t(apply(all_microbiome,1,function(i){scale(i,T,T)}))
zscore[ zscore > 4 ] <- 4
zscore[ zscore < -4 ] <- -4
rownames(zscore) <- rownames(all_microbiome); colnames(zscore) <- colnames(all_microbiome)
mypar(mar=c(2,15,2,2))
top_var <- apply(all_microbiome,1,var)
top_var_t <- sort(top_var,decreasing = T)[100]
is_top_var <- top_var > top_var_t 
barplot(sort(top_var[is_top_var],decreasing = T)[30:1],horiz = T,las=1)

#Compute sample-wise correlations
cors <- cor(all_microbiome[,], method = "pearson")
adj <- (1-cors)/2
# pheatmap::pheatmap(adj,border_color = NA,color = colorRampPalette(c("grey90","grey70","grey70","grey70","orange3","firebrick"))(90) )

# cors <- cor(all_microbiome, method = "pearson")
# adj <- (1-cors)/2
# pheatmap::pheatmap(adj,border_color = NA,color = colorRampPalette(c("grey90","grey70","grey70","grey70","orange3","firebrick"))(90) )

#Graph construction  
k <- 12 # k for KNN
p <- 0.1 # prunning threshold of SNN 0.14


#Compute KNN graph from correlation matrix
knn <- RANN::nn2(adj,k = k)
nn <- matrix(0, nrow(knn$nn.idx), nrow(knn$nn.idx))
for(idx in 1:nrow(knn$nn.idx) ){  nn[idx,knn$nn.idx[idx,]] <- 1 }

#Computing SNN matrix
N <- matrix(1,ncol = ncol(nn),nrow = nrow(nn))
P <- t(nn) %*% nn
B <- N - nn
C <- t(B) %*% B
Q <- nrow(nn)*N - C
J <- P/Q
J[is.nan(J)] <- 0
J[J < p] <- 0
colnames(J) <- colnames(all_microbiome)
rownames(J) <- colnames(all_microbiome)

#Transforming SNN into an igraph object
g <- graph_from_adjacency_matrix(J,mode = "undirected",diag = F,weighted = T)
write.csv(J,"../../results/patient_SNN_graph.csv",row.names = T)

mypar(3,3,mar=c(0,0,2,0))

set.seed(1)
set.seed(1)
l <- layout_with_fr(g,niter=3000,start.temp=30)

cl <- igraph::cluster_louvain(g)
# cl <- igraph::walktrap.community(g)

metadata$Louvain_clusters <- cl$membership
metadata$joint_clustering <- metadata$Luminal_gr
metadata$joint_clustering <- factor(metadata$joint_clustering) 
set.seed(1)
U <- uwot::umap(adj,n_neighbors = 30)
h <- hclust(as.dist(adj),method = "ward.D2")
# metadata$joint_clustering <- cutree(h,k = 5)
write.csv2(metadata,"../../results/metadata_integration.csv",row.names = T)

plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(cl$membership)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=l,
      vertex.size=10,main="FR embedding\nlouvain",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )

plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(cl$membership)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=U,
      vertex.size=10,main="umap",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
empty_plot()
legend("topleft",title = "Patient groups",
       legend = levels(factor(cl$membership)),
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1.2,xpd=T)



plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(metadata$group_M9_CVLv3)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=l,
      vertex.size=10,main="FR embedding",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(metadata$group_M9_CVLv3)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=U,
      vertex.size=10,main="umap",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
empty_plot()
legend("topleft",title = "Patient groups\ngroup_M9_CVLv3",
       legend = levels(factor(metadata$group_M9_CVLv3)),
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1.2,xpd=T)
# title("Louvain clusters",cex.main=1,font.main=1)


plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(metadata$Luminal_gr)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=l,
      vertex.size=10,main="FR embedding",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(metadata$Luminal_gr)] , 
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=U,
      vertex.size=10,main="umap",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
empty_plot()
legend("topleft",title = "Patient groups\nLuminal_gr",
       legend = levels(factor(metadata$Luminal_gr)),
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1.2,xpd=T)
# title("Louvain clusters",cex.main=1,font.main=1)

```


**Figure 1b.** Sample embedding of `r k`-SNN graph clustered using Louvain based on the normalized bacterial abundances across all microbiome datasets (including samples not present in all datasets).




```{r}
mypar(mar=c(3.2,15,2,2))

barlist( data = rbind( detected_counts),main = "detected features",
         genes = c(rownames(detected_counts)),labels = sub("_norm.*","",rownames(counts)),
         clustering = metadata$joint_clustering, 
         col=pal[factor(metadata$joint_clustering)],draw_mean_lines=F)
add_letter("b")


barlist( data = all_microbiome, main = "detected features",
         genes = c("Gardnerella",
                   "Lactobacillus iners",
                   "Lactobacillus jensenii",
                   "Lactobacillus crispatus/acidophilus",
                   "Prevotella",
                   "Atopobium"),
         clustering = metadata$joint_clustering,
         col=pal[factor(metadata$joint_clustering)], draw_mean_lines=F) # ,srt = 65
#legend(0,legend = levels(factor(metadata$Luminal_gr)))

# barlist( data = all_microbiome, main = "detected features",
#          genes = c("Gardnerella vaginalis",
#                    "Lactobacillus iners",
#                    "Lactobacillus jensenii",
#                    "Lactobacillus crispatus/acidophilus",
#                    "Prevotella amnii",
#                    "Prevotella timonensis",
#                    "Atopobium vaginae"),
#          clustering = metadata$joint_clustering,
#          col=pal,draw_mean_lines=F)
```

\clearpage


```{r, echo = FALSE,fig.width=5,fig.height=5}
x <- datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]]
xL <- rowsum(x , grepl("L.",rownames(x)) )[2,] / sum(grepl("L.",rownames(x)))
xM <- rowsum(x , grepl("Mobiluncus",rownames(x)) )[2,] / sum(grepl("Mobiluncus",rownames(x)))
xG <- rowsum(x , grepl("Gardnerella",rownames(x)) )[2,] / sum(grepl("Gardnerella",rownames(x)))

mL <- metadata$BV_Lactobacillus_v3 ; mL[is.na(mL)] <- 0
mM <- metadata$BV_Monbilicus_v3    ; mM[is.na(mM)] <- 0
mG <- metadata$BV_Vaginal_Garda_v3 ; mG[is.na(mG)] <- 0

mypar(3,1,mar=c(2,6,2,2))
barlist( data = rbind(  BV=mL ,
                        counts_16S=xL),
         main = "Lactobacillus",
         genes = c("BV","counts_16S"),
         clustering = metadata$joint_clustering, #srt = 60,
         draw_mean_lines=F,col = pal[factor(metadata$joint_clustering)])

barlist( data = rbind(  BV=mM ,
                        counts_16S=xM),
         main = "Monbilicus",
         genes = c("BV","counts_16S"), #srt = 60,
         clustering = metadata$joint_clustering,
         draw_mean_lines=F,col = pal[factor(metadata$joint_clustering)])

barlist( data = rbind(  BV=mG ,
                        counts_16S=xG),
         main = "Gardnerella",
         genes = c("BV","counts_16S"), #srt = 60,
         clustering = metadata$joint_clustering,
         draw_mean_lines=F,col = pal[factor(metadata$joint_clustering)])
```




\clearpage

```{r, eval=T,echo = FALSE,fig.width=10,fig.height=4}
myletters <- letters

mypar(1,2,mar=c(2,15,1,1))

for (m in c("ASV_tissue_V3_normalized_batch_corrected",
            "ASV_CVL_V3_normalized_batch_corrected",
            "ASV_CVL_V2_normalized_NOT_batch_corrected")){
  
  all_microbiome <- datasets_all_samples[[m]]
  
  datasets <- factor(metadata$joint_clustering)
  # datasets <- datasets[ colSums(all_microbiome)>0 ]
  # all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]
  NN <- min(table(datasets))
  
  res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
  for(i in levels(datasets)){
    for(j in rownames(all_microbiome) ){
      
      set.seed(1)
      a <- c(sample(all_microbiome[j,datasets == i],NN))
      set.seed(1)
      b <- sample(all_microbiome[j,datasets != i],NN)
      
      perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
      perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
      
      temp <- wilcox.test(x=a, y=b)
      
      fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
      
      res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                                 c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
      colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
    }
  }
  res <- res[-1,]
  res$pvalue <- as.numeric(res$pvalue)
  res$perc.1 <- as.numeric(res$perc.1)
  res$perc.2 <- as.numeric(res$perc.2)
  res$perc.diff <- res$perc.1 - res$perc.2
  
  res$fc <- as.numeric(res$fc)
  res$FDR <- p.adjust(res$pvalue)
  # res <- res[order(res$pvalue),]
  res <- res[abs(res$fc) > fct,]
  res <- res[abs(res$pvalue) < 0.05,]
  res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
  dim(res)
  

  
  ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

  plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
            cex.row = .8,cex.col = .8,srt = 0,main = m)
  add_letter(myletters); myletters <- myletters[-1]
  
  barlist( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
            cex.axis = .8,srt = 0,main = m)
  add_letter(myletters); myletters <- myletters[-1]

}
```


**Figure 2.** Differential bacterial abundance across joint clustered samples and compared across microbiome datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` were considered significant and were sorted by the highest expression.


\clearpage

```{r, eval=T,echo = FALSE,fig.width=10,fig.height=9}
all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])
datasets <- factor(c( rep("tissue_v3",ncol(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]])) , rep("CVL_v3",ncol(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])) , rep("CVL_v2",ncol(datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])) ))
datasets <- datasets[ colSums(all_microbiome)>0 ]
all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

NN <- min(table(datasets))

res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    a <- all_microbiome[j,datasets == i]
    b <- all_microbiome[j,datasets != i]
    
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
dim(res)

ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

mypar(1,2,mar=c(2,15,1,1))
plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.row = .8,cex.col = .8,srt = 0)
add_letter("a")
barlist( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.axis = .8,srt = 0)
add_letter("b")
```

**Figure 2.** Differential bacterial abundance across microbiome datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` were considered significant and were sorted by the highest expression.


\clearpage

```{r, eval=T,echo = FALSE,fig.width=10,fig.height=9}
all_microbiome <- cbind(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])
datasets <- factor(c( rep("CVL_v3",ncol(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])) , rep("CVL_v2",ncol(datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])) ))
datasets <- datasets[ colSums(all_microbiome)>0 ]
all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

NN <- min(table(datasets))


res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    a <- all_microbiome[j,datasets == i]
    b <- all_microbiome[j,datasets != i]
    
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
dim(res)

ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

mypar(1,2,mar=c(2,15,1,1))
plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.row = .8,cex.col = .8,srt = 0)
add_letter("a")
barlist( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.axis = .8,srt = 0)
add_letter("b")
```

**Figure 2.** Differential bacterial abundance across CVL2 and CVL3 microbiome datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` were considered significant and were sorted by the highest expression.


\clearpage

```{r, eval=T,echo = FALSE,fig.width=10,fig.height=9}
all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])
datasets <- factor(c( rep("tissue_v3",ncol(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]])) , rep("CVL_v3",ncol(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])) ))
datasets <- datasets[ colSums(all_microbiome)>0 ]
all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

NN <- min(table(datasets))


res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    a <- all_microbiome[j,datasets == i]
    b <- all_microbiome[j,datasets != i]
    
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
dim(res)

ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

mypar(1,2,mar=c(2,15,1,1))
plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.row = .8,cex.col = .8,srt = 0)
add_letter("a")
barlist( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.axis = .8,srt = 0)
add_letter("b")
```

**Figure 2b.** Differential bacterial abundance across tissue and CVL microbiome week3 datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` were considered significant and were sorted by the highest expression.


\clearpage

# Computing differential expression across microbiome datasets

```{r, echo = FALSE,fig.width=10,fig.height=9}
all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])
datasets <- factor(c( rep("tissue_v3",ncol(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]])) , rep("CVL_v3",ncol(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])) , rep("CVL_v2",ncol(datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])) ))

datasets <- datasets[ colSums(all_microbiome)>0 ]
all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

NN <- min(table(datasets))
```


```{r, eval=F,echo = FALSE,fig.width=10,fig.height=12}
res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    a <- all_microbiome[j,datasets == i]
    b <- all_microbiome[j,datasets != i]
    
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i)
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
dim(res)

ord <- factor(sapply(unique(as.character(res$bacteria)),function(x){getcluster(all_microbiome, x, datasets)}))

mypar(1,1,mar=c(5,15,1,1))
plot_dots( all_microbiome , names(sort(ord)) , clustering = datasets,show_grid = T,cex.main=1,font.main=1,
          cex.row = .7,cex.col = .8,srt = 30)
add_letter("a")
```

**Figure 4.** Differential bacterial abundance across all groups and all microbiome datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` were considered significant and were sorted by the highest expression.

```{r, echo = FALSE}
# groupings <- lapply( names(datasets_all_samples)[-1] ,function(x){
#   x <- read.csv2( paste0(PATH,"results/Sample_Louvain_clusters_",x,".csv"),row.names = 1 ) })
# names(groupings) <- names(datasets_all_samples)[-1]
# all_names <- unique(unlist(lapply(groupings,function(x)rownames(x))))
# 
# groupings_all_samples <- lapply(groupings,function(x){
#   df <- data.frame(group=rep("NA",length(all_names)), row.names = all_names,stringsAsFactors = F)
#   df[rownames(x),1] <- as.character(x[,1])
#   return(df)
# })
# names(groupings_all_samples) <- names(groupings)
# 
# df <- as.data.frame(groupings_all_samples)
# colnames(df) <- paste0( "louvain_",sub( "ASV_","",  sub( "_norma.*","", names(groupings_all_samples) )))
# metadata <- cbind(metadata,df[rownames(metadata),])


Louvain_clusters <- bind_cols(ID = cl$names, louvain_CVL_V3 = cl$membership)

df <- metadata %>%
  left_join(., Louvain_clusters, by="ID")

```

```{r, eval=F, echo = FALSE,fig.width=10, fig.height=10}
sample_order <- order(df$louvain_CVL_V3)

# zscore_datasets_all_samples <- lapply(datasets_all_samples,function(x){
#   zscore <- t(apply(x,1,function(i){scale(i,T,T)}))
#   zscore[ zscore > 6 ] <- 6
#   zscore[ zscore < -6 ] <- -6
#   rownames(zscore) <- rownames(x); colnames(zscore) <- colnames(x)
#   zscore[is.nan(zscore)] <- -6
#   return(zscore)
# })

mypar(3,1,mar=c(1,2,1,0.5))

image(t(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]]
        [ names(sort(ord)),sample_order]),
      col=heat_pal,axes=F,border=NA,
      main="tissue week 3",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
points( (0:(length(df$louvain_tissue_V3)-1) )/(length(df$louvain_tissue_V3)-1),
        rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_tissue_V3)), pch=16,
        col=pal[factor(df$louvain_tissue_V3)][sample_order] ,xpd=T)
points( (0:(length(df$louvain_tissue_V3)-1) )/(length(df$louvain_tissue_V3)-1),
        rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_tissue_V3)), pch=16,
        col=pal[factor(df$louvain_tissue_V3)][sample_order] ,xpd=T)


image(t(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]]
        [ names(sort(ord)),sample_order]),
      col=heat_pal,axes=F,border=NA,
      main="CVL week 3",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
points( (0:(length(df$louvain_CVL_V3)-1) )/(length(df$louvain_CVL_V3)-1),
        rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_CVL_V3)), pch=16,
        col=pal[factor(df$louvain_CVL_V3)][sample_order] ,xpd=T)
points( (0:(length(df$louvain_CVL_V3)-1) )/(length(df$louvain_CVL_V3)-1),
        rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_CVL_V3)), pch=16,
        col=pal[factor(df$louvain_CVL_V3)][sample_order] ,xpd=T)


image(t(datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]]
        [ names(sort(ord)),sample_order]),
      col=heat_pal,axes=F,border=NA,
      main="CVL week 2",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
points( (0:(length(df$louvain_CVL_V2)-1) )/(length(df$louvain_CVL_V2)-1),
        rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_CVL_V2)), pch=16,
        col=pal[factor(df$louvain_CVL_V2)][sample_order] ,xpd=T)
points( (0:(length(df$louvain_CVL_V2)-1) )/(length(df$louvain_CVL_V2)-1),
        rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
            length(df$louvain_CVL_V2)), pch=16,
        col=pal[factor(df$louvain_CVL_V2)][sample_order] ,xpd=T)


# **Figure 5.** Comparisson of microbiome datasets, showing only the significant bacteria. Samples are ordered by the CVL3 groupings. The colors represent their respective bacterial groupings for each dataset. 

```


\clearpage

```{r, echo = FALSE,fig.width=8,fig.height=8}
metadata_parameters <- c("Antibiotics","Antibiotics_v2","Antibiotics_v3",
                         "Bacterio_KOH_v2","Bacterio_KOH_v3","Bacterio_TV_v3",
                         "BV_Clue_Cells_v2","BV_Clue_Cells_v3",
                         "BV_Diagnosis_v2","BV_Diagnosis_v3",
                         "BV_Lactobacillus_v2","BV_Lactobacillus_v3",
                         "BV_Monbilicus_v2","BV_Monbilicus_v3",
                         "BV_Vaginal_Garda_v2","BV_Vaginal_Garda_v3",
                         "blood_cervix_v2","blood_cervix_v3",
                         "cervical_inflammation_v2","cervical_inflammation_v3",
                         "BV_WBC_v2", "BV_WBC_v3","marital_status",
                         "HIVstatus","Contraception",
                         "CT_v2","CT_v3","regular_partner",
                         "vaginal_douching_v2","vaginal_douching_v3","Luminal_gr", "Tissue_gr", 
                         "Louvain_clusters")

Ftest <- sapply(metadata_parameters,function(x){
  cat(paste0("\n",x,"\n"))
  temp <- sapply(metadata_parameters,x=x,function(x,y){
    if (x==y){
      return( 1 )
    } else {
      # if( !is.na(sum(as.numeric(metadata[,x]))) ){ f1 <- cut(as.numeric(metadata[,x]),breaks = 5 ) }
      if( is.numeric(metadata[,x]) ){ f1 <- cut(metadata[,x],breaks = 5 ) }
    f1 <- as.character(metadata[,x])
    f1[is.na(f1)] <- "NA"
    f1[f1==""] <- "NA"
    
    # if( !is.na(sum(as.numeric(metadata[,y]))) ){ f1 <- cut(as.numeric(metadata[,y]),breaks = 5 ) }
    if( is.numeric(metadata[,y]) ){ f2 <- cut(metadata[,y],breaks = 5 ) }
    f2 <- as.character(metadata[,y])
    f2[is.na(f2)] <- "NA"
    f2[f2==""] <- "NA"
    
    f11 <- f1 [ (f1!="NA") & (f2!="NA") ]
    f22 <- f2 [ (f1!="NA") & (f2!="NA") ]
    contT1  <- table(list(factor(f11,levels=unique(f1)),factor(f22,levels=unique(f2))))
    cat(".")
    f <- NULL
    try(f <- fisher.test(contT1,simulate.p.value = T),silent = T)
    if( is.null(f) ){ 
      return( 1  ) 
      } else {
      return( c(f$p.value)  ) 
    }
    # return( chisq.test(contT1,simulate.p.value = T)$p.value )
    }
  })
  
  return( temp )
})

clean_Ftest <- Ftest
# clean_Ftest <- ( clean_Ftest + t(clean_Ftest) ) / 2
clean_Ftest <- -log10(clean_Ftest)
clean_Ftest[clean_Ftest > 6] <- 6
# clean_Ftest[ clean_Ftest > 0.05 ] <- 1

write.csv2(Ftest , "../../results/metadata_association_results_complete.csv",row.names = T)
write.csv2(clean_Ftest , "../../results/metadata_association_results_filtered.csv",row.names = T)

mypar(1,1,mar=c(12,1,1,12))
image( t(clean_Ftest [nrow(Ftest):1,]),axes=F,
       col=colorRampPalette(c("grey95","grey70","orange3","firebrick"))(90) )
text(  par("usr")[2]+.02 ,seq(1,0,length.out = nrow(Ftest)) , 
       labels = rownames(Ftest), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
text( seq(0,1,length.out = nrow(Ftest)) , par("usr")[3]-.02 , 
      labels = rownames(Ftest), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
```


**Figure 6.** Association analysis across several patient categorical metadata parameters, including patient groupings annotations from microbiome.

\clearpage


```{r, echo = FALSE,fig.width=10,fig.height=10}
mypar(6,4,mar=c(1,4,1,4) )
x <- as.character(metadata[,"Contraception"])
x[is.na(x)] <- "NA"
y <- as.character(metadata[,"Antibiotics"])
y[is.na(y)] <- "NA"
x1 <- x [ (x!="NA") & (y!="NA") ]
y1 <- y [ (x!="NA") & (y!="NA") ]
plot_sankey( data.frame(x1,y1),color_by = 2,
             plot_labels = T, plot_weights = F,order_1_by = "total",
             order_2_by = "total",use_w1 = T,use_w2 = T,gap2v = 0,smoothing = .1,
             pal=hue_pal()(10))
text(c(0,1),c(1,1),labels = c( "Contraception" , "Antibiotics" ),pos=3,xpd=T)



x <- as.character(metadata[,"BV_Monbilicus_v2"])
x[is.na(x)] <- "NA"
y <- as.character(metadata[,"Antibiotics"])
y[is.na(y)] <- "NA"
x1 <- x [ (x!="NA") & (y!="NA") ]
y1 <- y [ (x!="NA") & (y!="NA") ]
plot_sankey( data.frame(x1,y1),color_by = 2,
             plot_labels = T, plot_weights = F,order_1_by = "total",
             order_2_by = "total",use_w1 = T,use_w2 = T,gap2v = 0,smoothing = .1,
             pal=hue_pal()(10))
text(c(0,1),c(1,1),labels = c( "BV_Monbilicus_v2" , "Antibiotics" ),pos=3,xpd=T)



x <- as.character(metadata[,"HIVstatus"])
x[is.na(x)] <- "NA"
y <- as.character(metadata[,"Antibiotics"])
y[is.na(y)] <- "NA"
x1 <- x [ (x!="NA") & (y!="NA") ]
y1 <- y [ (x!="NA") & (y!="NA") ]
plot_sankey( data.frame(x1,y1),color_by = 2,
             plot_labels = T, plot_weights = F,order_1_by = "total",
             order_2_by = "total",use_w1 = T,use_w2 = T,gap2v = 0,smoothing = .1,
             pal=hue_pal()(10))
text(c(0,1),c(1,1),labels = c( "HIVstatus" , "Antibiotics_v2" ),pos=3,xpd=T)


x <- as.character(metadata[,"blood_cervix_v3"])
x[is.na(x)] <- "NA"
y <- as.character(metadata[,"cervical_inflammation_v3"])
y[is.na(y)] <- "NA"
x1 <- x [ (x!="NA") & (y!="NA") ]
y1 <- y [ (x!="NA") & (y!="NA") ]
plot_sankey( data.frame(x1,y1),
             plot_labels = T, plot_weights = F,order_1_by = "total",
             order_2_by = "total",use_w1 = T,use_w2 = T,gap2v = 0,smoothing = .1,
             pal=hue_pal()(9))
text(c(0,1),c(1,1),labels = c( "blood_cervix_v3" , "cervical_inflammation_v3" ),pos=3,xpd=T)

```

**Figure 7.** A few examples of significant association between metadata parameters shown as sankey plots.

\clearpage


```{r, echo = FALSE,fig.width=10,fig.height=10}
mypar(min(length(metadata_parameters),6),4,mar=c(1,4,1,4))

# metadata$Contraception
#Antibiotics usage
for(i in metadata_parameters){
  empty_plot()
  text( par("usr")[2] , mean(par("usr")[3:4]),labels = sub("_","\n",i),pos=2 )
  for(j in c("joint_clustering")){
    x <- as.character(metadata[,i])
    x[is.na(x)] <- "NA"
    y <- as.character(metadata[,j])
    y[is.na(y)] <- "NA"
    x1 <- x [ (x!="NA") & (y!="NA") ]
    y1 <- y [ (x!="NA") & (y!="NA") ]
    plot_sankey( data.frame(x1,y1),
             plot_labels = T, plot_weights = F,order_1_by = "total",
             order_2_by = "total",use_w1 = T,use_w2 = T,gap2v = 0,smoothing = .1,
             pal=hue_pal()(13))
    text(c(0,1),c(1,1),labels = c(i,j),pos=3,xpd=T)
  }
}
```

**Figure 8.** Sankey plots for all tested associations between the patient groups identified in in the microbiome datasets.

\clearpage

```{r, eval=F, echo = FALSE,fig.width=10,fig.height=13}
myletter <- letters

for(i in c("ASV_tissue_V3_normalized_batch_corrected",
           "ASV_CVL_V3_normalized_batch_corrected",
           "ASV_CVL_V2_normalized_NOT_batch_corrected")){
  
  cat(paste0("Correlation between RNAseq and ",sub("_norm.*","",i)))
  
  TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
  bac_dataset <- datasets_all_samples[[i]]
  dge <- read.csv2(paste0(PATH,"results/DGE_ASV_CVL_V2_normalized_NOT_batch_corrected.csv"),row.names = 1)
  samples_TRX_and_CVL3 <- colnames(counts) [ colSums(counts[c("Tissue_RNAseq_V3_normalized",i),] > 0)==2 ]
  
  # mypar(4,3,mar=c(4,2,2,2))
  #filter TRX dataset
  TRX <- TRX [ , samples_TRX_and_CVL3 ]
  TRX <- TRX [ rowSums(TRX>0)>= 2 , ]
  top_vars_TRX <- names(sort(apply(TRX,1,var),decreasing = T)[1:5000] )
  # plot(apply(TRX,1,mean),apply(TRX,1,var),col=ifelse(rownames(TRX) %in% top_vars_TRX,"red","grey20"),cex=.2)
  TRX <- TRX[ top_vars_TRX , ]
  dim(TRX)
  
  #filter bac_dataset dataset
  bac_dataset <- bac_dataset [ , samples_TRX_and_CVL3 ]
  bac_dataset <- bac_dataset [ rowSums(bac_dataset>0)>= 2 , ]
  top_vars_bacs <- names(sort(apply(bac_dataset,1,var),decreasing = T)[1:100] )
  dim(bac_dataset)
  
  cors <- cor( t(rbind(TRX) ) ,  t(rbind(bac_dataset) ) )
  # hist(cors,xlim=c(-1,1),breaks = 100)
  dim(cors)
  
  
  bacteria_use <- colnames(cors)
  gmt <- gmtPathways("../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
  cl <- parallel::makePSOCKcluster(parallel::detectCores()-1)
  invisible(parallel::clusterEvalQ(cl, {c("gmt","cors");library(fgsea);library(stats);library(base)}))
  enrichments <- parallel::parLapply(cl,bacteria_use,gmt=gmt,cors=cors,function(x,gmt,cors){
    res <- fgsea(pathways = gmt,stats = sort(cors[,x],decreasing = T))
    return(res)
  })
  parallel::stopCluster(cl)
  names(enrichments) <- bacteria_use
  
  
  
  pvalues <- lapply(enrichments,function(x) setNames(x$pval,x$pathway) )
  pvalues <- t(as.data.frame(pvalues))
  pvalues <- -log10( pvalues )
  top_pathways <- names(sort(apply(pvalues,1,median),T))[1:50]
  # mypar(mar=c(4,15,2,2))
  # boxplot( t(pvalues[rev(top_pathways),]),cex=.1,horizontal=T,las=1 )
  
  # pvalues[ pvalues < -log10(0.05)] <- 0
  # sort(rowSums(pvalues>0),T)
  # sort(colSums(pvalues>0),T)
  NESes <- lapply(enrichments,function(x) setNames(x$NES,x$pathway) )
  NESes <- t(as.data.frame(NESes))
  NESes[is.na(NESes)] <- 0
  NESes[ pvalues < -log10(0.05) ] <- 0
  NESes <- NESes[rowSums(NESes!=0) >= 10 , colSums(NESes!=0) >= 10]
  
  # o_kegg <- hclust( as.dist( (1-cor(NESes))/2 ),"ward.D2")$order
  # o_bacs <- hclust( as.dist( (1-cor(t(NESes)))/2 ),"ward.D2")$order
  o_kegg <- hclust( as.dist( (1-cor(NESes))/2 ),"ward.D2")$order
  o_bacs <- hclust( as.dist( (1-cor(t(NESes) ))/2 ),"ward.D2")$order
  
  NESes <- NESes[o_bacs,o_kegg]
  
  mypar(1,1,mar=c(12,2,2,12))
  image( t(NESes[nrow(NESes):1,]),col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-3,3,length.out = 92),
         axes=F,border=NA,main="tissue RNAseq KEGG pathways",xlab="",ylab=i,line=.4,cex.main=1,font.main=1)
  text(  par("usr")[c(4)] , seq(1,0,length.out = nrow(NESes)),
        labels = rownames(NESes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=.4)
  text( seq(0,1,length.out = ncol(NESes)) , par("usr")[c(1)],
        labels = colnames(NESes), srt = 90, adj = c(1,.5), xpd = TRUE, cex=.4)
  add_letter(myletter); myletter <- myletter[-1]
  # empty_plot()
  cat("\\clearpage")

}


# **Figure 9.** Heatmaps showing the functional association of microbiome datasets with the expression profiles in the RNAseq dataset. Briefly, bacterial abundances from each dataset were correlated with the gene expression of the top 5000 highly variable genes from the RNAseq dataset, generating a correlation matrix between bacteria and genes. Then for each bacteria, we rank genes based on their correlation to that bacteria and perform gene set enrichment anlaysis (GSEA) using the KEGG gene annotation database. This, in turn, will result in a matrix associating every bacteria with every KEGG process in the tissue. The heatmap shows the normalized enrichment score (NES). Only enrichments with pvalue below 0.05 are shown. Bacteria and pathways significant in less that 10 pathways and bacteria, respectivelly, were omitted.

```


\clearpage


```{r, echo = FALSE,fig.width=10,fig.height=10}

all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]],
                        datasets_all_samples[["ASV_CVL_V2_normalized_NOT_batch_corrected"]])

cors <- cor(t(all_microbiome[rowSums(all_microbiome > 0)>=10,]))

#Graph construction 
k_nearest_neighbours <- 5 # k for KNN
snn_threshold <- 0.1 # prunning threshold of SNN

#Compute kNN graph
knn <- RANN::nn2( (1-cors)/2 ,k = k_nearest_neighbours)
nn <- matrix(0, nrow(knn$nn.idx), nrow(knn$nn.idx))
for(idx in 1:nrow(knn$nn.idx) ){  nn[idx,knn$nn.idx[idx,]] <- 1 }

#Calculate SNN and prune based on Jaccard index
N <- matrix(1,ncol = ncol(nn),nrow = nrow(nn))
P <- t(nn) %*% nn
B <- N - nn
C <- t(B) %*% B
Q <- nrow(nn)*N - C
J <- P/Q
J[is.nan(J)] <- 0
J[J < snn_threshold] <- 0
rownames(J) <- rownames(cors)
colnames(J) <- colnames(cors)

#Build a graph object, perform clustering and define the force-directed layout
g <- graph_from_adjacency_matrix(J,mode = "undirected",diag = F,weighted = T)
write.csv(J,"../../results/bacteria_SNN_graph.csv",row.names = T)

cl <- igraph::cluster_louvain(g)
# cl <- igraph::walktrap.community(g)
set.seed(1)
l <- layout_with_fr(g,niter=3000,start.temp=1)

#Define layout with UMAP
mU <- uwot::umap((1-cors)/2, n_neighbors = 10)

#Plot
mypar(3,3,mar=c(0,0,2,0))

plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(cl$membership)] , 
    edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=l,
    vertex.size=10, main="graph layout FR",
    edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(cl$membership)] , 
    edge.width=  ( E(g)$weight / max(E(g)$weight)) ,layout=mU,
    vertex.size=10, main="UMAP",
    edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
par(mar=c(0,0,0,0))
empty_plot()
legend("topleft",title = "bact. communities",
     legend = levels(factor(cl$membership)),
     bty = "n",pch = 21,pt.bg = pal,cex = .8,pt.cex = .9,xpd=T)
```


**Figure 9a.** Bacterial community embedding of `r k`-SNN graph clustered using walktrap community detection algorithm based on the normalized bacterial abundances across all microbiome datasets (including samples not present in all datasets).



```{r, echo = FALSE,fig.width  = 10, fig.height=12}
mypar(3,6,mar=c(1,1,2,1))
for(j in sort(unique(cl$membership)) ){
  empty_plot()
  text( 0,1,labels = paste0(rownames(cors)[ cl$membership == j ],collapse = "\n"),adj=c(0,1),cex=.7)
  title(main=paste0("community",j))
}
write.csv2(cl$membership,"../../results/bacterial_communities.csv",row.names = rownames(cors))
```


**Figure 9b.** List of bacterial belongin to each community.


```{r, echo = FALSE,fig.width  = 10, fig.height=5}
# for Genus_lacto taxonomy:
taxa <- read_csv("../../results/SeqID_to_gen-lacto_tax-oth.csv") %>%
  select(-Sequence, -Seq_ID, -taxa_other, -Species) %>%
  unique() %>%
  filter(!(is.na(.$Genus_lacto))) %>%
  filter(!(.$Genus == "Granulicatella" & .$Family == "Enterococcaceae")) %>%
  column_to_rownames(., var = "Genus_lacto")

# For other_taxa taxonomy:
# taxa <- read.csv2("../../results/taxonomy.csv",row.names = 1,stringsAsFactors = F,as.is = T)
# taxa[is.na(taxa)] <- "NA"
# rownames(cors) %in% rownames(taxa)

mypar(2,2,mar=c(1,10,2,10))
x <- taxa[match(rownames(cors),rownames(taxa)),2]
y <- taxa[match(rownames(cors),rownames(taxa)),3]
plot_sankey( data.frame( x[y!="NA"] , y[y!="NA"]), color_by = 1,
       plot_labels = T, plot_weights = F,order_1_by = "total",
       order_2_by = "total",use_w1 = T,use_w2 = T,gapv = .02,gap2v = 0,smoothing = .1,
       pal=hue_pal()(length(unique(x))))

plot_sankey( data.frame( y[y!="NA"] , cl$membership[y!="NA"]), color_by = 1,
     plot_labels = T, plot_weights = F,order_1_by = "total",
     order_2_by = "total",use_w1 = T,use_w2 = T,gapv = .02,gap2v = 0,smoothing = .1,
     pal=hue_pal()(length(unique(y))))

plot_sankey( data.frame( x[y!="NA"] , cl$membership[y!="NA"]), color_by = 1,
     plot_labels = T, plot_weights = F,order_1_by = "total",
     order_2_by = "total",use_w1 = T,use_w2 = T,gapv = .02,gap2v = 0,smoothing = .1,
     pal=hue_pal()(length(unique(x))))

```


**Figure 9b.** Taxonomic annotation of bacterial communities.


```{r, eval=F, echo = FALSE,fig.width=10, fig.height=10}


module_means_per_dataset <- lapply(datasets_all_samples[-1],function(x){
  return( rowsum( x[rownames(cors),] , group = cl$membership ) / c(table(cl$membership)) )
})

mypar(4,3,mar=c(3,5,2,2))


# mypar(3,1,mar=c(1,2,1,3))
# sample_order <- order(df$louvain_tissue_V3)
# image(t(module_means_per_dataset[["ASV_tissue_V3_normalized_batch_corrected"]]
#         [ ,sample_order]),
#       col=heat_pal,axes=F,border=NA,
#       main="tissue week 3",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
# points( (0:(length(df$louvain_tissue_V3)-1) )/(length(df$louvain_tissue_V3)-1),
#         rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_tissue_V3)), pch=16,
#         col=pal[factor(df$louvain_tissue_V3)][sample_order] ,xpd=T)
# points( (0:(length(df$louvain_tissue_V3)-1) )/(length(df$louvain_tissue_V3)-1),
#         rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_tissue_V3)), pch=16,
#         col=pal[factor(df$louvain_tissue_V3)][sample_order] ,xpd=T)
# text(  par("usr")[c(2)] , seq(1,0,length.out = nrow(NESes)),
#         labels = rownames(NESes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
# 
# 
# sample_order <- order(df$louvain_CVL_V3)
# image(t(module_means_per_dataset[["ASV_CVL_V3_normalized_batch_corrected"]]
#         [ ,sample_order]),
#       col=heat_pal,axes=F,border=NA,
#       main="CVL week 3",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
# points( (0:(length(df$louvain_CVL_V3)-1) )/(length(df$louvain_CVL_V3)-1),
#         rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_CVL_V3)), pch=16,
#         col=pal[factor(df$louvain_CVL_V3)][sample_order] ,xpd=T)
# points( (0:(length(df$louvain_CVL_V3)-1) )/(length(df$louvain_CVL_V3)-1),
#         rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_CVL_V3)), pch=16,
#         col=pal[factor(df$louvain_CVL_V3)][sample_order] ,xpd=T)
# text(  par("usr")[c(2)] , seq(1,0,length.out = nrow(NESes)),
#         labels = rownames(NESes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
# 
# 
# sample_order <- order(df$louvain_CVL_V2)
# image(t(module_means_per_dataset[["ASV_CVL_V2_normalized_NOT_batch_corrected"]]
#         [ ,sample_order]),
#       col=heat_pal,axes=F,border=NA,
#       main="CVL week 2",xlab=" ",ylab="",line=.4,cex.main=1,font.main=1)
# points( (0:(length(df$louvain_CVL_V2)-1) )/(length(df$louvain_CVL_V2)-1),
#         rep(par("usr")[c(3)]-.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_CVL_V2)), pch=16,
#         col=pal[factor(df$louvain_CVL_V2)][sample_order] ,xpd=T)
# points( (0:(length(df$louvain_CVL_V2)-1) )/(length(df$louvain_CVL_V2)-1),
#         rep(par("usr")[c(4)]+.01*diff(par("usr")[c(3,4)]),
#             length(df$louvain_CVL_V2)), pch=16,
#         col=pal[factor(df$louvain_CVL_V2)][sample_order] ,xpd=T)
# text(  par("usr")[c(2)] , seq(1,0,length.out = nrow(NESes)),
#         labels = rownames(NESes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)


mypar(3,3,mar=c(3,5,2,2))


  
for( i in c("louvain_tissue_V3","louvain_CVL_V3","louvain_CVL_V2")){
  for( j in c("ASV_tissue_V3_normalized_batch_corrected",
          "ASV_CVL_V3_normalized_batch_corrected",
          "ASV_CVL_V2_normalized_NOT_batch_corrected")){
    temp_group <- as.character(df[,i])
    temp_group[ is.na(temp_group) ] <- "NA"
    temp_group[ temp_group == "" ] <- "NA"
    temp_group <- factor(temp_group)
    
    barlist(data=module_means_per_dataset[[j]],
            genes=rownames(module_means_per_dataset[[j]]),col = pal,
            labels = paste0("bac.com.",rownames(module_means_per_dataset[[j]])),
            clustering=temp_group, main=paste0(i,"\n", sub("_norm.*","",j)),srt=45)
  }
}

```



```{r, eval=T, echo = FALSE,fig.width=10, fig.height=14}

module_means_per_dataset <- lapply(datasets_all_samples[-1],function(x){
  return( rowsum( x[rownames(cors),] , group = cl$membership ) / c(table(cl$membership)) )
})

mypar(3,3,mar=c(3,5,2,2))

for( i in c("Luminal_gr","group_M9_CVLv3","Louvain_clusters")){
  for( j in c("ASV_tissue_V3_normalized_batch_corrected",
          "ASV_CVL_V3_normalized_batch_corrected",
          "ASV_CVL_V2_normalized_NOT_batch_corrected")){
      
    temp_group <- as.character(metadata[,i])
    temp_group[ is.na(temp_group) ] <- "NA"
    temp_group[ temp_group == "" ] <- "NA"
    temp_group <- factor(temp_group)
    
    barlist(data=module_means_per_dataset[[j]],
            genes=rownames(module_means_per_dataset[[j]]),col=pal[factor(temp_group)],
            labels = paste0("bac.com.",rownames(module_means_per_dataset[[j]])),
            clustering=temp_group, main=paste0(i,"\n", sub("_norm.*","",j)),srt=45)
  }
}
```

**Figure 5.** Comparisson of microbiome communities across all datasets and all clustering methods.



```{r, echo = FALSE,fig.width=10,fig.height=8}
myletter <- letters

NESse_list <- list()
pvalues_list <- list()

for(i in c("ASV_tissue_V3_normalized_batch_corrected",
           "ASV_CVL_V3_normalized_batch_corrected")){
  
  cat(paste0("\nCorrelation between RNAseq and ",sub("_norm.*","",i)))
  
  TRX <- as.matrix(datasets_all_samples[["Tissue_RNAseq_V3_normalized"]])
  bac_dataset <- module_means_per_dataset[[i]]
  samples_TRX_and_CVL3 <- colnames(bac_dataset)
  
  # mypar(4,3,mar=c(4,2,2,2))
  #filter TRX dataset
  TRX <- TRX [ , samples_TRX_and_CVL3 ]
  TRX <- TRX [ rowSums(TRX>0)>= 5 , ]
  
  #filter bac_dataset dataset
  bac_dataset <- bac_dataset [ , samples_TRX_and_CVL3 ]
  dim(bac_dataset)
  
  cors <- cor( t(rbind(TRX) ) ,  t(rbind(bac_dataset) ) )
  dim(cors)
  
  bacteria_use <- colnames(cors)
  gmt <- gmtPathways("../../supplementary_files/h.all.v6.2.symbols.gmt.txt")
  cl <- parallel::makePSOCKcluster(parallel::detectCores()-1)
  invisible(parallel::clusterEvalQ(cl, {c("gmt","cors");library(fgsea);library(stats);library(base)}))
  enrichments <- parallel::parLapply(cl,bacteria_use,gmt=gmt,cors=cors,function(x,gmt,cors){
    res <- fgseaMultilevel(pathways = gmt,stats = cors[,x])
    return(res)
  })
  parallel::stopCluster(cl)
  names(enrichments) <- paste0( "bac.com.", bacteria_use )
  
  allpathways <- unique(unlist(lapply(enrichments,function(x){x[,1]})))
  allpvals <- setNames(rep(NA,length(allpathways)),allpathways)
  
  pvalues <- lapply(enrichments,function(x){
    temp <- allpvals
    temp[x$pathway] <- x$pval
    return(temp)
  })
  pvalues <- t(as.data.frame(pvalues))
  pvalues <- -log10( pvalues )
  pvalues[is.na(pvalues)] <- 0
  
  allNES <- setNames(rep(NA,length(allpathways)),allpathways)
  NESes <- lapply(enrichments,function(x) {
    temp <- allNES
    temp[x$pathway] <- x$NES
    return(temp)
  })
  NESes <- t(as.data.frame(NESes))
  NESes[is.na(NESes)] <- 0
  # NESes[ pvalues < -log10(0.1) ] <- 0

  NESse_list[[i]] <- NESes
  pvalues_list[[i]] <- pvalues
  
}

write.csv2(NESse_list,"../../results/Hallmark_NESse_list.csv")
write.csv2(pvalues_list,"../../results/Hallmark_pvalues_list.csv")

common_pathways <- unique( unlist( lapply( NESse_list , function(x) colnames(x)[colSums(x!=0)>=1] ) ) )

layout(matrix(c(1,2,3,4,4),nrow = 1))
par(mar=c(6,1,2,0))
for( i in names(NESse_list)){
  x <- NESse_list[[i]][,common_pathways]
  image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
         axes=F,border=NA,main=sub("_norm.*","",i),xlab="",ylab="",line=.4,cex.main=1,font.main=1)
  # add_letter(myletter); myletter <- myletter[-1]
  text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
      labels = rownames(x), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
}
empty_plot(xaxs="i",yaxs="i",
           xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)])
text(  par("usr")[c(1)],seq(1,0,length.out = ncol(x)) ,
      labels = sub("HALLMARK_","",colnames(x)), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
  
cat("\\clearpage")

#Combined representation of the two datasets
layout(matrix(c(1,2,2,2,2),nrow = 1))
par(mar=c(6,1,3,0))
x <- (NESse_list[[1]][,common_pathways] + NESse_list[[2]][,common_pathways]) / 2
x[NESse_list[[1]][,common_pathways] == 0] <- 0
x[NESse_list[[2]][,common_pathways] == 0] <- 0
x <- x[, colSums(x!=0)>0]
image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
       axes=F,border=NA,main="CVL_v3 + tissue_v3\ncombined",xlab="",ylab="",line=.4,cex.main=1,font.main=1)
# add_letter(myletter); myletter <- myletter[-1]
text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
    labels = rev(rownames(x)), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
empty_plot(xaxs="i",yaxs="i",
           xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)],main="HALLMARK")
text(  par("usr")[c(1)],seq(0,1,length.out = ncol(x)) ,
      labels = rev(sub("HALLMARK_","",colnames(x))), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)




#Combined representation of the two datasets
layout(matrix(c(1,1,1,1,1,2,2),nrow = 1))
par(mar=c(6,30,3,0))
x <- (NESse_list[[1]][,common_pathways] + NESse_list[[2]][,common_pathways]) / 2
x[NESse_list[[1]][,common_pathways] == 0] <- 0
x[NESse_list[[2]][,common_pathways] == 0] <- 0
x <- x[, colSums(x!=0)>0]

TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
non_zeros <- colSums(TRX)!=0
TRX <- TRX [  , non_zeros ]
pathway_means <- t(sapply( colnames(x) , function(i){
  i <- gmt[[i]]
  i <- as.matrix( TRX[ rownames(TRX) %in% i , ] )
  if(ncol(i)>1){
    i <- ( i - apply(i,1,min) ) / ( apply(i,1,max)- apply(i,1,min) )
    i <- colMeans(i)
  } else {
    i <- (i - min(i)) / (max(i) - min(i))
  }
  return( i - min(i) )
} ))
plot_dots(data=pathway_means,main="patient groups",cex.main=1,font.main=1,
        genes=rownames(pathway_means),
        clustering=factor(metadata$joint_clustering[non_zeros]),
        srt=0,pal=c("blue","blue","navy","navy","grey95","grey95","firebrick3","black"))


par(mar=c(6,2,3,2))
image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
       axes=F,border=NA,main="CVL_v3 + tissue_v3\ncombined",xlab="",ylab="",line=.4,cex.main=1,font.main=1)
# add_letter(myletter); myletter <- myletter[-1]
text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
    labels = rownames(x), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
# empty_plot(xaxs="i",yaxs="i",
#            xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)],main=" ")
# text(  par("usr")[c(1)],seq(1,0,length.out = ncol(x)) ,
      # labels = sub("KEGG_","",colnames(x)), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)


```

**Figure 9.** Heatmaps showing the functional association of microbiome datasets with the expression profiles in the RNAseq dataset. Briefly, bacterial abundances from each dataset were correlated with the gene expression of the top 5000 highly variable genes from the RNAseq dataset, generating a correlation matrix between bacteria and genes. Then for each bacteria, we rank genes based on their correlation to that bacteria and perform gene set enrichment anlaysis (GSEA) using the Hallmark MSigDB gene annotation database. This, in turn, will result in a matrix associating every bacteria with every Hallmark MSigDB process in the tissue. The heatmap shows the normalized enrichment score (NES). Only enrichments with pvalue below 0.05 are shown.

\clearpage

```{r, echo = FALSE,fig.width=10,fig.height=8}
myletter <- letters

NESse_list <- list()
pvalues_list <- list()

for(i in c("ASV_tissue_V3_normalized_batch_corrected",
           "ASV_CVL_V3_normalized_batch_corrected")){
  
  cat(paste0("\nCorrelation between RNAseq and ",sub("_norm.*","",i)))
  
  TRX <- as.matrix(datasets_all_samples[["Tissue_RNAseq_V3_normalized"]])
  bac_dataset <- module_means_per_dataset[[i]]
  samples_TRX_and_CVL3 <- colnames(bac_dataset)
  
  # mypar(4,3,mar=c(4,2,2,2))
  #filter TRX dataset
  TRX <- TRX [ , samples_TRX_and_CVL3 ]
  TRX <- TRX [ rowSums(TRX>0)>= 5 , ]
  
  #filter bac_dataset dataset
  bac_dataset <- bac_dataset [ , samples_TRX_and_CVL3 ]
  dim(bac_dataset)
  
  cors <- cor( t(rbind(TRX) ) ,  t(rbind(bac_dataset) ) )
  dim(cors)
  
  bacteria_use <- colnames(cors)
  gmt <- gmtPathways("../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
  cl <- parallel::makePSOCKcluster(parallel::detectCores()-1)
  invisible(parallel::clusterEvalQ(cl, {c("gmt","cors");library(fgsea);library(stats);library(base)}))
  enrichments <- parallel::parLapply(cl,bacteria_use,gmt=gmt,cors=cors,function(x,gmt,cors){
    res <- fgseaMultilevel(pathways = gmt,stats = cors[,x])
    return(res)
  })
  parallel::stopCluster(cl)
  names(enrichments) <- paste0( "bac.com.", bacteria_use )
  
  allpathways <- unique(unlist(lapply(enrichments,function(x){x[,1]})))
  allpvals <- setNames(rep(NA,length(allpathways)),allpathways)
  
  pvalues <- lapply(enrichments,function(x){
    temp <- allpvals
    temp[x$pathway] <- x$pval
    return(temp)
  })
  pvalues <- t(as.data.frame(pvalues))
  pvalues <- -log10( pvalues )
  pvalues[is.na(pvalues)] <- 0
  
  allNES <- setNames(rep(NA,length(allpathways)),allpathways)
  NESes <- lapply(enrichments,function(x) {
    temp <- allNES
    temp[x$pathway] <- x$NES
    return(temp)
  })
  NESes <- t(as.data.frame(NESes))
  NESes[is.na(NESes)] <- 0
  # NESes[ pvalues < -log10(0.1) ] <- 0

  NESse_list[[i]] <- NESes
  pvalues_list[[i]] <- pvalues
}

write.csv2(NESse_list,"../../results/Kegg_NESse_list.csv")
write.csv2(pvalues_list,"../../results/Kegg_pvalues_list.csv")

common_pathways <- unique( unlist( lapply( NESse_list , function(x) colnames(x)[colSums(x!=0)>=1] ) ) )

layout(matrix(c(1,2,3,4,4),nrow = 1))
par(mar=c(6,1,2,0))
for( i in names(NESse_list)){
  x <- NESse_list[[i]][,common_pathways]
  image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
         axes=F,border=NA,main=sub("_norm.*","",i),xlab="",ylab="",line=.4,cex.main=1,font.main=1)
  # add_letter(myletter); myletter <- myletter[-1]
  text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
      labels = rownames(x), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
}
empty_plot(xaxs="i",yaxs="i",
           xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)])
text(  par("usr")[c(1)],seq(1,0,length.out = ncol(x)) ,
      labels = sub("KEGG_","",colnames(x)), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
  
cat("\\clearpage")

#Combined representation of the two datasets
layout(matrix(c(1,2,2,2,2),nrow = 1))
par(mar=c(6,1,3,0))
x <- (NESse_list[[1]][,common_pathways] + NESse_list[[2]][,common_pathways]) / 2
x[NESse_list[[1]][,common_pathways] == 0] <- 0
x[NESse_list[[2]][,common_pathways] == 0] <- 0
x <- x[, colSums(x!=0)>0]
image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
       axes=F,border=NA,main="CVL_v3 + tissue_v3\ncombined",xlab="",ylab="",line=.4,cex.main=1,font.main=1)
# add_letter(myletter); myletter <- myletter[-1]
text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
    labels = rev(rownames(x)), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
empty_plot(xaxs="i",yaxs="i",
           xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)],main="KEGG")
text(  par("usr")[c(1)],seq(0,1,length.out = ncol(x)) ,
      labels = rev(sub("KEGG_","",colnames(x))), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)




#Combined representation of the two datasets
layout(matrix(c(1,1,1,1,1,2,2),nrow = 1))
par(mar=c(6,30,3,0))
x <- (NESse_list[[1]][,common_pathways] + NESse_list[[2]][,common_pathways]) / 2
x[NESse_list[[1]][,common_pathways] == 0] <- 0
x[NESse_list[[2]][,common_pathways] == 0] <- 0
x <- x[, colSums(x!=0)>0]

TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
non_zeros <- colSums(TRX)!=0
TRX <- TRX [ , non_zeros ]
pathway_means <- t(sapply( colnames(x) , function(i){
  i <- gmt[[i]]
  i <- as.matrix( TRX[ rownames(TRX) %in% i , ] )
  if(ncol(i)>1){
    i <- ( i - apply(i,1,min) ) / ( apply(i,1,max)- apply(i,1,min) )
    i <- colMeans(i)
  } else {
    i <- (i - min(i)) / (max(i) - min(i))
  }
  return( i - min(i) )
} ))
plot_dots(data=pathway_means,main="patient groups",cex.main=1,font.main=1,
        genes=rownames(pathway_means),
        clustering=factor(metadata$joint_clustering[non_zeros]),
        srt=0,pal=c("blue","blue","navy","navy","grey95","grey95","firebrick3","black"))


par(mar=c(6,2,3,2))
image( x[,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),breaks=seq(-5,5,length.out = 92),
       axes=F,border=NA,main="CVL_v3 + tissue_v3\ncombined",xlab="",ylab="",line=.4,cex.main=1,font.main=1)
# add_letter(myletter); myletter <- myletter[-1]
text(   seq(0,1,length.out = nrow(x)) , par("usr")[c(3)] ,
    labels = rownames(x), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)
# empty_plot(xaxs="i",yaxs="i",
#            xlim=par("usr")[c(1,2)],ylim=par("usr")[c(3,4)],main=" ")
# text(  par("usr")[c(1)],seq(1,0,length.out = ncol(x)) ,
      # labels = sub("KEGG_","",colnames(x)), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
```


**Figure 9.** Heatmaps showing the functional association of microbiome datasets with the expression profiles in the RNAseq dataset. Briefly, bacterial abundances from each dataset were correlated with the gene expression of the top 5000 highly variable genes from the RNAseq dataset, generating a correlation matrix between bacteria and genes. Then for each bacteria, we rank genes based on their correlation to that bacteria and perform gene set enrichment anlaysis (GSEA) using the KEGG gene annotation database. This, in turn, will result in a matrix associating every bacteria with every KEGG process in the tissue. The heatmap shows the normalized enrichment score (NES). Only enrichments with pvalue below 0.05 are shown. Bacteria and pathways significant in less that 10 pathways and bacteria, respectivelly, were omitted.


```{r, echo = FALSE,fig.width=10,fig.height=8}
metadata <- select(metadata, -joint_clustering, -group_M9_CVLv3)
write.csv2(metadata,"../../results/metadata_integration.csv",row.names = T)
```

