---
title: "DGE's and Enrichment (GO and KEGG)"
output: 
  pdf_document:
    fig_caption: yes
    keep_tex: yes
header-includes: 
- \usepackage{float}
editor_options: 
  chunk_output_type: console
---



```{r message=FALSE, warning=FALSE, include=FALSE}
##################
# LOAD LIBRARIES #
##################
#setwd("/Users/vilkal/work/Brolidens_work/Projects/Gabriella_repo/reports/rmarkdown/manuscript")
suppressWarnings({suppressMessages({suppressPackageStartupMessages({
  library(tidyverse)
  library(edgeR)
  library(openxlsx)
  library(scales)
  library(RColorBrewer)
  remotes::install_github("czarnewski/niceRplots",force=T)
  library(niceRplots)
})  })  })


#############
# LODA DATA #
#############
datasets_all_samples <- readRDS("../../../results/datasets_all_samples.RDS")
metadata <- read.csv2("../../../results/metadata_integration.csv",row.names = 1, stringsAsFactors = F)
sample_use <- metadata$ID

#color palettes
#################
# COLOR PALETTS #
#################
Set1 <- c("#4DAF4A", "#377EB8", "#E41A1C", "#FF7F00", "#984EA3", "#FFFF33", "#A65628", "#F781BF", "#999999")
pal <- c(Set1,RColorBrewer::brewer.pal(8,"Set2"),
         RColorBrewer::brewer.pal(9,"Pastel1"),RColorBrewer::brewer.pal(8,"Pastel2")) #color pallete for plots
pal <- c( "#0072B2", "#009E73","#D55E00", "#CC79A7", "#E69F00", "#999999")
taxa_pal <- c(RColorBrewer::brewer.pal(8,"Pastel2"),RColorBrewer::brewer.pal(8,"Pastel1"),"grey90")
```

## Heatmap DGEs and enrichment
```{r Heatmap functions, echo=FALSE, warning=FALSE, message=FALSE}
##################
# DEG's FUNCTION #
##################
#group <- "Tissue_gr"
#group <- "Luminal_gr"
edgR_dge.fun <- function(group, confound) {
  
  groups <- metadata[sample_use, group]
  n_gr <- length(na.omit(unique(groups)))
  
  if(any(is.na(groups))) {
    groups <- replace_na(groups, "X")
  }

  if(confound==FALSE) {
    c <- "_no_conf"
    design <- model.matrix(~groups, data=metadata[sample_use,]) # without confounders
  }else{
    c <- "_conf"
    design <- model.matrix(~groups+HIVstatus+Contraception,
                         data=metadata[sample_use,])
  }
  
  colnames(design) <- sub("groups","",colnames(design))
  colnames(design) <- sub("\\(Intercept\\)","Intercept",colnames(design))
  colnames(design) <- sub(" ","_",colnames(design))
  
  y <- DGEList(counts=TRX_counts[,sample_use], remove.zeros = T)
  y <- calcNormFactors(y,method = "TMM")
  y <- estimateGLMCommonDisp(y,design)
  y <- estimateGLMTagwiseDisp(y,design)
  fit <- glmFit(y, design)
  
  lrt <- glmLRT(fit,coef=2:n_gr) # with intercept
  top <- topTags(lrt,adjust.method = "BH",n = "all",sort.by = "p.value")[[1]]
  colnames(top) <- sub(group,"",colnames(top))
  top <- cbind(LogFC.intercept=0,top) # with intercept
  top <-  rownames_to_column(top, var = "Genes")
  
  write.xlsx(top, file=paste0("../../../results/DE/","dge_heatmap_",group,c,".xlsx"))
  
return(list(design=design, y=y, fit=fit, lrt=lrt, top=top, c=c))
}

#######################
# ENRICHMENT FUNCTION #
#######################
# group_name <- "Luminal_gr"
# top <- dge_L$top
# y <- dge_L$y
# c <- dge_L$c
gene_mod_enrich.fun <- function(top, y, c, db, group_name, cuttoff){
  p_val <- pull(top, "PValue")
  top_dge <- (p_val < 0.01) #& (rowSums( abs(top[,grep("logFC",colnames(top))]) >= log2(1.5) ) >= 1)
  top_dge <- top$Genes[ top_dge ]
  
  top_TRX_counts <- edgeR::cpm(y,normalized.lib.sizes = T,log = T)[top_dge, ]
  top_TRX_counts <- t(apply(top_TRX_counts,1,function(x){scale(x,T,T)}))
  top_TRX_counts[top_TRX_counts > 5] <- 5
  colnames(top_TRX_counts) <- colnames(TRX_counts)
  
  
  h <- hclust( as.dist( (1- cor(t(top_TRX_counts)))/2 ), method = "ward.D2")
  cuttoff <- cuttoff[names(cuttoff) == db]
  gene_module <- cutree(h, h=cuttoff)
  
  source("../enrichment_function.R")
  if(db == "GO"){
  gmt_list <- fgsea::gmtPathways("../../../supplementary_files/c5.bp.v6.2.symbols.gmt.txt")}else{
  gmt_list <- fgsea::gmtPathways("../../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")}
  res_list <- lapply( unique( gene_module[h$order] ),function(x){
    temp <- compute_enrichment(genes = names(gene_module)[gene_module==x],
                              gmt_list = gmt_list,
                              min_terms_pathway = 10,
                              max_terms_pathway = 300,
                              min_overlap = 3,
                              sort_by_pvalue = T)
    temp <- temp[!grepl("REGULATION",rownames(temp)),]
    return(temp)
  } )
  res <- map(res_list, ~rownames_to_column(.x, var = "Terms")) %>% 
    set_names(., paste0("gene_module_",  seq_along(res_list)))
  
  write.xlsx(res, file=paste0("../../../results/ENRICHMENT/Heatmap/",db,"_enrichment_heatmap_",group_name,c,".xlsx"))
  
  return(list(res_list=res_list, top_TRX_counts=top_TRX_counts, 
              gene_module=gene_module, cuttoff=cuttoff, h=h, c=c))
}

```

```{r Heatmap plots and tables, fig.height=4, fig.width=7, echo = FALSE, eval = TRUE, warning=F}

TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
TRX_counts <- round(2^TRX - 1)
TRX_counts <- TRX_counts [ rowSums(TRX_counts>0)>= 1 , ]
TRX_counts <- TRX_counts[,sample_use]

#########
# DEG's #
#########
# Replace missing values with average in sexwork_months column
m <- round(sum(metadata$sexwork_months, na.rm = T)/nrow(metadata))
metadata <- metadata %>%
  mutate(sexwork_months = ifelse(is.na(.$sexwork_months), m, .$sexwork_months))

dge_L <- edgR_dge.fun("Luminal_gr", confound=T)
dge_T <- edgR_dge.fun("Tissue_gr", confound=T)


###############
# ENRICHMENT #
##############
db <- c("GO", "KEGG") %>% set_names()
enrich_L <- db %>%
  map(., ~gene_mod_enrich.fun(dge_L$top,dge_L$y,dge_L$c,
                              db = .x,
                              group_name = "Luminal_gr", 
                              cuttoff = c(GO=2.5, KEGG=2.5)#3
                              )) 
enrich_T <- db %>%
  map(., ~gene_mod_enrich.fun(dge_T$top,dge_T$y,dge_T$c,
                              db = .x,
                              group_name = "Tissue_gr", 
                              cuttoff = c(GO=1.6, KEGG=1.6)
                              )) 


########################
# PLOT GO/KEGG HEATMAP #
########################

sample_use <- metadata$ID
gr = "Luminal_gr"
name = "Luminal"
enrich = enrich_L
db <- "GO"
db <- "KEGG"
# plot_heatmap.fun(enrich_T, db, "Tissue_gr", "Tissue")
plot_heatmap.fun <- function(enrich, db, gr, name, N = 3){
df <- na.omit(metadata[sample_use, c(gr, "ID")])
df <- arrange(df, df[1])
samples <- df[, "ID"]
names(samples) <- df[, gr]

conf <- pluck(enrich, db, "c")
h <- pluck(enrich, db, "h")
cuttoff <- pluck(enrich, db, "cuttoff")
res_list <- pluck(enrich, db, "res_list")
top_TRX_counts <- pluck(enrich, db, "top_TRX_counts")[,samples]
gene_module <- pluck(enrich, db, "gene_module")

layout(matrix(c(1,2,3,4,5,
                1,2,3,4, 5),
                nrow = 2,ncol = 5,byrow = T),widths = c(1,3.5,.36,8,8)) #1,2,.3,8
  ## Heatmap
  par(mar=c(4,.1,2,.1))
  plot( rev(as.dendrogram(h)) , xlim=c(max(h$height),-.2),horiz = T, leaflab = "none",xaxs="i",yaxs="i",axes=T)
  title(xlab="height", line=2)
  abline(v=cuttoff,xpd=F,col="red",lty=2)
  points( rep(-.1,length(gene_module)),
          seq(length(gene_module),1,length.out = length(gene_module)),
          col=taxa_pal[factor(gene_module[h$order])],
          pch=15,cex=.5,xpd=F)
  
  image( t(top_TRX_counts[h$order,pull(df, "ID")][nrow(top_TRX_counts):1,]),
         col = colorRampPalette(c("navy","navy","grey95","firebrick4","firebrick4") )(99),
         breaks = seq(-5,5,length.out = 100),axes=F)
  title(xlab="patient groups", line=2)
  
  end <- table(pal[factor(names(samples))[ order(df[1])]])
  end <- end[order(factor(names(end), levels = pal))]
  end <- map_dbl(cumsum(end), ~ (.x * (par("usr")[2])/length(samples)-.012) )
  start <- c(0, map_dbl(end[1:4], ~.x+0.012))
  
  axis(1, at = c(start, end), label = F, pos =1.02 , xpd=T, col="white",col.ticks="black")
  map2(start, end, ~lines(x=c(.y, .x), y= c(1.02, 1.02), xpd=T, cex=.8))
  # text labels
  t <- map2(start, end, ~(.x+.y)/2)
  map2(t,levels(factor(names(samples))), ~text(x=.x, y=1.03, .y,xpd=T, cex=.8))
  
  ## Meta info bars
  points( seq(0,1,length.out = length(samples)),rep(par("usr")[4],length(samples))+.005,
          col=pal[factor(names(samples))[ order(df[1])]],
          pch=15,xpd=T,cex=.7)
  points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.008,
          col=c("tomato","orange","#d7d7d7")[ factor(metadata[samples,"BV_Diagnosis_v3"][order(df[1])]) ],
          pch=15,xpd=T,cex=.45) 
  points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.022,
          col= c("#d7d7d7","tomato")[ factor(metadata[samples,"HIVstatus"][order(df[1])]) ],
          pch=15,xpd=T,cex=.45)
  
  ## Gene module enrichment
  names(res_list) <- unique( gene_module[h$order] )
  
  pvalues <- unlist(lapply(res_list,function(x){ -log10(x$pvalue) [1:N] }))
  # pvalues <- unlist(lapply(res_list,function(x){ x$jaccard_index [1:N] }))
  pvalues[is.na(pvalues)] <- 0
  terms <- unlist(lapply(res_list,function(x){ rownames(x) [1:N] }))
  terms[is.na(terms)] <- ""
  terms <- gsub("_", " ", terms)
  terms <- gsub("(GO |KEGG )|(\\w+)", "\\1\\L\\2", terms, perl = TRUE)
  genes <- unlist(lapply(res_list,function(x){ x$genes [1:N] }))
  genes[is.na(genes)] <- ""
  
  #pvalues <- setNames(pvalues,paste0(terms," (",genes,")"))
  pvalues <- setNames(pvalues,paste0(terms,""))
  module_color <- unlist(lapply(names(res_list),function(x){ rep(x,N) }))
  
  temp <- factor( rev(gene_module), levels = unique( gene_module[h$order] ))
  plot_sankey( data.frame(temp,temp), pal = taxa_pal[as.numeric(levels(temp))], 
               use_w2 = F , plot_labels = T ,gapv = .01,gap2v = 0 ,xaxs="i",
               yaxs="i",plot_weights = F)
  
  text(par("usr")[1],par("usr")[3]-.008,labels = "     BV",cex=.6, xpd=T)
  text(par("usr")[1],par("usr")[3]-.022,labels = "      HIV",cex=.6, xpd=T)
  
  par(mar=c(4,.1,2,5))

  # KEGG
  barplot( rev(pvalues) , horiz = T , col= taxa_pal[factor(rev(module_color))],
           xaxs="i",yaxs="i",xlim=c(0,1.2*max(pvalues)),ylab="",las=1,
           names.arg = "",xpd = FALSE)
  abline(v=2,xpd=F,col="red",lty=2)
  points( rev(pvalues), seq(1,length(pvalues) )*1.2-.5,pch=21,
          bg=taxa_pal[factor(rev(module_color))] )
  title(xlab="-log10(pvalue)", line=2)
  text( rev(pvalues)+strwidth("L") , seq(1,length(pvalues) )*1.2-.5, 
        adj=0,cex=.9,xpd=T,rev(gsub("\\(\\)","",names(pvalues)) ))
  text( rev(pvalues)+strwidth("L") , seq(1,length(pvalues) )*1.2-.5, 
        labels = rev(sub(" [(].*","",names(pvalues)) ), adj=0,cex=.9,
        col = paste0(taxa_pal[factor(rev(module_color))],90) ,xpd=T)
  
  # GO
  res_list <- pluck(enrich, "GO", "res_list")
  #top_TRX_counts <- pluck(enrich, "GO", "top_TRX_counts")[,samples]
  
  ## Gene module enrichment
  names(res_list) <- unique( gene_module[h$order] )
  
  pvalues <- unlist(lapply(res_list,function(x){ -log10(x$pvalue) [1:N] }))
  # pvalues <- unlist(lapply(res_list,function(x){ x$jaccard_index [1:N] }))

  # pvalues <- unlist(lapply(res_list,function(x){ 
  #   xxx <- -log10(x$pvalue) [1:N]
  #   return( xxx / xxx[1] )
  # }))

  pvalues[is.na(pvalues)] <- 0
  terms <- unlist(lapply(res_list,function(x){ rownames(x) [1:N] }))
  terms[is.na(terms)] <- ""
  terms <- gsub("_", " ", terms)
  terms <- gsub("(GO |KEGG )|(\\w+)", "\\1\\L\\2", terms, perl = TRUE)
  genes <- unlist(lapply(res_list,function(x){ x$genes [1:N] }))
  genes[is.na(genes)] <- ""
  
  #pvalues <- setNames(pvalues,paste0(terms," (",genes,")"))
  pvalues <- setNames(pvalues,paste0(terms,""))
  module_color <- unlist(lapply(names(res_list),function(x){ rep(x,N) }))
  
  barplot( rev(pvalues) , horiz = T , col= taxa_pal[factor(rev(module_color))],
           xaxs="i",yaxs="i",xlim=c(0,1.2*max(pvalues)),ylab="",las=1,
           names.arg = "",xpd = FALSE)
  abline(v=2,xpd=F,col="red",lty=2)
  points( rev(pvalues), seq(1,length(pvalues) )*1.2-.5,pch=21,
          bg=taxa_pal[factor(rev(module_color))] )
  title(xlab="-log10(pvalue)", line=2)
  text( rev(pvalues)+strwidth("L") , seq(1,length(pvalues) )*1.2-.5, 
        adj=0,cex=.9,xpd=T,rev(gsub("\\(\\)","",names(pvalues)) ))
  text( rev(pvalues)+strwidth("L") , seq(1,length(pvalues) )*1.2-.5, 
        labels = rev(sub(" [(].*","",names(pvalues)) ), adj=0,cex=.9,
        col = paste0(taxa_pal[factor(rev(module_color))],90) ,xpd=T)
  
  # dev.copy2pdf(file=paste0("../../../results/ENRICHMENT/","DGE_heatmap_",db,"_",name,".pdf"),
  #     width = 6.5,
  #     height = 6, paper = "a4"
  #     )
}
#max(strwidth(pvalues))
#strwidth("M")

map(db, ~plot_heatmap.fun(enrich_L, .x, "Luminal_gr", "Luminal"))
map(db, ~plot_heatmap.fun(enrich_T, .x, "Tissue_gr", "Tissue"))

pdf("./test.pdf", width=7, height = 5, useDingbats = FALSE)
plot_heatmap.fun(enrich_L, "KEGG", "Luminal_gr", "Luminal")
dev.off()

```

```{r UMAP, echo=FALSE, fig.height=2.5, fig.width=3}
# , fig.height=3, fig.width=3.5
###############
#  TRX UMAP   #
###############
plot_UMAP.fun <- function(top_TRX_counts, group){
  #par(mar=c(4,4,2,6)) #b,l,t,r
  par(mar=c(2.5,2.5,2,3)) #b,l,t,r
  #par(mar=c(10,12,2,12)) #b,l,t,r
  set.seed(1)
  UMAP_TRX <- uwot::umap(t(top_TRX_counts),n_neighbors = 10,
                   metric = "correlation",min_dist = 0.1,spread = 5,
                   negative_sample_rate = 10)
  plot(UMAP_TRX,bg=pal[factor(group)],main="UMAP on top DGE RNAseq",
       pch=21,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F)
  mtext("UMAP1", side=1, line=0.7, cex=0.8)
  mtext("UMAP2", side=2, line=0.7, cex=0.8)
  legend(par("usr")[2],par("usr")[4],title.adj = 0,
         #legend = levels(factor(metadata$Luminal_gr)),
         legend = sort(unique(group)),
         xjust = 0,yjust = 1, cex=0.7,
         bty = "n",pch = 21,pt.bg = pal,pt.cex = 1,xpd=T)
}

plot_UMAP.fun(enrich_L$GO$top_TRX_counts, metadata$Luminal_gr)
m <- filter(metadata, !is.na(Tissue_gr)) 
plot_UMAP.fun(enrich_T$GO$top_TRX_counts[,m$ID], m$Tissue_gr)

#plot_UMAP.fun(enrich_T$GO$top_TRX_counts, metadata$Tissue_gr)
dev.copy2pdf(file=paste0("./UMAP",".pdf"),
    width = 4,
    height = 3.5, paper = "a4" #units = "cm", res = 300 
    )


```


<!-- ## Pairwise DGEs and enrichment 
```{r Pairwise functions, message=FALSE, warning=FALSE, include=FALSE}
## individual group comparision ##
# for information about how to get pairwise comparison se:
# edgeR User’s Guide section 3.4.2 Blocking and 4.2.8  Differential expression
# https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
###########################
# PAIRWISE DEG's FUNCTION #
###########################
pairwise_dge.fun <- function(group, fit, design) {
  groups <- metadata[sample_use, group]
  levels <- as.character(sort(unique(groups),na.last = NA))
  
  # without intercept:
  n <- combn(levels,2) 
  g <- map_chr(seq_along(1:ncol(n)), ~paste(n[1,.x], n[2,.x], sep = "-"))
  s <- seq_along(1:ncol(n)) %>% set_names(g)
  #name <- map_chr(seq_along(1:ncol(n)), ~paste(str_extract(n[1,.x], "^g."), str_extract(n[2,.x], "^g."), sep = "_"))
  
  # with intercept:
  g <- sub(paste0("^(",levels[1], ")-(.*)"), '-\\2', g)
  
  CONTRASTS <- makeContrasts( contrasts=g,
                              levels = design )
  
  #fit2 <- contrasts.fit(fit, CONTRASTS)
  #lrt_ <- glmLRT(fit2, coef = 2:5)
  lrt <- map(s, ~ glmLRT(fit, contrast = CONTRASTS[,.x]))
  top <- map(lrt, ~topTags(.x,adjust.method = "BH",n = "all",sort.by = "p.value")[[1]]) %>% map(., ~rownames_to_column(.x, var = "Genes"))
  
  results <- map_df(lrt, ~decideTests(.x, method="separate", p.value = 0.05))
  summary <- bind_cols( "FDR<0.05"= c("Down","NotSig","Up"), summary(results))
  colnames(summary) <- c("FDR<0.05", g)
  top <- c(summary = list(summary), top)
  
return(list(lrt=lrt, top=top, CONTRASTS=CONTRASTS))
}

################################
# PAIRWISE ENRICHMENT FUNCTION #
################################
enrich_list <- function(x, gmt_list){
  temp <- compute_enrichment(genes = x,
                            gmt_list = gmt_list,
                            min_terms_pathway = 10,
                            max_terms_pathway = 300,
                            min_overlap = 3,
                            sort_by_pvalue = T)
  temp <- temp[!grepl("REGULATION",rownames(temp)),]
  temp <- rownames_to_column(temp, var = "Terms")
  return(temp)
} 
```

```{r Pairwise tables, include=FALSE}
##################
# PAIRWISE DEG's #
##################
pairwise_dge_L <- pairwise_dge.fun(group = "Luminal_gr", 
                                   fit=dge_L$fit, 
                                   design=dge_L$design )
pairwise_dge_T <- pairwise_dge.fun(group = "Tissue_gr",
                                   fit=dge_T$fit,
                                   design=dge_T$design)

# save TopTags 
top_dfs <- list(Luminal = pairwise_dge_L$top, Tissue = pairwise_dge_T$top) 
imap(top_dfs, ~write.xlsx(.x,file=paste0("../../../results/DE/","dge_pairwise_",.y,"_conf",".xlsx")))


###############################
# PAIRWISE ENRICHMENT GO/KEGG #
###############################
source("../enrichment_function.R")
gmt_list_KEGG <- fgsea::gmtPathways("../../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
gmt_list_GO <- fgsea::gmtPathways("../../../supplementary_files/c5.bp.v6.2.symbols.gmt.txt")

#info_logFC <- map(top_dfs[[1]], ~table(abs(.x$logFC) > 1))
#info_PValue <- map(top_dfs[[1]], ~table(.$PValue < 0.01))

l_top <- list(Luminal = pairwise_dge_L$top[-1], Tissue = pairwise_dge_T$top[-1])
# up and down regulated genes seperatly
t_split <- modify_depth(l_top, 2, ~ .x %>%
        dplyr::filter(., .$PValue < 0.01) %>%
        #dplyr::filter(., abs(.$logFC) < 1) %>%
        mutate(., reg = ifelse(.$logFC > 0, "up", "down")) %>%
        split(., .$reg)
  )
# up and down regulated genes together
t <- modify_depth(l_top, 2,~ .x %>%
        dplyr::filter(., .$PValue < 0.01) #%>%
        #dplyr::filter(., abs(.$logFC) < 1) 
  )

## GO enrichment 
GO_enrichment <- map_depth(t, 2, ~enrich_list(.x$Genes, gmt_list_GO))
GO_enrichment_split <- map_depth(t_split, 3, ~enrich_list(.x$Genes, gmt_list_GO)) %>%
  map_depth(., 2, ~bind_rows(.x, .id = "regulation")) %>% 
  set_names(., paste0(names(.), "_split"))

## KEGG enrichment
KEGG_enrichment <- map_depth(t, 2, ~enrich_list(.x$Genes, gmt_list_KEGG))
KEGG_enrichment_split <- map_depth(t_split, 3, ~enrich_list(.x$Genes, gmt_list_KEGG)) %>%
  map_depth(., 2, ~bind_rows(.x, .id = "regulation")) %>% 
  set_names(., paste0(names(.), "_split"))
 
## save files
enrich_KEGG <- list(KEGG_enrichment, KEGG_enrichment_split) %>% flatten()
enrich_GO <- list(GO_enrichment, GO_enrichment_split) %>% flatten()

imap(enrich_GO, ~write.xlsx(.x, 
file=paste0("../../../results/ENRICHMENT/Pairwise/","GO_enrichment_pairwise_",.y,"_conf",".xlsx"))) 
imap(enrich_KEGG, ~write.xlsx(.x, 
file=paste0("../../../results/ENRICHMENT/Pairwise/","KEGG_enrichment_pairwise_",.y, "_conf",".xlsx")))

```
 -->
