---
title: "Integrated multi-omics analysis reveals Lactobacillus anti-inflammatory process in vaginal tissue"
subtitle: A demonstration of Rmarkdown using Herman Bumpus' data
author: |
  Author One^1^,
  Author Two^2^,
  Author Three^1,2^
keywords: "pandoc, r markdown, knitr"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  bookdown::pdf_document2:
      keep_tex: true
      toc: false
  bookdown::html_document2: default
  bookdown::word_document2: default
header-includes:
- \usepackage{fancyhdr}
- \pagestyle{fancy}
- \fancyhead[L]{XXXXXXXXXXX et al.}
- \fancyhead[R]{MANUSCRIPT SHORT TITLE}
- \usepackage{lineno}
- \linenumbers
linestretch: 1.5
link-citations: yes
linkcolor: blue
csl: nature.csl
cslreferences: nature.csl
bibliography: refs.bib
editor_options:
  chunk_output_type: console #inline
#params:
  #taxa_level: "Genus_lacto"
---
\footnotetext[1]{University of Nowhere}
\footnotetext[2]{University of Somewhere}
\footnotetext[3]{University of Lalaland}


```{r Setup..., message=FALSE, warning=FALSE, include=FALSE}
params <- "Genus_lacto"
#setwd("/Users/vilkal/work/Brolidens_work/Projects/Gabriella_repo/reports/rmarkdown/manuscript")
options(tinytex.verbose = TRUE)
knitr::opts_chunk$set(out.width = '100%',
                      dpi=600,
                      fig.width=10,
                      echo = FALSE,
                      # fig.pos = 'p', # Places figures on their own pages
                      eval = TRUE,
                      results = 'hide',
                      warning = FALSE,
                      message = FALSE )

# I usually load my libraries up front to keep things organized
library(bookdown)
library(knitr)
library(kableExtra)
library(dplyr)
library(tidyverse)
library(igraph)
library(phyloseq)
library(vegan)
library(ggpubr)
library(rafalib)
library(scales)
library(dendextend)
library(edgeR)
library(openxlsx)
remotes::install_github("czarnewski/niceRplots",force=T)
library(niceRplots)


```



Abstract
===============================================================================

Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract Abstract
test test test


\clearpage

Introduction
================================================================================

Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction ([@Johnston1972]), Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction Introduction **Introduction Introduction** *Introduction* Introduction ([@Darwin1859]^,^[@Bumpus1898]) .

Problem / question to answer

\clearpage


Results
===============================================================================

```{r, echo = FALSE, eval = TRUE}
# Load reasult from the analysis
metadata <- read.csv2("../../../results/metadata_integration.csv",row.names = 1)
TissueGroupM4 <- read.csv("../../../data/210411_TissueGroupM4.csv", row.names = 1)
#patient_groups <- factor(setNames(metadata$joint_clustering,metadata$ID))
patient_groups <- factor(setNames(metadata$ModifiedM4,metadata$ID))
bac_communities <- as.matrix(read.csv2("../../../results/bacterial_communities.csv",row.names = 1))[,1]


SNN_patients <- read.csv("../../../results/patient_SNN_graph.csv",row.names = 1)
SNN_bacteria <- read.csv("../../../results/bacteria_SNN_graph.csv",row.names = 1)
clean_Ftest <- read.csv2("../../../results/metadata_association_results_complete.csv",row.names = 1)
datasets_all_samples <- readRDS("../../../results/datasets_all_samples.RDS")

picrust_data <- read.table("../../../data/pred_metagenome_unstrat.tsv",header = T,row.names = 1)
picrust_KO_seq <- read.table("../../../data/KO_predicted.tsv",header = T,row.names = 1)
picrust_taxa <- read.table("../../../data/SeqID_to_taxa.csv",header = T,row.names = 1,sep = ",")
```


```{r}
#color palettes
Set1 <- c("#4DAF4A", "#377EB8", "#E41A1C", "#FF7F00", "#984EA3", "#FFFF33", "#A65628", "#F781BF", "#999999")
pal <- c(Set1,RColorBrewer::brewer.pal(8,"Set2"),
         RColorBrewer::brewer.pal(9,"Pastel1"),RColorBrewer::brewer.pal(8,"Pastel2")) #color pallete for plots
heat_pal <- c("#000000", colorRampPalette(c("#000000","grey5","grey30","orange3","yellow","yellow","white"))(90))
cor_pal <- colorRampPalette(c("navy","white","firebrick"))(90)
```


```{r, echo = FALSE, eval = TRUE}
#Assign names to patient groups
# big_groups <- c(
#   g1_L.crispatus= "1 NonIners",
#   g2_L.iners=    "2 Iners",
#   g3_Gard=   "3 Gard",
#   g4_Prev=        "4 Prev",
#   g5_Oth=   "5 Other"
# )
big_groups <- c(
  g1_L.crispatus= "NonIners",
  g2_L.iners=     "Iners",
  g3_Gard=        "Gard",
  g4_Prev=        "Prev",
  g5_Oth=         "Oth"
)
# big_groups <- c(
#   g2_L.iners=    "5",
#   g3_L.iners2=   "4",
#   g1_L.crispatus= "2",
#   g5_Gard=        "3",
#   g6_Gard.Prev=   "6",
#   g7_Gard.Prev2=  "8",
#   g4_Gard.Lacto=  "7",
#   g8_Patho=       "1"
# )

group_annotation <- as.character(patient_groups)
group_annotation <- names(big_groups)[match(group_annotation,big_groups)]
group_annotation <- setNames(group_annotation,names(patient_groups))
#group_annotation <- as.factor(setNames(group_annotation,names(patient_groups)))

metadata$group_annotation <- group_annotation
#write.csv2(metadata,"../../../results/metadata_integration_.csv",row.names = T)

# modifications nessecary for dge analyser:
metadata <- metadata %>%
  mutate(ModifiedM4 = as.character(ModifiedM4_)) %>%
  mutate(sexwork_months=ifelse(is.na(.$sexwork_months),
                               mean(metadata$sexwork_months,na.rm = TRUE),.$sexwork_months))%>%
  mutate(Contraception=ifelse(.$Contraception=="no HC", "no_HC","DMPA")) %>%
  cbind(TissueGroupM4) %>%
  #as_data_frame() %>%
  #left_join(TissueGroupM4, by="ID") %>%
  #mutate(TissueGroupM4= factor(.$TissueGroupM4)) %>%
  select(ID, PatID, HIVstatus, Contraception, age, sexwork_months, everything())

#write.csv2(metadata,"../../../results/metadata_integration_.csv",row.names = T)

```



```{r, echo = FALSE, eval = TRUE}
bac_communities <- factor(paste0("BC",sprintf("%02d",bac_communities)))
bacs <- names(bac_communities)

#Assign names to bacterial communities
BC_groups <- c(
  BC01= "BC01_Peptococcus",
  BC02= "BC02_Lactobacillus.spp",
  BC03= "BC03_Micrococcus",
  BC04= "BC04_Gardnerella",
  BC05= "BC05_Paracoccus",
  BC06= "BC06_Pseudomonas",
  BC07= "BC07_Prvotella",
  BC08= "BC08_Faecalibacterium",
  BC09= "BC09_Escherichia/Shigella"
)
# BC_groups <- c(
#   BC01= "BC01_Bifidobacterium",
#   BC02= "BC02_Mycoplasma",
#   BC03= "BC03_Lactobacillus.spp",
#   BC04= "BC04_Escherichia.sp",
#   BC05= "BC05_Other1",
#   BC06= "BC06_Other2",
#   BC07= "BC07_Corynebacterium.spp",
#   BC08= "BC08_Other3",
#   BC09= "BC09_Ureaplasma.sp",
#   BC10= "BC10_Campylobacter",
#   BC11= "BC11_Mobiluncus_BVAB",
#   BC12= "BC12_Faecalibacterium",
#   BC13= "BC13_Dialister",
#   BC14= "BC14_P.timonesis",
#   BC15= "BC15_T.scotoductus"
# )

bac_communities <- as.character(bac_communities)
bac_communities <- BC_groups[match(bac_communities,names(BC_groups))]
bac_communities <- as.factor(setNames(bac_communities,bacs))
```



**Joint analysis of vaginal microbiome reveals distinct patient subgroups**

To understand the longitudinal and tissue-specific microbiome profile in vaginal samples, `r nrow(metadata)` adult female sex workers were enrolled in [...]. Among those, `r sum(metadata$HIVstatus=="pos")` were previously tested positive for HIV during the cohort's sampling procedure. [Describe here what was done and when, which samples, which tissues].


To be able to better undertand the differences in microbiome profile across all datasets collected, we  performed a joint graph-based clustering analysis in order to identify co-regulated bacterial communities (see "Methods" section for details). A total of `r length(unique(bac_communities))` bacterial communities were identified.


Noticebly, bacterial community `r bac_communities["Lactobacillus iners"]` consisted only of Lactobacillus species (*`r names(bac_communities)[bac_communities %in% bac_communities["Lactobacillus iners"]]`*).


Patients were thus subdivided into `r length(unique(group_annotation))` groups,

Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results


\clearpage


**Indentification of bacterial communities metabolic processes linked to Lactobacilli**

Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results


\clearpage



**Indentification of bacterial communities metabolic processes linked to Lactobacilli**


Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results Results


\clearpage


Discussion
================================================================================

I have analysed data collected by Herman Bumpus [@Bumpus1898] on the relationship between sparrow (*Passer domesticus*) total length and surival following an unusually severe storm. I found that sparrows that died in the storm were longer than sparrows that survived, which suggests that higher sparrow body length decreased survival. Of course, it is not possible to definitively conclude a causal relationship between any aspect of body size and sparrow survival<!--- BD Note: maybe explain this better --->, and even the available data collected by Bumpus would permit a more thoughtful analysis than that conducted in this study (see [Appendix Table 1](#appendix)).

<!---

Here is one way to add some more detailed comments into the manuscript itself, though this can also be done within the text (see above).

--->

Overall, this document demonstrates how high quality, professional looking documents can be written using Rmarkdown. The [underlying code](https://github.com/StirlingCodingClub/Manuscripts_in_Rmarkdown/blob/master/ms.Rmd) for this manuscript is publicly available, along with [accompanying notes](https://stirlingcodingclub.github.io/Manuscripts_in_Rmarkdown/Rmarkdown_notes.html) to understand how it was written. By using Rmarkdown to write manuscripts, authors can more easily use version control (e.g., git) throughout the writing process. The ability to easily integrate citations though BibTeX, LaTeX tools, and dynamic R code can also make writing much more efficient and more enjoyable. Further, obtaining the benefits of using Rmarkdown does not need to come with the cost of isolating colleagues who prefer to work with Word or LaTeX because Rmarkdown can easily be converted to these formats (in the case of Word, with the push of a button). By learning all of the tools used in this manuscript, readers should have all of the necessary knowledge to get started writing and collaborating in Rmarkdown.


\clearpage


Methods
================================================================================


\clearpage

References
================================================================================

<div id="refs"></div>


\clearpage



<a name="appendix">Appendix Table 1</a>
================================================================================

An example table is shown below, which includes all of the variables collected by @Bumpus1898 for the first 10 measured sparrows. The full data set can be found online in [GitHub](https://github.com/StirlingCodingClub/Manuscripts_in_Rmarkdown/blob/master/data/Bumpus_data.csv).

```{r, echo = FALSE, fig.pos='H',fig.show='hold'}
library(knitr)
```


\clearpage


FIGURES (MAIN)
================================================================================


## Figure 1

```{r, echo = FALSE, eval = TRUE , fig.asp=9/10}
# Define Layout and set labels
layout(matrix(c(1,1,1,1,      3,3,3,3,          3,4,4,4,
                2,2,2,2,      3,3,3,3,          3,5,5,5,
                6,6,6,6,      9,9,9,9,          9,9,9,9,
                7,7,7,7,     10,10,10,10,     10,10,10,10,
                8,8,8,8,     10,10,10,10,     10,10,10,10 ) , ncol= 12,byrow = T),
       widths = 1,heights = c(1,1,.7,.7,.7))
# layout.show(12)
figlabels <- letters


#################################
# COHORT SCHEMATIC ILLUSTRATION #
#################################
par(mar=c(0,0,0,0))
fig1a <- magick::image_read("schematics.pdf",density = 900)
plot(c(0,1),c(0,1),axes=F,type="n",xlab="",ylab="",xaxs="i",yaxs="i")
rasterImage(fig1a, 0, 0, 1, 1)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]







####################
# PATIENT GROUPING #
####################
g <- graph_from_adjacency_matrix(as.matrix(SNN_patients), mode = "undirected",diag = F,weighted = T)
set.seed(1)
l <- layout_with_fr(g,niter=3000,start.temp=30)

par(mar=c(0,0,0,6))
plot( g , vertex.label.cex=0.000001 , vertex.color = pal[factor(group_annotation)] ,
      edge.width=  ( E(g)$weight / max(E(g)$weight)) ,
      vertex.size=10,main="", layout=l,
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(g)$weight / max(E(g)$weight) * 88 )+1 ] )
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(factor(group_annotation)),
       xjust = 0,yjust = 1,
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1,xpd=T)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]




##############################
# METADATA CROSS-ASSOCIATION #
##############################
par(mar=c(12,1,1,12))
set.seed(1)
noise <- matrix(rnorm(prod(dim(clean_Ftest)),mean = 0,sd = 1e-10),nrow(clean_Ftest),ncol(clean_Ftest))
o <- hclust(as.dist( (1-cor(clean_Ftest+noise)/2) ),method = "ward.D2")$order

sizes <- as.matrix(clean_Ftest[o,o])
sizes[sizes > 15] <- 15
sizes <- sizes / max(sizes)
image( t(sizes [nrow(sizes):1,]),axes=F,
       col=colorRampPalette(c("grey95","grey70","orange3","firebrick"))(90) )
text(  par("usr")[2]+.02 ,seq(1,0,length.out = nrow(sizes)) ,
       labels = rownames(sizes), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
text( seq(0,1,length.out = nrow(sizes)) , par("usr")[3]-.02 ,
      labels = rownames(sizes), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]




###################################
# BACTERIAL VAGINOSIS COMPARISSON #
###################################
par(mar=c(1,1,1,1))
empty_plot(frame = T)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]

empty_plot(frame = T)




###################################
# BACTERIAL VAGINOSIS COMPARISSON #
###################################
par(mar=c(3,5,2,1))
x <- datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]]
xL <- rowsum(x , grepl("Lactobacillus",rownames(x)) )[2,] / sum(grepl("Lactobacillus",rownames(x)))
xM <- rowsum(x , grepl("Mobiluncus",rownames(x)) )[2,] / sum(grepl("Mobiluncus",rownames(x)))
xG <- rowsum(x , grepl("Gardnerella",rownames(x)) )[2,] / sum(grepl("Gardnerella",rownames(x)))

mL <- metadata$BV_Lactobacillus_v3 ; mL[is.na(mL)] <- 0
mM <- metadata$BV_Monbilicus_v3    ; mM[is.na(mM)] <- 0
mG <- metadata$BV_Vaginal_Garda_v3 ; mG[is.na(mG)] <- 0

barlist( data = rbind(  BV=mL ,
                        counts_16S=xL),
         main = "Lactobacillus",
         genes = c("BV","counts_16S"),
         clustering = group_annotation,
         draw_mean_lines=F,col = pal,srt=20)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]


barlist( data = rbind(  BV=mM ,
                        counts_16S=xM),
         main = "Monbilicus",
         genes = c("BV","counts_16S"),
         clustering = group_annotation,
         draw_mean_lines=F,col = pal,srt=20)

barlist( data = rbind(  BV=mG ,
                        counts_16S=xG),
         main = "Gardanerella",
         genes = c("BV","counts_16S"),
         clustering = group_annotation,
         draw_mean_lines=F,col = pal,srt=20)





###################################
# BACTERIAL VAGINOSIS COMPARISSON #
###################################
par(mar=c(2,5,1,12))

shann <- vegan::diversity(t(x),index = c("shannon") )
simp <- vegan::diversity(t(x),index = c("simpson") )
invsimp <- vegan::diversity(t(x),index = c("invsimpson") )
divers <- as.matrix(rbind(shann,simp,invsimp) )
divers[is.infinite(divers)] <- 0
rownames(divers) <- c("Shannon","simpson","Inv Simpson")
# barlist( data = divers,
#          main = "",
#          genes = c("Shannon","simpson","Inv Simpson"),
#          clustering = group_annotation,
#          draw_mean_lines=F,col = pal,srt=30)
  barlist( data = divers[,order(group_annotation)],
         main = "",
         genes = c("Shannon","simpson","Inv Simpson"),
         # clustering = group_annotation,
         draw_mean_lines=F,col = pal[factor(group_annotation[order(group_annotation)])])




#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]


#######################
# BACTERIAL ABUNDANCE #
#######################
# for Genus_lacto taxonomy:
taxa <- read_csv("../../../results/SeqID_to_gen-lacto_tax-oth.csv") %>%
  select(-Sequence, -Seq_ID, -taxa_other, -Species) %>%
  unique() %>%
  filter(!(is.na(.$Genus_lacto))) %>%
  filter(!(.$Genus == "Granulicatella" & .$Family == "Enterococcaceae")) %>%
  column_to_rownames(., var = "Genus_lacto")
# For other_taxa taxonomy:
#taxa <- read.csv2("../../../results/taxonomy.csv",row.names = 1,stringsAsFactors = F,as.is = T)
par(mar=c(4,5,1,12))
temp <- t(t(2^x-1)/colSums(2^x-1))

taxa[is.na(taxa)] <- "NA"
taxa <- taxa[rownames(temp),]
temp <- rowsum(temp , taxa$Class)
temp[is.nan(temp)] <- 0
o <- order( rowSums(temp) , decreasing = T)
sample_o <- order( group_annotation )
temp <- temp[o,sample_o]*100
#temp <- temp[-grep("NA",rownames(temp)),]
temp <- rbind(temp[1:16,], Other = colSums(temp[17:nrow(temp),]) )

taxa_pal <- c(RColorBrewer::brewer.pal(8,"Pastel2"),RColorBrewer::brewer.pal(8,"Pastel1"),"grey90")
barplot( temp, las=2 , border = NA , yaxs="i", xaxs="i",ylab="abundance",
         col = taxa_pal[factor(rownames(temp),levels = rownames(temp))] ,names.arg = rep("",ncol(temp)))
legend(x = par("usr")[c(2)],
       y = par("usr")[c(4)],title.adj = 0,
       legend = rownames(temp),xjust = 0,yjust = 1,
       bty = "n",pch = 22,pt.cex = 2,cex = 1,xpd=T,
       pt.bg = taxa_pal)
#
# points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-1,
#         col= c("grey","red")[(!is.na(metadata$RepeatedPCR))+1][sample_o],
#         pch=16,cex=.5,xpd=T)
# points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-2,
#         col= c("grey","red")[as.numeric(metadata$HIVstatus)][sample_o],
#         pch=16,cex=.5,xpd=T)
  points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-2,
        col= c("red","orange","grey")[as.numeric(metadata$BV_Diagnosis_v3)][order( group_annotation )],
        pch=15,cex=.8,xpd=T)
  text(par("usr")[1],par("usr")[3]-2,labels = "BV Diagnosis",pos=2,cex=.8,xpd=T)

  points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-6,
        col= c("grey","red")[as.numeric(metadata$HIVstatus)][order( group_annotation )],
        pch=15,cex=.8,xpd=T)
  text(par("usr")[1],par("usr")[3]-6,labels = "HIVstatus",pos=2,cex=.8,xpd=T)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]
```

**Figure 1. Identification of patient groups.**
(**a**) Schematic representation of ########## .
(**b**) Schematic representation of ########## .
(**c**) Schematic representation of ########## .
(**d**) Schematic representation of ########## .


\clearpage


## Figure 2

```{r, echo = FALSE, eval = TRUE,fig.asp=.6}
# Define Layout and set labels
layout(matrix(c(2,3,4,5,      6,7,11,11,          11,11,11,11,
                1,1,1,1,      1,8,11,11,          11,11,11,11,
                1,1,1,1,      1,9,12,12,         12,12,12,12,
                1,1,1,1,      1,10,13,13,         13,13,13,13,
                0,0,0,0,      0,0,14,14,         14,14,14,14 
                ) , 
              ncol= 12,byrow = T),widths = 1,heights = c(1.5,1.2,1.2,1.2,1.2))
# layout.show(3)
figlabels <- letters
#
# m <- matrix(as.character(c(1,2,3,4)),nrow = 2,ncol = 2)
# m
# m2 <- matrix(as.character(c(1,2,3,"m")),nrow = 2,ncol = 2)
# m2
#
# matrix(c(1,2,j,3),nrow = 2,ncol = 2)
# layout_matrix <- function(data, nrow, ncol){
#   return( matrix(as.character(data), nrow = nrow, ncol = ncol)  )
# }
# layout_matrix2 <- function(data, nrow, ncol, widths=c(1,2), heights=c(2,3), res=res){
#   temp <- matrix(as.character(data), nrow = nrow, ncol = ncol)
#   widths <- rep(widths,ncol(temp))[1:ncol(temp)]
#   heights <- rep(heights,nrow(temp))[1:nrow(temp)]
#   
#   cumsum(widths)*100/sum(widths)
#   cumsum(heights)*100/sum(heights)
#   
#   temp2 <- matrix(1,nrow = nrow(data),ncol = ncol(data))
#   temp3 <- temp2 * heights
#   temp4 <- t( t( temp2 ) * widths )
#   temp2 <- matrix( "" , ncol = res , nrow = res )
#   
#   
#   for(i in 1:nrow(temp)){
#     for(j in 1:ncol(temp)){
#       temp2 <-
#     }
#   }
#   temp <- apply(temp,1,function(x){rep( x , widths/ )})
#   return(   )
# }
#
# m <- layout_matrix(c(1,2,3,4),2,2)
# m2 <- layout_matrix(c(1,2,3,"m"),2,2)




#########################
# BACTERIAL COMMUNITIES #
#########################
gB <- graph_from_adjacency_matrix(as.matrix(SNN_bacteria), mode = "undirected",diag = F,weighted = T)
# set.seed(1)
lB <- layout_nicely(gB,niter=3000,start.temp=30)

par(mar=c(0,0,0,10))
plot( gB , vertex.label.cex=0.000001 , vertex.color = pal[factor(bac_communities)] ,
      edge.width=  ( E(gB)$weight / max(E(gB)$weight)) ,
      vertex.size=10,main="",
      edge.color=colorRampPalette(c("grey95","black"))(90) [ round( E(gB)$weight / max(E(gB)$weight) * 88 )+1 ] ,layout=lB)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(factor(bac_communities)),xjust = 0,yjust = 1,
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1,xpd=T)

#add label
add_letter(figlabels[1]); figlabels <- figlabels[-1]


par(mar=c(1,1,1,1))
for(j in sort(unique(bac_communities)) ){
  empty_plot()
  text( 0,1,labels = paste0(rownames(SNN_bacteria)[ bac_communities == j ],collapse = "\n"),adj=c(0,1),cex=.5)
  title(main=j,cex.main=.5)
}





###################################
# BACTERIAL VAGINOSIS COMPARISSON #
###################################
# for Genus_lacto taxonomy:
taxa <- read_csv("../../../results/SeqID_to_gen-lacto_tax-oth.csv") %>%
  select(-Sequence, -Seq_ID, -taxa_other, -Species) %>%
  unique() %>%
  filter(!(is.na(.$Genus_lacto))) %>%
  filter(!(.$Genus == "Granulicatella" & .$Family == "Enterococcaceae")) %>%
  column_to_rownames(., var = "Genus_lacto")
# for other taxa
# taxa <- read.csv2("../../../results/taxonomy.csv",row.names = 1,stringsAsFactors = F,as.is = T)
# taxa[is.na(taxa)] <- "NA"

#par(mar=c(1,10,1,15))
par(mar=c(1,10,1,15))
x <- taxa[match(rownames(SNN_bacteria),rownames(taxa)),2]
y <- taxa[match(rownames(SNN_bacteria),rownames(taxa)),3]

plot_sankey( data.frame( x[y!="NA"] , bac_communities[y!="NA"]), color_by = 1,
     plot_labels = T, plot_weights = F,order_1_by = "total",
     order_2_by = "total",use_w1 = T,use_w2 = T,gapv = .02,gap2v = 0,smoothing = .1,
     pal=hue_pal()(length(unique(x))),xaxs="i",yaxs="i")

add_letter(figlabels[1]); figlabels <- figlabels[-1]




##############################################
# BACTERIAL COMMUNITY DIFFERENTIAL ABUNDANCE #
##############################################
module_means_per_dataset <- lapply(datasets_all_samples[-1],function(x){
  return( rowsum( x[rownames(SNN_bacteria),] , group = bac_communities ) / c(table(bac_communities)) )
})
#
par(mar=c(2,10,2,6))
plot_dots_difference(data = module_means_per_dataset[[2]],
                     data2 = module_means_per_dataset[[3]],
                     clustering = as.character(group_annotation),
                     clustering2 = as.character(group_annotation),
                     genes = rownames(module_means_per_dataset[[1]]),
                     main = "CVL3 vs CVL2",cex.row = .8,srt=30)

plot_dots_difference(data = module_means_per_dataset[[2]],
                     data2 = module_means_per_dataset[[1]],
                     clustering = as.character(group_annotation),
                     clustering2 = as.character(group_annotation),
                     genes = rownames(module_means_per_dataset[[1]]),
                     main = "CVL3 vs tissue_V3",cex.row = .8,srt=30)


#####################
# BACTERIAL PICRUST #
#####################
picrust_data <- round(picrust_data[ rowSums(picrust_data>5)>=10, ])
sample_use <- colnames(picrust_data)

ko_names <- readLines('../../../data/ko_info.tsv')
ko_names <- strsplit(ko_names,"\t")
ko_names <- setNames(unlist(lapply(ko_names,function(x){x[2]})), unlist(lapply(ko_names,function(x){x[1]})))
ko_names <- sub(" [[].*","",ko_names)
ko_names <- sub(".*; ","",ko_names)

named_picrust_data <- rowsum( picrust_data , group = ko_names[ rownames(picrust_data) ] )
named_picrust_data[1:20,1:3]
dim(named_picrust_data)
named_picrust_data <- round(named_picrust_data[ rowSums(named_picrust_data>5)>=5, ])

design <- model.matrix(~ group_annotation+Contraception+HIVstatus+age+sexwork_months, 
                       data=metadata[sample_use,])

y <- DGEList(counts=named_picrust_data[,sample_use])
y <- calcNormFactors(y,method = "TMM")
y <- estimateGLMCommonDisp(y)
y <- estimateGLMTagwiseDisp(y)
fit <- glmFit(y, design)

lrt <- glmLRT(fit,coef=2:9)
top <- topTags(lrt,adjust.method = "BH",n = "all",sort.by = "p.value")[[1]]
colnames(top) <- sub("group_annotation","",colnames(top))
top <- cbind(LogFC.intercept=0,top) # with intercept


saveRDS( y , "../../../results/Picrust_EdgeR_estimations.rds" )
write.csv2( top , "../../../results/Picrust_EdgeR_top_dge.csv" )


top_dge <- (top$FDR < 1e-5) #& (rowSums( abs(top[,grep("logFC",colnames(top))]) >= log2(1.5) ) >= 1)
top_dge <- rownames(top)[ top_dge ]

top_picrust_data <- cpm(y)[top_dge, ]
top_picrust_data <- t(apply(top_picrust_data,1,function(x){scale(x,T,T)}))
top_picrust_data[top_picrust_data > 5] <- 5
colnames(top_picrust_data) <- colnames(picrust_data)


layout(matrix(c(1,2,3,4,
              1,2,3,4),
              nrow = 2,ncol = 4,byrow = T),widths = c(1,2,.5,6))
par(mar=c(4,.1,2,.1))

h <- hclust( as.dist( (1- cor(t(top_picrust_data)))/2 ), method = "ward.D2")
plot( rev(as.dendrogram(h)) ,xlim=c(max(h$height),-1), horiz = T, leaflab = "none",xaxs="i",yaxs="i",axes=T)
cutoff <- 4
abline(v=cutoff,xpd=F,col="red",lty=2)
gene_module <- cutree(h, h = cutoff)

points( rep(-.5,length(gene_module)),
        seq(length(gene_module),1,length.out = length(gene_module)),
        col=taxa_pal[factor(gene_module[h$order])],
        pch=16,cex=1,xpd=T)

image( t(top_picrust_data[h$order,][nrow(top_picrust_data):1,order(group_annotation[sample_use])]),
       col = colorRampPalette(c("navy","navy","grey95","firebrick4","firebrick4") )(99),
       breaks = seq(-5,5,length.out = 100),axes=F)
points( seq(0,1,length.out = length(sample_use)),rep(par("usr")[4],length(sample_use)),
        col=pal[factor(group_annotation[sample_use])[ order(group_annotation[sample_use])]],
        pch=16)
points( seq(0,1,length.out = ncol(top_picrust_data)),rep(par("usr")[3],ncol(top_picrust_data))-.005,
        col= c("red","orange","grey")[as.numeric(metadata$BV_Diagnosis_v3)][order( group_annotation )],
        pch=16,xpd=T)
points( seq(0,1,length.out = ncol(top_picrust_data)),rep(par("usr")[3],ncol(top_picrust_data))-.01,
        col= c("grey","red")[as.numeric(metadata$HIVstatus)][order(group_annotation[sample_use])],
        pch=16,xpd=T)



source("../enrichment_function.R")
gmt_list <- fgsea::gmtPathways("../../../data/KEGG_pathways_to_KO.tsv")
kegg_names <- read.delim('../../../data/KEGG_pathways_info.tsv',header = F)
names(gmt_list) <- kegg_names [ match ( names(gmt_list), as.character(kegg_names[,1]) ) , 2 ]
gmt_list <- lapply(gmt_list , function(x){
  na.omit(ko_names[x])
})
res_list <- lapply( unique( gene_module[h$order] ),function(x){
  temp <- compute_enrichment(genes = names(gene_module)[gene_module==x],
                            gmt_list = gmt_list,
                            min_terms_pathway = 10,
                            max_terms_pathway = 300,
                            min_overlap = 3,
                            sort_by_pvalue = T)
  temp <- temp[!grepl("REGULATION",rownames(temp)),]
  return(temp)
} )
names(res_list) <- unique( gene_module[h$order] )

pvalues <- unlist(lapply(res_list,function(x){ -log10(as.numeric(x$pvalue)) [1:5] }))
pvalues[is.na(pvalues)] <- 0
terms <- unlist(lapply(res_list,function(x){ rownames(x) [1:5] }))
terms[is.na(terms)] <- ""
genes <- unlist(lapply(res_list,function(x){ x$genes [1:5] }))
genes[is.na(genes)] <- ""

pvalues <- setNames(pvalues,paste0(terms," (",genes,")"))
module_color <- unlist(lapply(names(res_list),function(x){ rep(x,5) }))

temp <- factor( rev(gene_module), levels = unique( gene_module[h$order] ))
plot_sankey( data.frame(temp,temp), pal = taxa_pal[as.numeric(levels(temp))] , use_w2 = F , plot_labels = T ,gapv = .01,gap2v = 0 ,xaxs="i",yaxs="i",plot_weights = F)

par(mar=c(4,.1,2,20))
barplot( rev(pvalues) , horiz = T , col= taxa_pal[factor(rev(module_color))],
         xaxs="i",yaxs="i",xlim=c(0,1.2*max(pvalues)),ylab="",xlab="-log10(pvalue)",las=1,names.arg = "")
points( rev(pvalues), seq(1,length(pvalues) )*1.2-.5,pch=21,bg=taxa_pal[factor(rev(module_color))] )
text( rev(pvalues)+1 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,xpd=T,
      labels = rev(names(pvalues)),xpd=T)
text( rev(pvalues)+1 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,
      labels = rev(sub(" [(].*","",names(pvalues)) ),col = paste0(taxa_pal[factor(rev(module_color))],90) ,xpd=T)
# text( rep(max(pvalues)+.5,length(pvalues)) , seq(length(pvalues),1 )*1.2-.5,
#       adj=0,cex=.8,labels = names(pvalues), col = taxa_pal[factor(module_color)] )


mypar(2.5,2.5,mar=c(2.5,2.5,2,6))
set.seed(1)
UMAP_picrust <- uwot::umap(t(named_picrust_data),n_neighbors = 20,
                 metric = "correlation",min_dist = .2,spread = .4,
                 repulsion_strength = .4,negative_sample_rate = 3)
plot(UMAP_picrust,bg=pal[factor(group_annotation[sample_use])],main="UMAP on top functional profile",
     pch=21,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(factor(group_annotation)),xjust = 0,yjust = 1,
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1,xpd=T)
```

**Figure 2. Identification and characterization of vaginal bacterial communities.**
(**a**) Schematic representation of ########## .
(**b**) Schematic representation of ########## .
(**c**) Schematic representation of ########## .
(**d**) Schematic representation of ########## .



```{r, echo = FALSE, eval = TRUE}
genus_lacto.fun <- function(df){
  genus_lacto <- df %>%
    mutate(Genus_lacto = .$Genus, .after=Seq_ID) %>% 
    mutate(Genus_lacto = ifelse(grepl("crispatus",.$Species), paste(.$Genus_lacto, "crispatus/acidophilus"), 
                                ifelse(grepl("reuteri",.$Species), paste(.$Genus_lacto, "reuteri/oris/frumenti/antri"),
                                  ifelse(grepl("gasseri",.$Species), paste(.$Genus_lacto,"gasseri/johnsonii/taiwanensis"),
                                    ifelse(grepl("murinus",.$Species), paste(.$Genus_lacto,"murinus/animalis/apodemi/salivarius"),
                                      ifelse(grepl("plantarum",.$Species), 
                                           paste(.$Genus_lacto, "plantarum/fabifermentans/composti/paraplantarum/graminis/fuchuensis"),
                                        ifelse(grepl("Lactobacillus",.$Genus), paste(.$Genus, .$Species), 
                                              as.character(.$Genus_lacto))))))) )
  return(genus_lacto)
} 

picrust_taxa <- read_csv("../../../results/SeqID_to_gen-lacto_tax-oth.csv") %>%
  column_to_rownames(var = "Seq_ID") %>%
  mutate(Genus_lacto = ifelse(is.na(.$Genus_lacto), "NA", .$Genus_lacto))
  #mutate(Genus_lacto = ifelse(is.na(.$Genus_lacto), "NA_", .$Genus_lacto))


# for Genus_lacto taxonomy:
taxa <- read_csv("../../../results/SeqID_to_gen-lacto_tax-oth.csv") %>%
  select(-Sequence, -Seq_ID, -taxa_other, -Species) %>%
  unique() %>%
  filter(!(is.na(.$Genus_lacto))) %>%
  filter(!(.$Genus == "Granulicatella" & .$Family == "Enterococcaceae")) %>%
  column_to_rownames(., var = "Genus_lacto")
# for other taxa
#taxa <- read.csv2("../../../results/taxonomy.csv",row.names = 1,stringsAsFactors = F,as.is = T)
dim(picrust_taxa)
dim(picrust_KO_seq)

#Convert picrust to the same bacterial names in the manuscript
temp <- as.matrix(picrust_taxa)
temp[is.na(temp)] <- "NA"
sps <- apply( temp[,8:9] ,1,function(x) paste(x,collapse ="_"))
not_complete <- grepl("_NA",sps)
taxa <- apply( temp ,1,function(x) paste(x,collapse ="_"))
taxa[not_complete] <- sub("_NA.*","",taxa[not_complete])
taxa[not_complete] <- sub(".*_","other ",taxa[not_complete])
taxa[!not_complete] <- sps[!not_complete]
taxa <- sub("_"," ",taxa)
length(taxa)
picrust_taxa$final_name <- taxa


dim(picrust_data)
dim(picrust_KO_seq)

common <- rownames(picrust_data)[rownames(picrust_data) %in% colnames(picrust_KO_seq)]
res <- t( t(as.matrix(picrust_data[common,])) %*% t(as.matrix(picrust_KO_seq[,common])) )
picrust_KO_seq <- picrust_KO_seq[,common]

picrust_KO_seq[1:100,1:100]

picrust_KO_seq <- rowsum(picrust_KO_seq , picrust_taxa[ match( rownames(picrust_KO_seq) , rownames(picrust_taxa)  ), params])
picrust_KO_seq <- as.matrix(picrust_KO_seq)

picrust_KO_seq[,1]

sub(" [(].*","",names(pvalues))[1]

#p <- sub(" [(].*","",names(pvalues))
#x <- p 

ress <- sapply( sub(" [(].*","",names(pvalues)) , function(x){
  inn <- colnames(picrust_KO_seq) [ colnames(picrust_KO_seq) %in% names(gmt_list[[x]] ) ]
  temp <- picrust_KO_seq[,inn]
  temp <- temp[ order(rowMeans(temp),decreasing = T) ,  ]
  common_bacs <- rownames(SNN_bacteria)[rownames(SNN_bacteria) %in% rownames(temp)]
  temp <- rowsum(temp[common_bacs , ] , group = bac_communities [ rownames(SNN_bacteria) %in% common_bacs ] )
  return( temp )
} )

ress

mypar(1,2,mar=c(4,10,2,4))
x <- sapply(ress,function(x){
  rowMeans(x)
})
x [ is.nan(x) ] <- 0
x <- t(x)

sizes <- sapply(ress,function(x){
  return( rowSums( x > 0) / ncol(x) )
})
sizes [ is.nan(sizes) ] <- 0
sizes <- t(sizes)

plot( sort(rep(1:ncol(x),nrow(x))),
      rep(nrow(x):1,ncol(x)),
        col=colorRampPalette(c("grey95","firebrick"))(91)[cut( c(x) , breaks = seq(0,15,length.out = 92) )],pch=16 ,
      cex= sizes*2 + 0.2, ylim=c(0.5,nrow(x)+.5),yaxs="i",
      axes=F,xlab="",ylab="", frame=T)
srt=30
text(1:ncol(x), par("usr")[3] - (par("usr")[4])/200, labels = colnames(x), srt = srt,
     adj = c(ifelse(srt==0,.5,ifelse(srt==90,1,1)),ifelse(srt==0,1,ifelse(srt==90,.5,1))),
     xpd = TRUE, cex=.6)
text(par("usr")[1] - (par("usr")[2])/200, nrow(x):1 , labels = rownames(x), srt = 0, adj = c(1,0.5), xpd = TRUE, cex=.6)

add_scale_legend(pal = colorRampPalette(c("grey95","firebrick"))(91),labels = c(min(x),round(max(x)) ))
add_size_legend(labels = c("0%","50%","100%"),max.cex = 2.2, min.cex = .2)

```




\clearpage

## Figure 3
## individual group comparison DGE
```{r}
TRX <- datasets_all_samples[["Tissue_RNAseq_V3_normalized"]]
TRX_counts <- round(2^TRX - 1)
TRX_counts <- TRX_counts [ rowSums(TRX_counts>0)>= 1 , ]
sample_use <- colnames(TRX_counts)[colSums(TRX_counts)!=0]
TRX_counts <- TRX_counts[,sample_use]

library(edgeR)
#group <- "TissueGroupM4"
#group <- "group_annotation"
edgR_dge.fun <- function(group, confound) {
  groups <- metadata[sample_use, group]
  
  if(confound==FALSE) {
    c <- "_no_conf"
    design <- model.matrix(~  groups, data=metadata[sample_use,])} # without confounders
  else{
    c <- "_w_conf"
    # design <- model.matrix(~groups+HIVstatus+Contraception+age+sexwork_months,
    #                      data=metadata[sample_use,])
    design <- model.matrix(~groups+HIVstatus+Contraception,
                         data=metadata[sample_use,])
  }
  #design <- model.matrix(~0 + groups+Contraception+HIVstatus+age+sexwork_months, 
                         #data=metadata[sample_use,]) # without intercept
  
  colnames(design) <- sub("groups","",colnames(design))
  colnames(design) <- sub("(Intercept)","Intercept",colnames(design))
  
  y <- DGEList(counts=TRX_counts[,sample_use], remove.zeros = T)
  y <- calcNormFactors(y,method = "TMM")
  y <- estimateGLMCommonDisp(y,design)
  y <- estimateGLMTagwiseDisp(y,design)
  fit <- glmFit(y, design)
  
  #lrt <- glmLRT(fit,coef=1:ncol(design)) # without intercept
  lrt <- glmLRT(fit,coef=2:5) # with intercept
  top <- topTags(lrt,adjust.method = "BH",n = "all",sort.by = "p.value")[[1]]
  colnames(top) <- sub(group,"",colnames(top))
  top <- cbind(LogFC.intercept=0,top) # with intercept
  top <-  rownames_to_column(top, var = "Genes")
  
  saveRDS( y , paste0("../../../results/TRX_EdgeR_estimations_",group,c,".rds" ))
  #write.xlsx(top, file=paste0("../../../results/DGE/","dge_heatmap_",group,c,".xlsx"))
  write.xlsx(top, file=paste0("../../../results/DGE/","dge_heatmap_",group,"_2:5_conf_2.xlsx"))
  #write.csv2( top , paste0("../../../results/DGE/TRX_EdgeR_top_dge_",group,".csv" ))
  
return(list(design=design, y=y, fit=fit, lrt=lrt, top=top, c=c))
}

dge_M4 <- edgR_dge.fun("ModifiedM4", confound=T)
#dge_M4_nonconfound <- edgR_dge.fun("ModifiedM4", confound=F)
#dge_TissueM4 <- edgR_dge.fun("TissueGroupM4", confound=T)

```


```{r, echo = FALSE, eval = TRUE}
# Define Layout and set labels
layout(matrix(c(1) , ncol= 1,byrow = T),
       heights = c(1))
figlabels <- letters

# group_name <- "ModifiedM4"
# top <- dge_M4$top
# y <- dge_M4$y
# c <- dge_M4$c
gene_mod_enrich.fun <- function(top, y, c, db, group_name, cuttoff){
  p_val <- pull(top, "PValue")
  top_dge <- (p_val < 0.01) #& (rowSums( abs(top[,grep("logFC",colnames(top))]) >= log2(1.5) ) >= 1)
  top_dge <- top$Genes[ top_dge ]
  
  top_TRX_counts <- cpm(y,normalized.lib.sizes = T,log = T)[top_dge, ]
  top_TRX_counts <- t(apply(top_TRX_counts,1,function(x){scale(x,T,T)}))
  top_TRX_counts[top_TRX_counts > 5] <- 5
  colnames(top_TRX_counts) <- colnames(TRX_counts)
  
  
  h <- hclust( as.dist( (1- cor(t(top_TRX_counts)))/2 ), method = "ward.D2")
  cuttoff <- cuttoff[names(cuttoff) == db]
  gene_module <- cutree(h, h=cuttoff)
  
  source("../enrichment_function.R")
  if(db == "GO"){
  gmt_list <- fgsea::gmtPathways("../../../supplementary_files/c5.bp.v6.2.symbols.gmt.txt")}else{
  gmt_list <- fgsea::gmtPathways("../../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")}
  res_list <- lapply( unique( gene_module[h$order] ),function(x){
    temp <- compute_enrichment(genes = names(gene_module)[gene_module==x],
                              gmt_list = gmt_list,
                              min_terms_pathway = 10,
                              max_terms_pathway = 300,
                              min_overlap = 3,
                              sort_by_pvalue = T)
    temp <- temp[!grepl("REGULATION",rownames(temp)),]
    return(temp)
  } )
  res <- map(res_list, ~rownames_to_column(.x, var = "Terms")) %>% 
    set_names(., paste0("gene_module_",  seq_along(res_list)))
  #write.xlsx(res, file=paste0("../../../results/ENRICHMENT/",db,"_enrichment_heatmap_",group_name,c,".xlsx"))
  write.xlsx(res, file=paste0("../../../results/ENRICHMENT/",db,"_enrichment_heatmap_",group_name,"_2:5_conf_2.xlsx"))
  
  return(list(res_list=res_list, top_TRX_counts=top_TRX_counts, 
              gene_module=gene_module, cuttoff=cuttoff, h=h, c=c))
}

db <- c("GO", "KEGG")
heatmap_cvlM4 <- map(db, ~gene_mod_enrich.fun(dge_M4$top,dge_M4$y,dge_M4$c,
                                              .x,"ModifiedM4", 
                                              #cuttoff= c(GO=4, KEGG=3.5)
                                              cuttoff= c(GO=2.5, KEGG=3)
                                              )) %>% set_names(db)

heatmap_cvlM4_nonconfound <- map(db,~gene_mod_enrich.fun(dge_M4_nonconfound$top,
                                                      dge_M4_nonconfound$y,
                                                      dge_M4_nonconfound$c,
                                                      .x,"ModifiedM4",
                                                      cuttoff= c(GO=2.5, KEGG=2)
                                                      )) %>% set_names(db)

# heatmap_tissueM4 <- map(db, ~gene_mod_enrich.fun(dge_TissueM4$top,
#                                                  dge_TissueM4$y,
#                                                  .x,"TissueM4", 
#                                                  cuttoff=)) %>% set_names(db)

###########
#   GO    #
###########
gr <- heatmap_cvlM4_nonconfound
gr <- heatmap_cvlM4

name <- "ModifiedM4"
groups <- metadata[sample_use, name]
conf <- pluck(gr, "GO", "c")
h <- pluck(gr, "GO", "h")
cuttoff <- pluck(gr, "GO", "cuttoff")
res_list <- pluck(gr, "GO", "res_list")
top_TRX_counts <- pluck(gr, "GO", "top_TRX_counts")
gene_module <- pluck(gr, "GO", "gene_module")


layout(matrix(c(1,2,3,4,
              1,2,3,4),
              nrow = 2,ncol = 4,byrow = T),widths = c(1,2,.4,6)) #1,2,.3,8
## Heatmap
par(mar=c(4,.1,2,.1))
plot( rev(as.dendrogram(h)) , xlim=c(max(h$height),-.2),horiz = T, leaflab = "none",xaxs="i",yaxs="i",axes=T,xlab="height")
# rect.dendrogram(rev(as.dendrogram(h)),h = 2,horiz = TRUE)
abline(v=cuttoff,xpd=F,col="red",lty=2)
points( rep(-.1,length(gene_module)),
        seq(length(gene_module),1,length.out = length(gene_module)),
        col=taxa_pal[factor(gene_module[h$order])],
        pch=16,cex=.5,xpd=T)


image( t(top_TRX_counts[h$order,order(groups)][nrow(top_TRX_counts):1,]),
       col = colorRampPalette(c("navy","navy","grey95","firebrick4","firebrick4") )(99),
       breaks = seq(-5,5,length.out = 100),axes=F , xlab="patient groups")

## Meta info bars
points( seq(0,1,length.out = length(sample_use)),rep(par("usr")[4],length(sample_use)),
        col=pal[factor(groups)[ order(groups)]],
        pch=16)
points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.005,
        col= c("red","orange","grey")[as.numeric(metadata[sample_use,"BV_Diagnosis_v3"])][order( groups )],
        pch=16,xpd=T) 
points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.015,
        col= c("grey","red")[as.numeric(metadata[sample_use,"HIVstatus"])][order(groups)],
        pch=16,xpd=T)

## Gene module enrichment
names(res_list) <- unique( gene_module[h$order] )

pvalues <- unlist(lapply(res_list,function(x){ -log10(x$pvalue) [1:5] }))
pvalues[is.na(pvalues)] <- 0
terms <- unlist(lapply(res_list,function(x){ rownames(x) [1:5] }))
#terms <- unlist(lapply(res_list,function(x){ x$Terms [1:5] }))
terms[is.na(terms)] <- ""
genes <- unlist(lapply(res_list,function(x){ x$genes [1:5] }))
genes[is.na(genes)] <- ""

pvalues <- setNames(pvalues,paste0(terms," (",genes,")"))
module_color <- unlist(lapply(names(res_list),function(x){ rep(x,5) }))

temp <- factor( rev(gene_module), levels = unique( gene_module[h$order] ))
plot_sankey( data.frame(temp,temp), pal = taxa_pal[as.numeric(levels(temp))] , use_w2 = F , plot_labels = T ,gapv = .01,gap2v = 0 ,xaxs="i",yaxs="i",plot_weights = F)

par(mar=c(4,.1,2,20))

barplot( rev(pvalues) , horiz = T , col= taxa_pal[factor(rev(module_color))],
         xaxs="i",yaxs="i",xlim=c(0,1.2*max(pvalues)),ylab="",xlab="-log10(pvalue)",las=1,names.arg = "",xpd = FALSE)
points( rev(pvalues), seq(1,length(pvalues) )*1.2-.5,pch=21,bg=taxa_pal[factor(rev(module_color))] )
text( rev(pvalues)+1.5 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,xpd=T,
      labels = rev(names(pvalues)))
text( rev(pvalues)+1.5 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,
      labels = rev(sub(" [(].*","",names(pvalues)) ),col = paste0(taxa_pal[factor(rev(module_color))],90) ,xpd=T)
# text( rep(max(pvalues)+.5,length(pvalues)) , seq(length(pvalues),1 )*1.2-.5,
#       adj=0,cex=.8,labels = names(pvalues), col = taxa_pal[factor(module_color)] )

#dev.copy2pdf(file=paste0("../../../results/Plots/","2:5_DGE_heatmap_GO_",name,conf,".pdf"), 
dev.copy2pdf(file=paste0("../../../results/Plots/","2:5_DGE_heatmap_GO_","M4_conf_2",".pdf"),
    width = 6.5,
    height = 6, paper = "a4" #units = "cm", res = 300 
    )

###########
#  KEGG   #
###########
gr <- heatmap_cvlM4_nonconfound
gr <- heatmap_cvlM4

name <- "ModifiedM4"
groups <- metadata[sample_use, name]
conf <- pluck(gr, "KEGG", "c")
cuttoff <- pluck(gr, "KEGG", "cuttoff")
h <- pluck(gr, "KEGG", "h")
res_list <- pluck(gr, "KEGG", "res_list")
top_TRX_counts <- pluck(gr, "KEGG", "top_TRX_counts")
gene_module <- pluck(gr, "KEGG", "gene_module")


layout(matrix(c(1,2,3,4,
              1,2,3,4),
              nrow = 2,ncol = 4,byrow = T),widths = c(1,2,.4,6)) #1,2,.3,8
par(mar=c(4,.1,2,.1))

## Heatmap
plot( rev(as.dendrogram(h)) , xlim=c(max(h$height),-.2),horiz = T, leaflab = "none",xaxs="i",yaxs="i",axes=T,xlab="height")
# rect.dendrogram(rev(as.dendrogram(h)),h = 2,horiz = TRUE)
abline(v=cuttoff,xpd=F,col="red",lty=2)
points( rep(-.1,length(gene_module)),
        seq(length(gene_module),1,length.out = length(gene_module)),
        col=taxa_pal[factor(gene_module[h$order])],
        pch=16,cex=.5,xpd=T)


image( t(top_TRX_counts[h$order,][nrow(top_TRX_counts):1,order(groups)]),
       col = colorRampPalette(c("navy","navy","grey95","firebrick4","firebrick4") )(99),
       breaks = seq(-5,5,length.out = 100),axes=F , xlab="patient groups")

## Meta info bars
points( seq(0,1,length.out = length(sample_use)),rep(par("usr")[4],length(sample_use)),
        col=pal[factor(groups)[ order(groups)]],
        pch=16)
points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.005,
        col= c("red","orange","grey")[as.numeric(metadata[sample_use,"BV_Diagnosis_v3"])][order( groups )],
        pch=16,xpd=T)
points( seq(0,1,length.out = ncol(top_TRX_counts)),rep(par("usr")[3],ncol(top_TRX_counts))-.015,
        col= c("grey","red")[as.numeric(metadata[sample_use,"HIVstatus"])][order(groups)],
        pch=16,xpd=T)

## Gene module enrichment
names(res_list) <- unique( gene_module[h$order] )

pvalues <- unlist(lapply(res_list,function(x){ -log10(x$pvalue) [1:5] }))
pvalues[is.na(pvalues)] <- 0
terms <- unlist(lapply(res_list,function(x){ rownames(x) [1:5] }))
#terms <- unlist(lapply(res_list,function(x){ x$Terms [1:5] }))
terms[is.na(terms)] <- ""
genes <- unlist(lapply(res_list,function(x){ x$genes [1:5] }))
genes[is.na(genes)] <- ""

pvalues <- setNames(pvalues,paste0(terms," (",genes,")"))
module_color <- unlist(lapply(names(res_list),function(x){ rep(x,5) }))

temp <- factor( rev(gene_module), levels = unique( gene_module[h$order] ))
plot_sankey( data.frame(temp,temp), pal = taxa_pal[as.numeric(levels(temp))] , use_w2 = F , plot_labels = T ,gapv = .01,gap2v = 0 ,xaxs="i",yaxs="i",plot_weights = F)

par(mar=c(4,.1,2,20))

barplot( rev(pvalues) , horiz = T , col= taxa_pal[factor(rev(module_color))],
         xaxs="i",yaxs="i",xlim=c(0,1.2*max(pvalues)),ylab="",xlab="-log10(pvalue)",las=1,names.arg = "")
points( rev(pvalues), seq(1,length(pvalues) )*1.2-.5,pch=21,bg=taxa_pal[factor(rev(module_color))] )
text( rev(pvalues)+1 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,xpd=T,
      labels = rev(names(pvalues)))
text( rev(pvalues)+1 , seq(1,length(pvalues) )*1.2-.5, adj=0,cex=.8,
      labels = rev(sub(" [(].*","",names(pvalues)) ),col = paste0(taxa_pal[factor(rev(module_color))],90) ,xpd=T)
# text( rep(max(pvalues)+.5,length(pvalues)) , seq(length(pvalues),1 )*1.2-.5,
#       adj=0,cex=.8,labels = names(pvalues), col = taxa_pal[factor(module_color)] )


#dev.copy2pdf(file=paste0("../../../results/Plots/","2:5_DGE_heatmap_KEGG_",name,conf,".pdf"), 
dev.copy2pdf(file=paste0("../../../results/Plots/","2:5_DGE_heatmap_KEGG_","M4_conf_2",".pdf"), 
    width = 6.5,
    height = 6, paper = "a4" #units = "cm", res = 300
)

###########
#  UMAP   #
###########
mypar(3,3,mar=c(3,3,2,6))
set.seed(1)
UMAP_TRX <- uwot::umap(t(top_TRX_counts),n_neighbors = 10,
                 metric = "correlation",min_dist = 0.1,spread = 5,
                 negative_sample_rate = 10)
plot(UMAP_TRX,bg=pal[factor(groups)],main="UMAP on top DGE RNAseq",
     pch=21,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       #legend = levels(factor(groups)),
       legend = sort(unique(groups)),
       xjust = 0,yjust = 1,
       bty = "n",pch = 21,pt.bg = pal,pt.cex = 1,xpd=T)


###############
#  HALLMARK   #
###############

NESse_list <- read.csv2("../../../results/Hallmark_NESse_list.csv",row.names = 1)
NESse_list <- lapply(unique(sub("[.].*","",colnames(NESse_list))),function(x){
  temp <- NESse_list[ grep( x, colnames(NESse_list)) ]
  colnames(temp) <- sub(".*[.]","",colnames(temp))
  temp[is.na(temp)]<-0
  return(t(temp))
})
pvalue_list <- read.csv2("../../../results/Hallmark_pvalues_list.csv",row.names = 1)
pvalue_list <- lapply(unique(sub("[.].*","",colnames(pvalue_list))),function(x){
  temp <- pvalue_list[ grep( x, colnames(pvalue_list)) ]
  colnames(temp) <- sub(".*[.]","",colnames(temp))
  temp[is.na(temp)]<-0
  return(t(temp))
})
x <- pvalue_list[[2]]
df <- data.frame( pathway=c(sapply(1:ncol(x),x=x,function(i,x){rownames(x)} )),
                  bc=c(sapply(colnames(x),x=x,function(i,x){ rep(i,nrow(x)) }) ),
                  NES=c(NESse_list[[1]]),
                  abs_NES=abs(c(NESse_list[[1]])),
                  pvalue=c(x))
df <- df[df$pvalue!=0,]
df <- df[df$pvalue<0.05,]
df <- df[df$NES!=0,]

df %>% group_by(bc)  %>% top_n(3,NES) -> top_per_bc
df %>% group_by(bc)  %>% top_n(-3,NES) ->  bottom_per_bc

as.data.frame(top_per_bc)

mypar(1,2,mar=c(4,10,2,4))
for(i in 1:2){
x <- NESse_list[[i]]
x [ pvalue_list[[i]] < 1 ] <- 0
x <- x[unique(top_per_bc$pathway),]
# h <- hclust( as.dist( (1-cor(t(x)))/2 ) , method = "ward.D2" )
# image( x[h$order,ncol(x):1],col=colorRampPalette(c("navy","grey95","firebrick"))(91),
#        breaks=seq(-5,5,length.out = 92),
#        axes=F,border=NA,xlab="",ylab="",line=.4,cex.main=1,font.main=1)

sizes <- pvalue_list[[i]]
sizes[sizes < 2] <- 0
sizes[sizes > 10] <- 10
sizes <- sizes / max(sizes)

plot( sort(rep(1:ncol(x),nrow(x))),
      rep(nrow(x):1,ncol(x)),
        col=colorRampPalette(c("navy","grey95","firebrick"))(91)[cut( c(x) , breaks = seq(-5,5,length.out = 92) )],pch=16 ,
      cex= sizes*2 + 0.2,
      axes=F,xlab="",ylab="", frame=T)
srt=30
text(1:ncol(x), par("usr")[3] - (par("usr")[4])/200, labels = colnames(x), srt = srt,
     adj = c(ifelse(srt==0,.5,ifelse(srt==90,1,1)),ifelse(srt==0,1,ifelse(srt==90,.5,1))),
     xpd = TRUE, cex=.6)
text(par("usr")[1] - (par("usr")[2])/200, 1:nrow(x) , labels = rownames(x), srt = 0, adj = c(1,0.5), xpd = TRUE, cex=.6)

add_scale_legend(pal = colorRampPalette(c("navy","grey95","firebrick"))(91),labels = c("-5","0","5"))
add_size_legend(labels = c("> 0.01","< 1e-6","< 1e-10"),max.cex = 2.2, min.cex = .2)
}

```

**Figure 1. Identification of patient groups.**
(**a**) Schematic representation of ########## .
(**b**) Schematic representation of ########## .
(**c**) Schematic representation of ########## .
(**d**) Schematic representation of ########## .


```{r}
## individual group comparision ##
# for information about how to get pairwise comparison se:
# edgeR Users Guide section 3.4.2 Blocking and 4.2.8  Differential expression
# https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
pairwise_dge.fun <- function(group, fit, design) {
  groups <- metadata[sample_use, group]
  levels <- as.character(sort(unique(groups)))
  
  # without intercept:
  n <- combn(levels,2) 
  g <- map_chr(seq_along(1:ncol(n)), ~paste(n[1,.x], n[2,.x], sep = "-"))
  s <- seq_along(1:ncol(n)) %>% set_names(g)
  #name <- map_chr(seq_along(1:ncol(n)), ~paste(str_extract(n[1,.x], "^g."), str_extract(n[2,.x], "^g."), sep = "_"))
  
  # with intercept:
  g <- sub(paste0("^(",levels[1], ")-(.*)"), '-\\2', g)
  
  CONTRASTS <- makeContrasts( contrasts=g,
                              levels = design )
  
  lrt <- map(s, ~ glmLRT(fit, contrast = CONTRASTS[,.x]))
  top <- map(lrt, ~topTags(.x,adjust.method = "BH",n = "all",sort.by = "p.value")[[1]]) %>% map(., ~rownames_to_column(.x, var = "Genes"))
  
  #saveRDS(top, file = "./TopTable_all_groups_06.04.2021.RDS")
  #saveRDS(top, file = "./TopTable_ModifiedM4_pairwise_17.05.2021.RDS")
return(list(lrt=lrt, top=top, CONTRASTS=CONTRASTS))
}
pairwise_dge_M4 <- pairwise_dge.fun("ModifiedM4", 
                                    fit=dge_M4$fit, 
                                    design=dge_M4$design )
# pairwise_dge_TissueM4 <- pairwise_dge.fun("TissueGroupM4", 
#                                           fit=dge_TissueM4$fit, 
#                                           design=dge_TissueM4$design)

# save TopTables 
l_top <- list(ModifiedM4 = pairwise_dge_M4$top) # TissueM4 = pairwise_dge_TissueM4$top
imap(l_top, ~write.xlsx(.x, #file=paste0("../../../results/DGE","dge_pairwise_",.y,".xlsx"))) # Excel
file=paste0("../../../results/DGE/","dge_pairwise_",.y,"_2:5_conf_2",".xlsx"))) # Excel
## end of individual group comparision ##

## individual GO/KEGG comparison ##
source("../enrichment_function.R")
gmt_list_KEGG <- fgsea::gmtPathways("../../../supplementary_files/c2.cp.kegg.v6.2.symbols.gmt.txt")
gmt_list_GO <- fgsea::gmtPathways("../../../supplementary_files/c5.bp.v6.2.symbols.gmt.txt")
enrich_list <- function(x, gmt_list){
  temp <- compute_enrichment(genes = x,
                            gmt_list = gmt_list,
                            min_terms_pathway = 10,
                            max_terms_pathway = 300,
                            min_overlap = 3,
                            sort_by_pvalue = T)
  temp <- temp[!grepl("REGULATION",rownames(temp)),]
  temp <- rownames_to_column(temp, var = "Terms")
  return(temp)
} 

#info_logFC <- map(top, ~table(abs(.x$logFC) > 1))
#info_PValue <- map(top, ~table(.$PValue < 0.01))

# up and down regulated genes seperatly
t_split <- modify_depth(l_top, 2, ~ .x %>%
        dplyr::filter(., .$PValue < 0.01) %>%
        #dplyr::filter(., abs(.$logFC) < 1) %>%
        mutate(., reg = ifelse(.$logFC > 0, "up", "down")) %>%
        split(., .$reg)
  )
# up and down regulated genes toghether
t <- modify_depth(l_top, 2,~ .x %>%
        dplyr::filter(., .$PValue < 0.01) #%>%
        #dplyr::filter(., abs(.$logFC) < 1) 
  )

## GO enrichment rownames(.x)
GO_enrichment <- map_depth(t, 2, ~enrich_list(.x$Genes, gmt_list_GO))
GO_enrichment_split <- map_depth(t_split, 3, ~enrich_list(.x$Genes, gmt_list_GO)) %>%
  map_depth(., 2, ~bind_rows(.x, .id = "regulation")) %>% 
  set_names(., paste0(names(.), "_split"))

## KEGG enrichment
KEGG_enrichment <- map_depth(t, 2, ~enrich_list(.x$Genes, gmt_list_KEGG))
KEGG_enrichment_split <- map_depth(t_split, 3, ~enrich_list(.x$Genes, gmt_list_KEGG)) %>%
  map_depth(., 2, ~bind_rows(.x, .id = "regulation")) %>% 
  set_names(., paste0(names(.), "_split"))
 
## save files
enrich_KEGG <- list(KEGG_enrichment, KEGG_enrichment_split) %>% flatten()
enrich_GO <- list(GO_enrichment, GO_enrichment_split) %>% flatten()
imap(enrich_KEGG, ~write.xlsx(.x, #file=paste0("../../../results/ENRICHMEN/T","GO_enrichment_pairwise_",.y,".xlsx"))) # Excel
file=paste0("../../../results/ENRICHMENT/","GO_enrichment_pairwise_",.y,"_2:5_conf_2",".xlsx"))) # Excel
imap(enrich_GO, ~write.xlsx(.x, #file=paste0("../../../results/ENRICHMENT/","KEGG_enrichment_pairwise_",.y,".xlsx"))) # Excel
file=paste0("../../../results/ENRICHMENT/","KEGG_enrichment_pairwise_",.y, "_2:5_conf_2",".xlsx"))) # Excel

## end of individual GO comparison ##
```



\clearpage



FIGURES (SUPPL)
================================================================================

## Figure S1

```{r, echo = FALSE, eval = TRUE, fig.asp=1}
figlabels <- letters
temp_legend <- NULL

layout(matrix(c(1,7,
                2,7,
                3,8,
                4,8,
                5,8,
                6,8) , ncol= 2,byrow = T),widths = c(4,1),heights = c(.5,1,.5,1,.5,1))

for(i in c("ASV_CVL_V3_normalized_batch_corrected",
           "ASV_CVL_V2_normalized_NOT_batch_corrected",
           "ASV_tissue_V3_normalized_batch_corrected")){
  x <- datasets_all_samples[[i]]

  par(mar=c(1,5,1,1))
  temp <- t(t(2^x-1)/colSums(2^x-1))

  shann <- vegan::diversity(t(temp),index = c("shannon") )
  simp <- vegan::diversity(t(temp),index = c("simpson") )
  invsimp <- vegan::diversity(t(temp),index = c("invsimpson") )
  divers <- as.matrix(rbind(shann,simp,invsimp) )
  divers[is.infinite(divers)] <- 0
  rownames(divers) <- c("Shannon","simpson","Inv Simpson")
  barlist( data = divers[,order(group_annotation)],
         main = "",
         genes = c("Shannon","simpson","Inv Simpson"),
         # clustering = group_annotation,
         draw_mean_lines=F,col = pal[factor(group_annotation[order(group_annotation)])])


  #add label
  add_letter(figlabels[1]); figlabels <- figlabels[-1]

  par(mar=c(3,5,1,1))
  # taxa <- read.csv2("../../../results/taxonomy.csv",row.names = 1,stringsAsFactors = F,as.is = T)
  # taxa[is.na(taxa)] <- "NA"
  # taxa <- taxa[rownames(temp),]
  # temp <- rowsum(temp , taxa$Class)
  temp[is.nan(temp)] <- 0

  if(is.null(temp_legend)){
    temp_legend <- temp   #Use the same legend from the 1st dataset
  }

  o <- order( rowSums(temp_legend) , decreasing = T)
  sample_o <- names(group_annotation)[ order( group_annotation )]
  temp <- temp[o,sample_o]*100
  # temp <- temp[-grep("NA",rownames(temp)),]
  temp <- rbind(temp[1:16,], Other = colSums(temp[17:nrow(temp),]) )

  taxa_pal <- c(RColorBrewer::brewer.pal(8,"Pastel2"),RColorBrewer::brewer.pal(8,"Pastel1"),"grey90")
  barplot( temp, las=2 , border = NA , yaxs="i", xaxs="i",ylab="abundance",main=sub("_nor.*","",i),font.main=1,
           col = taxa_pal[factor(rownames(temp),levels = rownames(temp))] ,names.arg = rep("",ncol(temp)))
  # legend(x = par("usr")[c(2)],
  #        y = par("usr")[c(4)],title.adj = 0,
  #        legend = rownames(temp),xjust = 0,yjust = 1,
  #        bty = "n",pch = 22,pt.cex = 2,cex = 1,xpd=T,
  #        pt.bg = taxa_pal)
  points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-2,
        col= c("grey","red", "black")[(!is.na(metadata$BV_Diagnosis_v3))+1][order( group_annotation )],
        pch=15,cex=1,xpd=T)
  text(par("usr")[1],par("usr")[3]-2,labels = "RepeatedPCR",pos=2,cex=.8,xpd=T)

  points( (1:ncol(temp))*1.2 -.5, rep(par("usr")[3],ncol(temp))-6,
        col= c("grey","red")[as.numeric(metadata$HIVstatus)][order( group_annotation )],
        pch=15,cex=1,xpd=T)
  text(par("usr")[1],par("usr")[3]-6,labels = "HIVstatus",pos=2,cex=.8,xpd=T)

}


par(mar=c(4,0,1,0))
empty_plot()
legend(x = par("usr")[c(1)],
         y = par("usr")[c(4)],title.adj = 0,
         legend = sort(unique(group_annotation)),
         #legend = levels(group_annotation),
         xjust = 0,yjust = 1,
         bty = "n",pch = 22,pt.cex = 2,cex = 1,xpd=T,title = "patient groups",
         pt.bg = pal)


empty_plot()
legend(x = par("usr")[c(1)],
         y = par("usr")[c(4)],title.adj = 0,
         legend = rownames(temp),xjust = 0,yjust = 1,
         bty = "n",pch = 22,pt.cex = 2,cex = 1,xpd=T,
          title="top abundant bacteria",
         pt.bg = taxa_pal)
```

\clearpage

## Figure S2

```{r, echo = FALSE, eval = TRUE, fig.asp=.5}
figlabels <- letters
temp_legend <- NULL

continuous_parameters <- c("age","Count_CD4_v2","Count_CD4_v3","Plasma_S_Prog_ng_mL_v3","Plasma_S_Estradiol_pg_mL_v3","Conc_T4_v2","Conc_T3_v2","Conc_Progesteron_v2","Conc_Estradiol_v2","Conc_Cortisol_v2","Freq_CD4_v2","Freq_CD4_v3","Freq_Lympho_v2","Freq_Lympho_v3","visit_clients7days_v2")
continuous_metadata <- metadata[,continuous_parameters]
# continuous_metadata <- cbind(continuous_metadata , t(module_means_per_dataset[[i]]) )


mypar(1,4,mar=c(12,1,1,0))

for(i in c("ASV_CVL_V3_normalized_batch_corrected",
           "ASV_CVL_V2_normalized_NOT_batch_corrected",
           "ASV_tissue_V3_normalized_batch_corrected")){
  metadata_cor <- cor(t(module_means_per_dataset[[i]]),continuous_metadata,use = "pairwise.complete.obs")
  # metadata_cor_test <- cor.test(t(module_means_per_dataset[[i]]),continuous_metadata)

  image( t(metadata_cor [nrow(metadata_cor):1,]),axes=F,
         breaks = seq(-1,1,length.out = 90),main=sub("_nor.*","",i),font.main=1,
         col=colorRampPalette(c("navy","grey95","firebrick"))(89) )
  # text(  par("usr")[2]+.02 ,seq(1,0,length.out = nrow(metadata_cor)) ,
  #        labels = rownames(metadata_cor), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
  text( seq(0,1,length.out = ncol(metadata_cor)) , par("usr")[3]-.02 ,
        labels = colnames(metadata_cor), srt = 90, adj = c(1,.5), xpd = TRUE, cex=1)

  #add label
  add_letter(figlabels[1]); figlabels <- figlabels[-1]
}
par(mar=c(12,0,1,0))
image( t(metadata_cor [nrow(metadata_cor):1,]),axes=F,
         breaks = seq(-1,1,length.out = 90),main="",
         col=colorRampPalette(c("white"))(89) )
text(  0 ,seq(1,0,length.out = nrow(metadata_cor)) ,
  labels = rownames(metadata_cor), srt = 0, adj = c(0,.5), xpd = TRUE, cex=1)
```

\clearpage

## Figure S2 v2

```{r, echo = FALSE, eval = TRUE, fig.asp=1}
figlabels <- letters
temp_legend <- NULL

mypar(1,4,mar=c(12,1,1,0))

for(k in c("ASV_CVL_V3_normalized_batch_corrected",
           "ASV_CVL_V2_normalized_NOT_batch_corrected",
           "ASV_tissue_V3_normalized_batch_corrected")){
res2 <- lapply( 1:nrow(module_means_per_dataset[[k]]),k=k, function(i,k){
  res <- lapply( c( colnames(clean_Ftest) , continuous_parameters ) , i=i,k=k,function(j,i,k) {
    model <- paste0("module_means_per_dataset[['",k,"']][",i,",] ~ ",j)
    m <- summary(lm( eval(parse(text=model)), data = metadata ))
    f <- m$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    temp <- c(m$r.squared,m$adj.r.squared,m$fstatistic,p)
    return(temp)
  })
  res <- t(as.data.frame(res))
  rownames(res) <- c( colnames(clean_Ftest) , continuous_parameters )
  colnames(res) <- c("r.squared","adj.r.squared","F-statistic","df1","df2","pvalue")
  return(res)
})
names(res2) <- rownames(module_means_per_dataset[[k]])

pvals <- sapply(res2, function(x){x[,"pvalue"]})
rs <- sapply(res2, function(x){x[,"r.squared"]})
rs[is.nan(rs)] <- 0
pvals[is.nan(pvals)] <- 1

sizes <- -log10( pvals )
sizes[sizes < 2] <- 0
sizes[sizes > 11] <- 10
sizes <- sizes / max(sizes)

plot(0,0, las=1,xlim=c(.5,ncol(rs)+.5),ylim=c(.5,nrow(rs)+.5),
       axes=F,ylab="",xlab="",main=gsub("^ASV_|_norm.*","",k),xaxs="i",yaxs="i", frame=T)
for(i in 1:ncol(rs)){
  lines(c(i,i),c(nrow(rs)+.5,0),col="grey95",lwd=.5,xpd=F)
}
for(i in 1:nrow(rs) ){
  lines(c(0,ncol(rs)+0.5),c(i,i),col="grey95",lwd=.5,xpd=F)
}
points(sort(rep(1:ncol(rs),nrow(rs))),
      rep(nrow(rs):1,ncol(rs)),
        col=colorRampPalette(c("navy","grey95","firebrick"))(91)[cut( c(rs) , breaks = seq(-1,1,length.out = 92) )],pch=16 ,
      cex= sizes*2 + 0.2)
srt=90
text(1:ncol(rs), par("usr")[3] - (par("usr")[4])/200, labels = colnames(rs), srt = srt,
     adj = c(ifelse(srt==0,.5,ifelse(srt==90,1,1)),ifelse(srt==0,1,ifelse(srt==90,.5,1))),
     xpd = TRUE, cex=.8)
add_letter(figlabels[1]); figlabels <- figlabels[-1]
}

empty_plot(xlim=c(.5,ncol(rs)+.5),ylim=c(.5,nrow(rs)+.5),xaxs="i",yaxs="i")
text(par("usr")[1] , 1:nrow(rs) , labels = rownames(rs), srt = 0, adj = c(0,0.5), xpd = TRUE, cex=.8)

add_scale_legend(pal = colorRampPalette(c("grey90","firebrick"))(91),labels = c("0","0.5","1"))
add_size_legend(labels = c("> 0.01","< 1e-5","< 1e-8","< 1e-11"),max.cex = 2.2, min.cex = .2)

```


\clearpage

## Figure S2 v3

```{r, echo = FALSE, eval = TRUE, fig.asp=.9}
metadata_parameters <- colnames(metadata)[-c(1:2)]

set.seed(1)
noise <- matrix(rnorm(prod(dim(clean_Ftest)),mean = 0,sd = 1e-10),nrow(clean_Ftest),ncol(clean_Ftest))
o <- hclust(as.dist( (1-cor(clean_Ftest+noise)/2) ),method = "ward.D2")$order

sizes <- as.matrix(clean_Ftest[o,o])
# sizes[sizes < 2] <- 0
sizes[sizes > 10] <- 10
sizes <- sizes / max(sizes)
sizes <- sizes * lower.tri(sizes)


mypar(1,1,mar=c(10,10,2,6))
plot(0,0, las=1,xlim=c(.5,ncol(sizes)+.5),ylim=c(.5,nrow(sizes)+.5),
       axes=F,ylab="",xlab="",main="",xaxs="i",yaxs="i", frame=T)
for(i in 1:ncol(sizes)){
  lines(c(i,i),c(nrow(sizes)+.5,0),col="grey95",lwd=.5,xpd=F)
}
for(i in 1:nrow(sizes) ){
  lines(c(0,ncol(sizes)+0.5),c(i,i),col="grey95",lwd=.5,xpd=F)
}
points(sort(rep(1:ncol(sizes),nrow(sizes))),
      rep(nrow(sizes):1,ncol(sizes)),
        col=colorRampPalette(c("grey95","orange3","firebrick","red"))(91)[cut( c(sizes) , breaks = seq(0,1,length.out = 92) )],pch=16 ,
      cex= sizes*1 + 0.2)
srt=90
text(1:ncol(sizes), par("usr")[3] - (par("usr")[4])/200, labels = colnames(sizes), srt = srt,
     adj = c(ifelse(srt==0,.5,ifelse(srt==90,1,1)),ifelse(srt==0,1,ifelse(srt==90,.5,1))),
     xpd = TRUE, cex=1)
text(par("usr")[1] , nrow(sizes):1 , labels = rownames(sizes), srt = 0, adj = c(1,0.5), xpd = TRUE, cex=1)

add_scale_legend(pal = colorRampPalette(c("grey95","orange3","firebrick","red"))(91),labels = c("> 0.01","< 1e-5","< 1e-8","< 1e-11"))
add_size_legend(labels = c("> 0.01","< 1e-5","< 1e-8","< 1e-11"),max.cex = 1.2, min.cex = .2)
```

\clearpage

## Figure S3

```{r, echo = FALSE, eval = TRUE}
figlabels <- letters
mypar(1,2,mar=c(3,3,3,10))

datasets <- datasets_all_samples[grep("ASV",names(datasets_all_samples))]
datasets <- lapply(datasets,t)
datasets <- t( as.data.frame(datasets) )

group_means <- sapply(levels(factor(group_annotation)), function(x){
  rowMeans( datasets[ , group_annotation == x ] )
})

d <- as.dist( (1 - cor(group_means)) / 2 )
h <- hclust( d , method = "ward.D2")
plot( as.dendrogram(h) , horiz=T, xlab="height")
points( rep(0 , length(h$labels)) , 1:length(h$labels) , pch=21, cex=1.5,
        bg=pal[factor(h$labels[ h$order ],levels = h$labels)] )
add_letter(figlabels[1]); figlabels <- figlabels[-1]



datasets <- datasets_all_samples[grep("ASV",names(datasets_all_samples))]
datasets <- as.data.frame(datasets)

group_means <- sapply(levels(factor(bac_communities)), function(x){
  colMeans( datasets[ bac_communities == x , ] )
})

d <- as.dist( (1 - cor(group_means)) / 2 )
h <- hclust( d , method = "ward.D2")
plot( as.dendrogram(h) , horiz=T, xlab="height")
points( rep(0 , length(h$labels)) , 1:length(h$labels) , pch=21, cex=1.5,
        bg=taxa_pal[factor(h$labels[ h$order ],levels = h$labels)] )
add_letter(figlabels[1]); figlabels <- figlabels[-1]
```

\clearpage

## Figure S4

```{r, echo = FALSE, eval = TRUE,fig.asp=1.3}
#layout( matrix(1:(16*8), 16,8,byrow = F) )
layout( matrix(1:(10*8), 10,8,byrow = F) )
par(mar=c(.5,.5,1,.5))

figlabels <- letters
for(j in names(module_means_per_dataset)){

  plot(UMAP_picrust,bg=pal[factor(group_annotation[sample_use])],
       main=gsub("ASV_|_norm.*","",j),
     pch=21,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F)
  add_letter(figlabels[1]); figlabels <- figlabels[-1]

  for(i in 1:nrow(module_means_per_dataset[[j]])){

    xx <- module_means_per_dataset[[j]][i,]
    xx <- xx / max(xx)
    o <- order(xx)
    plot(UMAP_picrust[o,],
         col=colorRampPalette( c("grey95","navy")) (99)[ 1 + (xx[o]*98) ],
         main="",
         pch=16,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F,font.main=1,cex.main=1)
  }
}
empty_plot()

text( par("usr")[1], mean(par("usr")[3:4]),
        adj=c(0,0.5),xpd=T,
        labels = "patient groups" )
for(i in 1:nrow(module_means_per_dataset[[j]])){
  empty_plot()
  text( par("usr")[1], mean(par("usr")[3:4]),
        adj=c(0,0.5),xpd=T,
        labels = sub( "_","\n",rownames(module_means_per_dataset[[j]])[i]) )
}

```


\clearpage

## Figure S5

```{r, echo = FALSE, eval = TRUE,fig.asp=1.3}
#layout( matrix(1:(16*8), 16,8,byrow = F) )
layout( matrix(1:(10*8), 10,8,byrow = F) )
par(mar=c(.5,.5,1,.5))


figlabels <- letters
for(j in names(module_means_per_dataset)){

  plot(UMAP_TRX,bg=pal[factor(group_annotation[sample_use])],main=gsub("ASV_|_norm.*","",j),
     pch=21,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F)
  add_letter(figlabels[1]); figlabels <- figlabels[-1]

  for(i in 1:nrow(module_means_per_dataset[[j]])){

    xx <- module_means_per_dataset[[j]][i,][sample_use]
    xx <- xx / max(xx)
    o <- order(xx)
    plot(UMAP_TRX[o,],
         col=colorRampPalette( c("grey95","navy")) (99)[ 1 + (xx[o]*98) ],
         main="",xpd=T,
         pch=16,xlab="UMAP1",ylab="UMAP2",frame=F,axes=F,font.main=1,cex.main=1)
  }
}

empty_plot()
text( par("usr")[1], mean(par("usr")[3:4]),
        adj=c(0,0.5),
        labels = "patient groups" )
for(i in 1:nrow(module_means_per_dataset[[j]])){
  empty_plot()
  text( par("usr")[1], mean(par("usr")[3:4]),
        adj=c(0,0.5),xpd=T,
        labels = sub( "_","\n",rownames(module_means_per_dataset[[j]])[i]) )
}

```


\clearpage

## Figure S6



```{r, echo = FALSE, eval = TRUE,fig.asp=1.2}
combined_bc <- rbind( top_per_bc  , bottom_per_bc )

mypar(8,2,mar=c(2,25,1,1))
for(i in sort(levels(combined_bc$bc)) ){
  barplot( c(combined_bc [combined_bc$bc == i,"NES" ])[[1]] , horiz = T,las=1,
           main=i,
           names.arg = as.character(combined_bc$pathway)[combined_bc$bc == i])
}
```



\clearpage

## Figure S7

```{r, echo = FALSE, eval = TRUE,fig.asp=.8}
figlabels <- letters
mypar(3,3,mar=c(4,12,2,4))
for( j in c("ASV_tissue_V3_normalized_batch_corrected",
        "ASV_CVL_V3_normalized_batch_corrected",
        "ASV_CVL_V2_normalized_NOT_batch_corrected")){
  temp_group <- as.character(group_annotation)
  temp_group[ is.na(temp_group) ] <- "NA"
  temp_group[ temp_group == "" ] <- "NA"
  temp_group <- factor(temp_group)

  plot_dots(data=module_means_per_dataset[[j]],srt=45,
          genes=rownames(module_means_per_dataset[[j]]),col = pal,
          clustering=temp_group, main=paste0(sub("_norm.*","",j)))
  add_letter(figlabels[1]); figlabels <- figlabels[-1]
  empty_plot()
  empty_plot()
}
```



\clearpage

## Figure S8

```{r, echo = FALSE, eval = TRUE,fig.asp=.5}
mypar(mar=c(10,6,1,1))

top_all <- lapply( datasets_all_samples[2:4] , function(x){
   sort(apply(x,1,var),decreasing = T)[1:20]} )
top <- unique( sub( ".*[.]" , "" , names(unlist(top_all)) ) )

x <- t(sapply(datasets_all_samples[2:4],function(x){
  c(t(x))
}))
colnames(x) <- c(sapply(rownames(datasets_all_samples[[2]]),
                        function(i){rep(i,ncol(datasets_all_samples[[2]]))}))
rownames(x) <- gsub("ASV_|_norm.*", "",rownames(x))

hist(x,breaks = 400)

top <- sort(sapply( unique(colnames(x)),
               function(i) { mean( x[,colnames(x)==i ] ) } ),decreasing = T)
top <- names(top)[1:30]


violist(data = x[,colnames(x)%in%top],
        genes = rownames(x),
        clustering = factor( colnames(x)[colnames(x)%in%top],levels = top) ,
        srt=45,
        transparency = 50,col = pal)
```


\clearpage

## Figure S9


```{r, echo = FALSE, eval = TRUE,fig.asp=.5}
figlabels <- letters

cytokine_data <- read.csv("../../../data/210125_Cytokines_visit3CVL.csv")
cytokine_data$ID <- metadata [ match( cytokine_data$PatID , as.character(metadata$PatID)  ) ,"ID" ]
cytokine_data <- cytokine_data[ !is.na(cytokine_data$ID) , ]
rownames(cytokine_data) <- cytokine_data$ID
cytokine_data <- cytokine_data[,grepl("CVL",colnames(cytokine_data))]
colnames(cytokine_data) <- sub("CVL_","",colnames(cytokine_data))
cytokine_data <- t(cytokine_data)
cytokine_data[is.na(cytokine_data)] <- 0.35
cytokine_data <- log2(cytokine_data+1)

groupings <- metadata$group_annotation [ match( colnames(cytokine_data) , as.character(metadata$ID)  ) ]
mypar(1,3,mar=c(6,5,1,5))

plot_dots(data=cytokine_data,
          srt=45,
          genes=rownames(cytokine_data),
          col = pal,
          clustering=groupings )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

barlist(data=cytokine_data,
          srt=45,
          genes=rownames(cytokine_data),
          col = pal,
          clustering=groupings )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

violist(data=cytokine_data,
          srt=45,
          genes=rownames(cytokine_data),
          col = pal, transparency = 50,
          clustering=groupings,smooth = 2 )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

```



\clearpage

## Figure S10


```{r, echo = FALSE, eval = TRUE,fig.asp=.8}
figlabels <- letters

IF_data <- read.csv("../../../data/200105_CD4Ecad_Visit3LAMIQ.csv")
IF_data$ID <- metadata [ match( IF_data$PatID , as.character(metadata$PatID)  ) ,"ID" ]
IF_data <- IF_data[ !is.na(IF_data$ID) , ]
rownames(IF_data) <- IF_data$ID
IF_data <- IF_data [ , -c(1:3)]
IF_data <- IF_data [ , -ncol(IF_data)]
IF_data <- t(IF_data)
# IF_data[is.na(IF_data)] <- 0
# IF_data <- log2(IF_data+1)

groupings <- metadata$group_annotation [ match( colnames(IF_data) , as.character(metadata$ID)  ) ]


mypar(1,3,mar=c(6,15,1,4))
plot_dots(data=IF_data,
          srt=45,
          genes=rownames(IF_data),
          col = pal,
          clustering=groupings )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

barlist(data=IF_data,
          srt=45,
          genes=rownames(IF_data),
          col = pal,
          clustering=groupings )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

violist(data=IF_data,
          srt=45,
          genes=rownames(IF_data),
          col = pal, transparency = 50,
          clustering=groupings,smooth = 2 )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

```



```{r, echo = FALSE, eval = TRUE,fig.asp=.5}

continuous_parameters <- c("age","Count_CD4_v2","Count_CD4_v3","Plasma_S_Prog_ng_mL_v3","Plasma_S_Estradiol_pg_mL_v3","Conc_T4_v2","Conc_T3_v2","Conc_Progesteron_v2","Conc_Estradiol_v2","Conc_Cortisol_v2","Freq_CD4_v2","Freq_CD4_v3","Freq_Lympho_v2","Freq_Lympho_v3","visit_clients7days_v2")


mypar(1,3,mar=c(6,15,1,4))
plot_dots(data=as.matrix(t(metadata[,continuous_parameters])),
          srt=45,
          genes=continuous_parameters,
          col = pal,
          clustering=metadata$group_annotation )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

barlist(data=as.matrix(t(metadata[,continuous_parameters])),
          srt=45,
          genes=continuous_parameters,
          col = pal,
          clustering=metadata$group_annotation )
add_letter(figlabels[1]); figlabels <- figlabels[-1]

# violist(data=as.matrix(t(metadata[,continuous_parameters])),
#           srt=45,
#           genes=continuous_parameters,
#           col = pal, transparency = 50,
#           clustering=metadata$group_annotation,smooth = 2 )
# add_letter(figlabels[1]); figlabels <- figlabels[-1]
# 
# m <-  na.omit(as.matrix(t(metadata[,continuous_parameters])))
# m <- metadata[,continuous_parameters] %>% as.data.frame() %>%
#   rownames_to_column(var = "ID") %>% pivot_longer(-ID, names_to= "Param", values_to= "Values") %>% pivot_wider(names_from = "ID", values_from="Values") %>%
#   column_to_rownames(var = "Param")
```


\clearpage

## Figure S11


```{r, echo = FALSE, eval = TRUE,fig.asp=1.2}
figlabels <- letters

g <- graph_from_adjacency_matrix(as.matrix(SNN_patients), mode = "undirected",diag = F,weighted = T)
set.seed(1)
l <- layout_with_fr(g,niter=3000,start.temp=30)


continuous_parameters <- c("age","Count_CD4_v2","Count_CD4_v3","Plasma_S_Prog_ng_mL_v3","Plasma_S_Estradiol_pg_mL_v3","Conc_T4_v2","Conc_T3_v2","Conc_Progesteron_v2","Conc_Estradiol_v2","Conc_Cortisol_v2","Freq_CD4_v2","Freq_CD4_v3","Freq_Lympho_v2","Freq_Lympho_v3","visit_clients7days_v2")


mypar(6,6,mar=c(1,1,1,1))

plot(l,
       col=pal[factor(group_annotation[sample_use])],
       main="SNN Louvain clustering",xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
empty_plot()
legend("topleft",title.adj = 0,
       legend = sort(unique(metadata$group_annotation)),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)

for(i in continuous_parameters){
  xx <- metadata[,i]
  xx[is.na(xx)] <- 0
  xx <- xx / max(xx)
  o <- order(xx)
  plot(l[o,],
       col=colorRampPalette( c("grey95","navy")) (99)[ 1 + (xx[o]*98) ],
       main=i,xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)

}

```



\clearpage

## Figure S12


```{r, echo = FALSE, eval = TRUE,fig.asp=1}
figlabels <- letters

g <- graph_from_adjacency_matrix(as.matrix(SNN_patients), mode = "undirected",diag = F,weighted = T)
set.seed(1)
l <- layout_with_fr(g,niter=3000,start.temp=30)


categorical_parameters <- c("Antibiotics","Antibiotics_v2","Antibiotics_v3",
                         "Bacterio_KOH_v2","Bacterio_KOH_v3","Bacterio_TV_v3",
                         "BV_Clue_Cells_v2","BV_Clue_Cells_v3",
                         "BV_Diagnosis_v2","BV_Diagnosis_v3",
                         "BV_Lactobacillus_v2","BV_Lactobacillus_v3",
                         "BV_Monbilicus_v2","BV_Monbilicus_v3",
                         "BV_Vaginal_Garda_v2","BV_Vaginal_Garda_v3",
                         "blood_cervix_v2","blood_cervix_v3",
                         "cervical_inflammation_v2","cervical_inflammation_v3",
                         "HIVstatus","Contraception",
                         "CT_v2","CT_v3","regular_partner",
                         "vaginal_douching_v2","vaginal_douching_v3")

mypar(6,5,mar=c(1,1,1,8))
plot(l,
       col=pal[factor(group_annotation[sample_use])],
       main="SNN Louvain clustering",xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(factor(group_annotation[sample_use])),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)

for(i in categorical_parameters){
  xx <- factor(metadata[,i])
  plot(l[,],
       col=pal[xx],
       main=i,xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(xx),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)
}
```


\clearpage

## Figure S13

```{r, echo = FALSE, eval = TRUE,fig.asp=.5}

continuous_parameters <- c("age","Count_CD4_v2","Count_CD4_v3","Plasma_S_Prog_ng_mL_v3","Plasma_S_Estradiol_pg_mL_v3","Conc_T4_v2","Conc_T3_v2","Conc_Progesteron_v2","Conc_Estradiol_v2","Conc_Cortisol_v2","Freq_CD4_v2","Freq_CD4_v3","Freq_Lympho_v2","Freq_Lympho_v3","visit_clients7days_v2")

common <- colnames(top_TRX_counts)[colnames(top_TRX_counts) %in% rownames(metadata)]

mypar(6,6,mar=c(1,1,1,1))

plot(UMAP_TRX,
       col=pal[factor(group_annotation[sample_use])],
       main=""UMAP trx"",xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
empty_plot()
legend("topleft",title.adj = 0,
       legend = levels(factor(group_annotation[sample_use])),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)

for(i in continuous_parameters){
  xx <- metadata[common,i]
  xx[is.na(xx)] <- 0
  xx <- xx / max(xx)
  o <- order(xx)
  plot(UMAP_TRX[o,],
       col=colorRampPalette( c("grey95","navy")) (99)[ 1 + (xx[o]*98) ],
       main=i,xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)

}
```


\clearpage

## Figure S14

```{r, echo = FALSE, eval = TRUE,fig.asp=1}
figlabels <- letters

# g <- graph_from_adjacency_matrix(as.matrix(SNN_patients), mode = "undirected",diag = F,weighted = T)
# set.seed(1)
# l <- layout_with_fr(g,niter=3000,start.temp=30)
# 

categorical_parameters <- c("Antibiotics","Antibiotics_v2","Antibiotics_v3",
                         "Bacterio_KOH_v2","Bacterio_KOH_v3","Bacterio_TV_v3",
                         "BV_Clue_Cells_v2","BV_Clue_Cells_v3",
                         "BV_Diagnosis_v2","BV_Diagnosis_v3",
                         "BV_Lactobacillus_v2","BV_Lactobacillus_v3",
                         "BV_Monbilicus_v2","BV_Monbilicus_v3",
                         "BV_Vaginal_Garda_v2","BV_Vaginal_Garda_v3",
                         "blood_cervix_v2","blood_cervix_v3",
                         "cervical_inflammation_v2","cervical_inflammation_v3",
                         "HIVstatus","Contraception",
                         "CT_v2","CT_v3","regular_partner",
                         "vaginal_douching_v2","vaginal_douching_v3")

common <- colnames(top_TRX_counts)[colnames(top_TRX_counts) %in% rownames(metadata)]

mypar(6,5,mar=c(1,1,1,8))
plot(UMAP_TRX,
       col=pal[factor(group_annotation[sample_use])],
       main="UMAP trx",xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(factor(group_annotation[sample_use])),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)

for(i in categorical_parameters){
  xx <- factor(metadata[common,i])
  plot(UMAP_TRX[,],
       col=pal[xx],
       main=i,xpd=T,
       pch=16,xlab="",ylab="",frame=F,axes=F,font.main=1,cex.main=1)
legend(par("usr")[2],par("usr")[4],title.adj = 0,
       legend = levels(xx),xjust = 0,yjust = 1,
       bty = "n",pch = 16,col = pal,pt.cex = 1,xpd=T)
}
```




\clearpage

## Figure S15


```{r, echo = FALSE, eval = F, fig.width=6, fig.height=10}
paths_p <- c('../../../data/phyloseq_boston_r1.RDS',   # CVL V3
             '../../../data/phyloseq_boston_r2.RDS',   # Tissue V3 & CVL V2
             '../../../data/phyloseq_Sthlm.RDS')       # CVL V2, only v4 region

n <- c("boston_r1", "boston_r2", "Sthlm")
ps_list <- map(paths_p, ~readRDS(.x)) %>% set_names(n)

# remove controls and duplicates
ps_filt <- map(ps_list, ~ .x %>%
  subset_samples(., Include_final == "yes"))

# separate tissue from CVLv2 samples in the second Boston run
CVLv2 <- ps_filt[["boston_r2"]] %>% subset_samples(., Sample_type == "CVL" & Visit == "v2")
tissue <- ps_filt[["boston_r2"]] %>% subset_samples(., Sample_type == "tissue" & Visit == "v3")

# merge the lists
ps_types <- c(ps_filt, "CVLv2"=CVLv2, "tissue"=tissue)
# Get ASV tables
seqtabs <- map(ps_types, ~data.frame(.x@otu_table@.Data))

# Get rarefaction curves
rare_runs <- map(seqtabs, ~ .x %>%
             rarecurve(., step = 20, label = FALSE)
)

ggplot_df_fun <- function(rare_obj, count) {
  # Add sample names
  names(rare_obj) <- c(paste0("sample",sprintf("%03.0f", 1:length(rare_obj))))

  # Coerce data into "long" form.
  rare_df <- imap_dfr(rare_obj, ~ .x %>%
              as_tibble(., rownames = "Sample size") %>%
              mutate(Sample = .y) %>%
              mutate(`Sample size` = as.numeric(gsub("N","",`Sample size`)))
  )
  # Get nr. of counts in each sample
  c <- formatC(count, format="f", big.mark = " ", digits=0)
  rare_df_col <- rare_df %>%
    dplyr::rename(ASV = "value") %>%
    group_by(Sample) %>%
    nest() %>%
    mutate(Counts = map(data, ~max(.$`Sample size`))) %>%
    unnest(c(data, Counts)) %>%
    mutate(Counts = ifelse(Counts > count, paste0(">", c), paste0("<", c)))

  return(rare_df_col)
}

# Define some plot variables
count <- c(33000, 10000, 33000, 20000, 2500)
limit <- c(65000, 65000, 65000, 65000, 10000)
title <- c("Run B1. CVLv3", "Run B2. CVLv2 + TISSUEv3","Run Sthlm CVLv2", "Run B2. CVLv2", "Run B2. TISSUEv3")

# Extract data frame from the rarecurve object
rare_df_col<- map2(rare_runs, count, ~ggplot_df_fun(.x, .y))
# Create Plot
rare_plot <- map(rare_df_col,
              ~ggplot(.x, aes(x = `Sample size`, y = ASV)) +
              theme_bw() +
              scale_color_manual(values=c('tomato', alpha('black', 0.1))) +
              geom_line(aes(color = Counts, group = Sample)) +
              theme(plot.title = element_text(hjust = 0.5),
                    plot.margin=unit(c(0,0,0.5,0),"cm"))
            )
raremax <- map(seqtabs, ~min(rowSums(.x)))
r_plot <- pmap(list(rare_plot, raremax, limit, title),
               ~ ..1 + xlim(c(0,..3)) +
               labs(title=paste0("Sequencing ", ..4)) +
               geom_vline(xintercept = ..2,
                          linetype='dashed',
                          color='slategray' )
)

# Save png
#imap(r_plot, ~ggsave(filename = paste0('Rarefaction_plot_', .y, '.png'), plot = .x,
#        path = paste0("../../../results/"),
#        width = 10, height = 5,)
#)

```

```{r Print rare curves, eval=F, fig.width=6, fig.height=10, fig.align="center"}
ggarrange(r_plot[[1]] + theme(axis.title.x=element_blank()),
          r_plot[[3]] + theme(axis.title.x=element_blank()),
          r_plot[[4]] + theme(axis.title.x=element_blank()),
          r_plot[[5]],
          ncol = 1,
          labels = c("A","B","C","D")
          )
```

**Supp. Figure 10. Rarefraction curves**. **A** Boston 1 sequencing run, **B** Sthlm sequencing run and **C** and **D** shows Boston sequencing run 2 split in two plots showing the CVL visit 2 samples and the the TISSUE visit 3 samples respectively












\clearpage

TABLES (MAIN)
================================================================================

## Table 1

```{r, echo = FALSE, eval = TRUE}
```

\clearpage

## Table 2

```{r, echo = FALSE, eval = TRUE}
```

\clearpage

## Table 3

```{r, echo = FALSE, eval = TRUE}
```

\clearpage

TABLES (SUPPL)
================================================================================

## Table S1

```{r, echo = FALSE, eval = TRUE}
```

\clearpage

## Table S2

```{r, echo = FALSE, eval = TRUE}
```

\clearpage

## Table S3

```{r, echo = FALSE, eval = TRUE}
```

\clearpage
