---
title: "Suppl. Figure 2. Differential bacterial abundance across Tissue and Luminal microbiome"
geometry: "left=2cm,right=2cm,top=2cm,bottom=2cm"
header-includes: 
- \usepackage{float}
editor_options: 
  chunk_output_type: console
knit: (function(inputFile, out_dir, ...) {
    source("../../code/knit_function.R");
    custom_knit(inputFile, "../../lab_book/SupplFigure2/", ...)
    })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.path="./Suppl.Figures/",
  fig.align = "center",
  fig.process = function(filename){
    new_filename <- stringr::str_remove(string = filename, 
                                        pattern = "-1")
    fs::file_move(path = filename, new_path = new_filename)
    ifelse(fs::file_exists(new_filename), new_filename, filename)
})
# setwd("/Users/vilkal/work/Brolidens_work/Projects/broliden_5325/reports/manuscript")
```

```{r Load data, message=FALSE, warning=FALSE}
##################
# LOAD LIBRARIES #
##################
suppressWarnings({suppressMessages({suppressPackageStartupMessages({
  #remotes::install_github("czarnewski/niceRplots",force=T)
  library(niceRplots)
  library(openxlsx)
})  })  })

#############
# LODA DATA #
#############
datasets_all_samples <- readRDS("../../results/03_normalize_data_output/datasets_all_samples.RDS")
metadata <- read.csv("../../data/metadata.csv", row.names = 1, stringsAsFactors = F)

# setwd("/Users/vilkal/work/Brolidens_work/Projects/Gabriella_repo/reports/rmarkdown/manuscript")
# datasets_all_samples <- readRDS("../../../results/datasets_all_samples.RDS")
# metadata <- read.csv2("../../../results/metadata_integration.csv",row.names = 1, stringsAsFactors = F)
group_annotation <- factor(setNames(metadata$Luminal_gr,metadata$ID))

#Graph construction  
fct <- .25 # FC threshold for differential expression
pvt <- 0.01 # Pvalue threshold for differential expression
min_pct <- .05 # minimun level of detected bacteria in each sample group

#################
# COLOR PALETTS #
#################
pal <- c( "#0072B2", "#009E73","#D55E00", "#CC79A7", "#E69F00", "#999999")
```


```{r Suppl.Fig.2 Diff_bact_abundance, warning=FALSE, message=FALSE, fig.width=10, fig.height=9}
# all_microbiome <- cbind(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]],
#                         datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])
# datasets <- factor(c( rep("Tissue",ncol(datasets_all_samples[["ASV_tissue_V3_normalized_batch_corrected"]])) , rep("Luminal",ncol(datasets_all_samples[["ASV_CVL_V3_normalized_batch_corrected"]])) ))
# datasets <- datasets[ colSums(all_microbiome)>0 ]
# all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

all_microbiome <- cbind(datasets_all_samples[["ASV_Tissue_normalized"]],
                        datasets_all_samples[["ASV_Luminal_normalized"]])
datasets <- factor(c( rep("Tissue",ncol(datasets_all_samples[["ASV_Tissue_normalized"]])) , rep("Luminal",ncol(datasets_all_samples[["ASV_Luminal_normalized"]])) ))
datasets <- datasets[ colSums(all_microbiome)>0 ]
all_microbiome <- all_microbiome[,colSums(all_microbiome)>0]

NN <- min(table(datasets))


res <- data.frame( matrix(0,nrow = 1,ncol = 6) )
for(i in levels(datasets)){
  for(j in rownames(all_microbiome) ){
    
    a <- all_microbiome[j,datasets == i]
    b <- all_microbiome[j,datasets != i]
    
    # calc % of samples that have a presence of bacteria j 
    perc1 <- sum(all_microbiome[j,datasets == i]>0) / sum(datasets == i) 
    perc2 <- sum(all_microbiome[j,datasets != i]>0) / sum(datasets != i)
    
    temp <- wilcox.test(x=a, y=b)
    
    fc <- log2( (mean(a)+1e-3) / (mean(b)+1e-3) )
    
    res <- rbind(res, setNames(c(j,i,fc,perc1,perc2,unlist(temp)[2] ),
                          c("bacteria","cluster","fc","perc.1","perc.2","pvalue")) )
    colnames(res) <- c("bacteria","cluster","fc","perc.1","perc.2","pvalue")
  }
}
res <- res[-1,]
res$pvalue <- as.numeric(res$pvalue)
res$perc.1 <- as.numeric(res$perc.1)
res$perc.2 <- as.numeric(res$perc.2)
res$perc.diff <- res$perc.1 - res$perc.2

res$fc <- as.numeric(res$fc)
res$FDR <- p.adjust(res$pvalue)
res <- res[order(res$pvalue),]
res <- res[abs(res$fc) > fct,]
res <- res[abs(res$pvalue) < pvt,]
res <- res[ (res$perc.1 > min_pct) | (res$perc.2 > min_pct) ,]
# dim(res)

#write.xlsx(res, file=paste0("../../../results/","DBA_across_Luminal_vs_Tissue",".xlsx"))

ord <- getcluster(data = all_microbiome, genes =  unique(as.character(res$bacteria)), clustering =  datasets)
ord <- ord[unique(as.character(res$bacteria))]

### A & B
####################################
# DIFFERENTIAL BACTERIAL ABUNDANCE #
####################################
figlabels <- letters
par(mfrow=c(1,2),mar=c(2,12,1,3.1))
plot_dots( all_microbiome, genes = names(sort(ord)) , clustering = datasets,
           show_grid = T,cex.main=1,font.main=1,
           cex.row = .8,cex.col = .8,srt = 0)
add_letter("a")
barlist( all_microbiome, names(sort(ord)) , clustering = datasets,
         show_grid = T,cex.main=1,font.main=1,
         cex.axis = .8,srt = 0)
add_letter("b")

dev.off()
```

**Suppl. Figure 2.** Differential bacterial abundance across tissue and CVL microbiome week3 datasets. The results are shown both as **a)** Dot plots and **b)** barplots Bacteria with log2FC above `r fct` and p-value below `r pvt` (from the wilcoxon test) were considered significant and were sorted by the highest expression. The colour scale indicates the difference in total abundance between the datsets as a proportion, where Max is the higest abundace of the two datasets and the other becomes a proportion of this value.The size of the dots indicates the average abundance of the given bacteria in the given dataset. 
